<!DOCTYPE html>



  


<html class="theme-next muse use-motion" lang="zh-Hans">
<head><meta name="generator" content="Hexo 3.8.0">
  <meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
<meta name="theme-color" content="#222">



  
  
    
    
  <script src="/lib/pace/pace.min.js?v=1.0.2"></script>
  <link href="/lib/pace/pace-theme-minimal.min.css?v=1.0.2" rel="stylesheet">







<meta http-equiv="Cache-Control" content="no-transform">
<meta http-equiv="Cache-Control" content="no-siteapp">
















  
  
    
  
  <link href="//cdn.jsdelivr.net/fancybox/2.1.5/jquery.fancybox.min.css" rel="stylesheet" type="text/css">







  

<link href="//maxcdn.bootstrapcdn.com/font-awesome/4.6.2/css/font-awesome.min.css" rel="stylesheet" type="text/css">

<link href="/css/main.css?v=5.1.4" rel="stylesheet" type="text/css">


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.4">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.4">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.4">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.4" color="#222">





  <meta name="keywords" content="RNASeq数据下载处理全流程,">





  <link rel="alternate" href="/atom.xml" title="sylvia" type="application/atom+xml">






<meta name="description" content="RNASeq数据下载处理全流程最近需要重新跑一个物种的RNASeq数据处理，所以趁此机会记录一下。ssh win10@pcrent365.com -A 22522 一、下载数据首先在NCBI官网上选择SRA数据库，输入想要的数据关键词搜索“Zea mays drought”  点击发送结果到数据框，点击Metadata下载SraRunTable。 接着对数据条进行筛选，可手动可写代码，看个人需求。">
<meta name="keywords" content="RNASeq数据下载处理全流程">
<meta property="og:type" content="article">
<meta property="og:title" content="RNASeq数据下载处理全流程">
<meta property="og:url" content="https://janvia.github.io/2019/01/07/RNASeq数据下载处理全流程/index.html">
<meta property="og:site_name" content="sylvia">
<meta property="og:description" content="RNASeq数据下载处理全流程最近需要重新跑一个物种的RNASeq数据处理，所以趁此机会记录一下。ssh win10@pcrent365.com -A 22522 一、下载数据首先在NCBI官网上选择SRA数据库，输入想要的数据关键词搜索“Zea mays drought”  点击发送结果到数据框，点击Metadata下载SraRunTable。 接着对数据条进行筛选，可手动可写代码，看个人需求。">
<meta property="og:locale" content="zh-Hans">
<meta property="og:image" content="https://jiangvia.oss-cn-shenzhen.aliyuncs.com/dxz/RNAseq.png">
<meta property="og:updated_time" content="2024-05-29T02:02:57.000Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="RNASeq数据下载处理全流程">
<meta name="twitter:description" content="RNASeq数据下载处理全流程最近需要重新跑一个物种的RNASeq数据处理，所以趁此机会记录一下。ssh win10@pcrent365.com -A 22522 一、下载数据首先在NCBI官网上选择SRA数据库，输入想要的数据关键词搜索“Zea mays drought”  点击发送结果到数据框，点击Metadata下载SraRunTable。 接着对数据条进行筛选，可手动可写代码，看个人需求。">
<meta name="twitter:image" content="https://jiangvia.oss-cn-shenzhen.aliyuncs.com/dxz/RNAseq.png">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Muse',
    version: '5.1.4',
    sidebar: {"position":"left","display":"hide","offset":12,"b2t":false,"scrollpercent":true,"onmobile":true},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '博主'
    },
    algolia: {
      applicationID: '9CNLK347LC',
      apiKey: '5d000d330392ce1b44b614961e14f53e',
      indexName: 'dev_janviablog',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="https://janvia.github.io/2019/01/07/RNASeq数据下载处理全流程/">






  <title>RNASeq数据下载处理全流程 | sylvia</title>
  








</head>

<body itemscope="" itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope="" itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">sylvia</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle">Viva La Vida</p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br>
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br>
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br>
            
            分类
          </a>
        </li>
      

      
        <li class="menu-item menu-item-search">
          
            <a href="javascript:;" class="popup-trigger">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br>
            
            搜索
          </a>
        </li>
      
    </ul>
  

  
    <div class="site-search">
      
  <div class="popup search-popup local-search-popup">
  <div class="local-search-header clearfix">
    <span class="search-icon">
      <i class="fa fa-search"></i>
    </span>
    <span class="popup-btn-close">
      <i class="fa fa-times-circle"></i>
    </span>
    <div class="local-search-input-wrapper">
      <input autocomplete="off" placeholder="搜索..." spellcheck="false" type="text" id="local-search-input">
    </div>
  </div>
  <div id="local-search-result"></div>
</div>



    </div>
  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-normal" itemscope="" itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="https://janvia.github.io/2019/01/07/RNASeq数据下载处理全流程/">

    <span hidden itemprop="author" itemscope="" itemtype="http://schema.org/Person">
      <meta itemprop="name" content="sylvia">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/uploads/avatar.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope="" itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="sylvia">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">RNASeq数据下载处理全流程</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-01-07T17:07:32+08:00">
                2019-01-07
              </time>
            

            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope="" itemtype="http://schema.org/Thing">
                  <a href="/categories/dxz/" itemprop="url" rel="index">
                    <span itemprop="name">dxz</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
                <a href="/2019/01/07/RNASeq数据下载处理全流程/#comments" itemprop="discussionUrl">
                  <span class="post-comments-count disqus-comment-count" data-disqus-identifier="2019/01/07/RNASeq数据下载处理全流程/" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  13.7k
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  77
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h1 id="RNASeq数据下载处理全流程"><a href="#RNASeq数据下载处理全流程" class="headerlink" title="RNASeq数据下载处理全流程"></a>RNASeq数据下载处理全流程</h1><p>最近需要重新跑一个物种的RNASeq数据处理，所以趁此机会记录一下。ssh win10@pcrent365.com -A 22522</p>
<h2 id="一、下载数据"><a href="#一、下载数据" class="headerlink" title="一、下载数据"></a>一、下载数据</h2><p>首先在NCBI官网上选择SRA数据库，输入想要的数据关键词搜索“Zea mays drought”</p>
<p><img src="https://jiangvia.oss-cn-shenzhen.aliyuncs.com/dxz/RNAseq.png" alt="image-20240415210211959"></p>
<p>点击发送结果到数据框，点击Metadata下载SraRunTable。</p>
<p>接着对数据条进行筛选，可手动可写代码，看个人需求。筛选 ‘Assay Type’ 列为 ‘RNA Seq’ 的行和‘LibraryLayout’为‘PAIRED’</p>
<figure class="highlight python"><table><tr><td class="code"><pre><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">from</span> icecream <span class="keyword">import</span> install,ic</span><br><span class="line">install()</span><br><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">filter_rna_seq</span><span class="params">(input_file, output_file)</span>:</span></span><br><span class="line">    <span class="comment"># 读取输入文件</span></span><br><span class="line">    df = pd.read_csv(input_file)</span><br><span class="line">    ic(df.shape)</span><br><span class="line">    ic(df.columns)</span><br><span class="line">    <span class="comment"># 筛选 'Assay Type' 列为 'RNA Seq' 的行</span></span><br><span class="line">    ic(df[<span class="string">'Assay Type'</span>] == <span class="string">'RNA Seq'</span>)</span><br><span class="line">    filtered_df = df[df[<span class="string">'Assay Type'</span>] == <span class="string">'RNA-Seq'</span>]</span><br><span class="line">    ic(filtered_df.shape)</span><br><span class="line">    ic(filtered_df[<span class="string">'LibraryLayout'</span>].unique())</span><br><span class="line">    filtered_df = filtered_df[filtered_df[<span class="string">'LibraryLayout'</span>] == <span class="string">'PAIRED'</span>]</span><br><span class="line">    ic(filtered_df.shape)</span><br><span class="line">    <span class="comment"># 将筛选后的数据写入到输出文件</span></span><br><span class="line">    filtered_df.to_csv(output_file, index=<span class="keyword">False</span>, sep=<span class="string">','</span>, encoding=<span class="string">'utf-8'</span>)                    </span><br><span class="line">filter_rna_seq(<span class="string">'/home/win/16t2/study/deep_learning/gene/RNA_Seq/SRR_download_list/Zea_Mays_SraRunTable.txt'</span>, <span class="string">'/home/win/16t2/study/deep_learning/gene/RNA_Seq/SRR_download_list/Zea_Mays_filter_SraRunTable.csv'</span>)</span><br></pre></td></tr></table></figure>
<p>筛选完之后对srr_id生成数据下载地址</p>
<figure class="highlight python"><table><tr><td class="code"><pre><span class="line"><span class="keyword">import</span> os</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line">url_template = <span class="string">"https://sra-pub-run-odp.s3.amazonaws.com/sra/&#123;srr_id&#125;/&#123;srr_id&#125;"</span></span><br><span class="line"></span><br><span class="line">download_urls = []</span><br><span class="line">srrid_data = pd.read_csv(<span class="string">'/home/win/16t2/study/deep_learning/gene/RNA_Seq/SRR_download_list/Zea_Mays_filter_SraRunTable.csv'</span>)</span><br><span class="line">srr_ids = srrid_data[<span class="string">'Run'</span>]</span><br><span class="line"><span class="keyword">for</span> srr_id <span class="keyword">in</span> srr_ids:</span><br><span class="line">    srr_id = srr_id.strip()  <span class="comment"># 去除每行末尾可能存在的空白字符</span></span><br><span class="line">    download_url = url_template.format(srr_id=srr_id)</span><br><span class="line">    download_urls.append(download_url)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 打印或者保存到文件中</span></span><br><span class="line"><span class="keyword">for</span> url <span class="keyword">in</span> download_urls:</span><br><span class="line">    print(url)</span><br><span class="line"></span><br><span class="line"><span class="keyword">with</span> open(<span class="string">f'/home/win/16t2/study/deep_learning/gene/RNA_Seq/SRR_download_list/zea_mays_download_urls_sra.txt'</span>, <span class="string">'a'</span>) <span class="keyword">as</span> file_out:</span><br><span class="line">    <span class="keyword">for</span> url <span class="keyword">in</span> download_urls:</span><br><span class="line">        file_out.write(url + <span class="string">'\n'</span>)</span><br></pre></td></tr></table></figure>
<p>windows的话得到下载链接之后复制到迅雷下载，一次复制一千条，正常在几十Mb/s</p>
<p>linux下采用多线程下载,正常在400Mb/s(看网络情况)，存储够的话最好是提前下载</p>
<figure class="highlight bash"><table><tr><td class="code"><pre><span class="line">aria2c -i <span class="string">"/home/win/4T/GeneDL/RNASeq/SRR/zea_mays_download_urls_sra.txt"</span> -x 16</span><br></pre></td></tr></table></figure>
<p>下载也可参考第四步在全流程里面一起做</p>
<h2 id="二、参考基因组索引-建立"><a href="#二、参考基因组索引-建立" class="headerlink" title="二、参考基因组索引 建立"></a>二、参考基因组索引 建立</h2><p>首先在ensembl plants数据库下载对应物种的参考基因组序列.fa和注释文件.gff3，利用hisat进行索引建立</p>
<figure class="highlight bash"><table><tr><td class="code"><pre><span class="line">hisat2-build -p 4 <span class="string">"/home/win/4T/GeneDL/RNASeq/Reference/Zea_mays.fa"</span> Zea_mays</span><br></pre></td></tr></table></figure>
<p>比对完之后，将.ht2文件放在一个文件夹Zea_mays下，后续比对需要用到这个文件夹</p>
<figure class="highlight bash"><table><tr><td class="code"><pre><span class="line">mv *.ht2 /home/win/4T/GeneDL/RNASeq/Reference/Zea_mays/</span><br></pre></td></tr></table></figure>
<h2 id="三、数据全流程自动处理"><a href="#三、数据全流程自动处理" class="headerlink" title="三、数据全流程自动处理"></a>三、数据全流程自动处理</h2><p>这里将数据下载，解压 ，比对，定量的步骤采用自动化的方式处理，需提前搭建包含parallel-fastq-dump、hisat2、samtools、featureCounts等软件。</p>
<p>环境配置</p>
<figure class="highlight bash"><table><tr><td class="code"><pre><span class="line">conda create -n rnaseq python=3.7</span><br><span class="line">conda install -c bioconda hisat2</span><br><span class="line">conda install parallel-fastq-dump</span><br><span class="line">conda install samtools</span><br><span class="line">sudo apt-get install fastp</span><br><span class="line">conda install subread</span><br><span class="line">pip install tqdm</span><br><span class="line">pip install pandas</span><br><span class="line">pip install icecream</span><br><span class="line">sudo apt-get update</span><br><span class="line">sudo apt-get install aria2</span><br></pre></td></tr></table></figure>
<p>采用shell自动化命令</p>
<figure class="highlight bash"><table><tr><td class="code"><pre><span class="line">nj=8 <span class="comment">#多少个进程</span></span><br><span class="line">run_cmd=/mnt/d/RNASeq/src/run.pl  <span class="comment">#多线程的启动命令</span></span><br><span class="line"></span><br><span class="line">dir=/mnt/d/RNASeq/data  <span class="comment">#数据路径</span></span><br><span class="line"><span class="comment"># data_dir=/home/win/16t2/study/deep_learning/gene/RNA_Seq/FastQ/paired </span></span><br><span class="line"><span class="comment"># mkdir -p $&#123;dir&#125;/logs   #保存中间数据的目录</span></span><br><span class="line"></span><br><span class="line">cat <span class="string">"/mnt/d/RNASeq/data/Indica_download_urls_sra.txt"</span> &gt; <span class="variable">$&#123;dir&#125;</span>/SRR_data.scp</span><br><span class="line"><span class="comment">#ls $&#123;data_dir&#125; | grep _1.fastq.gz &gt; $&#123;dir&#125;/data.scp   #写总的数据文件信息</span></span><br><span class="line"></span><br><span class="line">nutt=$(&lt;<span class="variable">$&#123;dir&#125;</span>/SRR_data.scp wc -l)   <span class="comment">#获取多少个文件</span></span><br><span class="line"><span class="built_in">echo</span> <span class="variable">$nutt</span></span><br><span class="line">nj=$((nj&lt;nutt?nj:nutt))   <span class="comment">#判断文件数有没有小于线程数</span></span><br><span class="line"></span><br><span class="line">split_scps=<span class="string">""</span></span><br><span class="line"><span class="keyword">for</span> n <span class="keyword">in</span> $(seq <span class="variable">$&#123;nj&#125;</span>); <span class="keyword">do</span></span><br><span class="line">    split_scps=<span class="string">"<span class="variable">$&#123;split_scps&#125;</span> <span class="variable">$&#123;dir&#125;</span>/logs/SRR_data.scp.<span class="variable">$&#123;n&#125;</span>"</span>   <span class="comment">#获取输出文件名</span></span><br><span class="line">    <span class="built_in">echo</span> <span class="variable">$split_scps</span></span><br><span class="line"><span class="keyword">done</span></span><br><span class="line"><span class="comment">#echo $split_scps</span></span><br><span class="line"></span><br><span class="line">/home/win/16t2/study/deep_learning/gene/course_try/1Sequence/utils/split_scp.pl <span class="variable">$&#123;dir&#125;</span>/SRR_data.scp <span class="variable">$&#123;split_scps&#125;</span>  <span class="comment">#将总的文件划分成nj个子文件</span></span><br><span class="line"></span><br><span class="line"><span class="variable">$&#123;run_cmd&#125;</span> <span class="string">"JOB=1:<span class="variable">$&#123;nj&#125;</span>"</span> <span class="string">"<span class="variable">$&#123;dir&#125;</span>/logs/process.JOB.log"</span>\</span><br><span class="line">    python3 /mnt/d/RNASeq/src/RNASeq_allprocess.py <span class="variable">$&#123;dir&#125;</span>/logs/SRR_data.scp.JOB</span><br></pre></td></tr></table></figure>
<p>里面的py文件内容如下</p>
<figure class="highlight python"><table><tr><td class="code"><pre><span class="line"><span class="comment"># import sys</span></span><br><span class="line"><span class="comment"># import shutil</span></span><br><span class="line"><span class="comment"># import time</span></span><br><span class="line"><span class="comment"># from tqdm import tqdm</span></span><br><span class="line"><span class="comment"># scp_path = sys.argv[1]</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># with open(scp_path,'r') as f:</span></span><br><span class="line"><span class="comment">#     lines = f.readlines()</span></span><br><span class="line"><span class="comment">#     lines = [line.strip() for line in lines]</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># for line in tqdm(lines):</span></span><br><span class="line"><span class="comment">#     #time.sleep(0.1)</span></span><br><span class="line"><span class="comment">#     print(f"processed &#123;line&#125;")</span></span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"><span class="keyword">import</span> os</span><br><span class="line"><span class="keyword">import</span> subprocess</span><br><span class="line"><span class="keyword">from</span> tqdm <span class="keyword">import</span> tqdm</span><br><span class="line"><span class="keyword">import</span> logging</span><br><span class="line"><span class="keyword">import</span> glob</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line">logger  = logging.getLogger(__file__)</span><br><span class="line"><span class="keyword">from</span> icecream <span class="keyword">import</span> install,ic</span><br><span class="line">install()</span><br><span class="line"><span class="keyword">import</span> fnmatch</span><br><span class="line">species=<span class="string">'Zea_mays'</span> <span class="comment">#需与制定的参考基因组文件夹名称一致</span></span><br><span class="line">folder_path = <span class="string">'/home/win10/3T/RNASeq/data'</span></span><br><span class="line"><span class="comment"># files = os.listdir(folder_path)</span></span><br><span class="line">SRRdir  = <span class="string">f'<span class="subst">&#123;folder_path&#125;</span>/SRR'</span></span><br><span class="line">scp_path = sys.argv[<span class="number">1</span>]</span><br><span class="line">FastQdir = <span class="string">f'<span class="subst">&#123;folder_path&#125;</span>/FastQ'</span></span><br><span class="line">SAMdir =  <span class="string">f'<span class="subst">&#123;folder_path&#125;</span>/SAM'</span></span><br><span class="line">BAMdir =<span class="string">f'<span class="subst">&#123;folder_path&#125;</span>/BAM'</span></span><br><span class="line">Referencedir = <span class="string">f'<span class="subst">&#123;folder_path&#125;</span>/Reference'</span></span><br><span class="line">Countdir = <span class="string">f'<span class="subst">&#123;folder_path&#125;</span>/Count'</span></span><br><span class="line">Countlogdir = <span class="string">f'<span class="subst">&#123;folder_path&#125;</span>/Countlog'</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">with</span> open(scp_path,<span class="string">'r'</span>) <span class="keyword">as</span> f:</span><br><span class="line">    lines = f.readlines()</span><br><span class="line">    lines = [line.strip() <span class="keyword">for</span> line <span class="keyword">in</span> lines]</span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> file <span class="keyword">in</span> lines:</span><br><span class="line">    </span><br><span class="line">    SRR_id = file.split(<span class="string">'/'</span>)[<span class="number">-1</span>]</span><br><span class="line">    logger.info(<span class="string">f'process <span class="subst">&#123;SRR_id&#125;</span>'</span>)</span><br><span class="line">    <span class="keyword">if</span> <span class="string">f'<span class="subst">&#123;SRR_id&#125;</span>_featureCounts.tsv'</span> <span class="keyword">in</span> os.listdir(Countdir):</span><br><span class="line">        ic(<span class="string">f'<span class="subst">&#123;SRR_id&#125;</span>已定量'</span>)</span><br><span class="line">        <span class="keyword">continue</span></span><br><span class="line">    <span class="keyword">if</span> SRR_id <span class="keyword">in</span> os.listdir(SRRdir):</span><br><span class="line">        ic(<span class="string">f'<span class="subst">&#123;SRR_id&#125;</span>已下载'</span>)</span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        cmd1 = <span class="string">f'aria2c <span class="subst">&#123;file&#125;</span> -d <span class="subst">&#123;SRRdir&#125;</span>'</span></span><br><span class="line">        <span class="comment">#cmd1 = f'curl -o &#123;SRRdir&#125;/&#123;SRR_id&#125; &#123;file&#125;'</span></span><br><span class="line">        <span class="comment">#cmd1 = f'wget -O &#123;SRRdir&#125;/&#123;SRR_id&#125; &#123;file&#125;'</span></span><br><span class="line"></span><br><span class="line">        <span class="comment">#cmd1 = f'wget &#123;file&#125; -P &#123;SRRdir&#125;'</span></span><br><span class="line">        <span class="comment">#os.system(cmd1)</span></span><br><span class="line">        subprocess.run(cmd1, shell=<span class="keyword">True</span>, check=<span class="keyword">True</span>)</span><br><span class="line"></span><br><span class="line">    matching_files = glob.glob(os.path.join(FastQdir, <span class="string">f'<span class="subst">&#123;SRR_id&#125;</span>*'</span>))</span><br><span class="line">    <span class="keyword">if</span> matching_files:</span><br><span class="line">        ic(<span class="string">f'<span class="subst">&#123;SRR_id&#125;</span>已分解'</span>)</span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        cmd2 = <span class="string">f'parallel-fastq-dump -t 24 -O <span class="subst">&#123;FastQdir&#125;</span> --split-3  --gzip --sra-id <span class="subst">&#123;SRRdir&#125;</span>/<span class="subst">&#123;SRR_id&#125;</span>'</span></span><br><span class="line">        <span class="comment">#os.system(cmd2)</span></span><br><span class="line">        subprocess.run(cmd2, shell=<span class="keyword">True</span>, check=<span class="keyword">True</span>)</span><br><span class="line">    </span><br><span class="line">    <span class="comment"># if f'&#123;SRR_id&#125;.sam' in os.listdir(SAMdir):</span></span><br><span class="line">    <span class="comment">#     ic(f'&#123;SRR_id&#125;已比对')</span></span><br><span class="line">    <span class="comment"># else:</span></span><br><span class="line">    <span class="comment">#     if sum(1 for filename in os.listdir(FastQdir) if fnmatch.fnmatch(filename, f"&#123;SRR_id&#125;*")) == 1:</span></span><br><span class="line">    <span class="comment">#         cmd4 = f'hisat2 -x &#123;Referencedir&#125;/&#123;species&#125;/&#123;species&#125; -p 4 -U &#123;FastQdir&#125;/&#123;SRR_id&#125;.fastq.gz -S &#123;SAMdir&#125;/&#123;SRR_id&#125;.sam'</span></span><br><span class="line">    <span class="comment">#         subprocess.run(cmd4, shell=True, check=True)</span></span><br><span class="line">    <span class="comment">#         logger.info("hisat2 end")</span></span><br><span class="line">    <span class="comment">#     else:</span></span><br><span class="line">    <span class="comment">#         cmd4 = f'hisat2 -x &#123;Referencedir&#125;/&#123;species&#125;/&#123;species&#125; -p 4 -1 &#123;FastQdir&#125;/&#123;SRR_id&#125;_1.fastq.gz -2 &#123;FastQdir&#125;/&#123;SRR_id&#125;_2.fastq.gz -S &#123;SAMdir&#125;/&#123;SRR_id&#125;.sam'</span></span><br><span class="line">    <span class="comment">#         subprocess.run(cmd4, shell=True, check=True)</span></span><br><span class="line">    <span class="comment">#         logger.info("hisat2 end")</span></span><br><span class="line"></span><br><span class="line">    <span class="comment"># if f'&#123;SRR_id&#125;.bam' in os.listdir(BAMdir):</span></span><br><span class="line">    <span class="comment">#     ic(f'&#123;SRR_id&#125;已排序')</span></span><br><span class="line">    <span class="comment"># else:</span></span><br><span class="line">    <span class="comment">#     cmd6 = f'samtools sort -n -@ 5 &#123;SAMdir&#125;/&#123;SRR_id&#125;.sam -o &#123;BAMdir&#125;/&#123;SRR_id&#125;.bam'</span></span><br><span class="line">    <span class="comment">#     subprocess.run(cmd6, shell=True, check=True)</span></span><br><span class="line">    <span class="comment">#     logger.info('samtools end')</span></span><br><span class="line">    <span class="keyword">if</span> <span class="string">f'<span class="subst">&#123;SRR_id&#125;</span>.bam'</span> <span class="keyword">in</span> os.listdir(BAMdir):</span><br><span class="line">        logger.info(<span class="string">f'<span class="subst">&#123;SRR_id&#125;</span>已排序'</span>)</span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        <span class="comment"># 检查是否有单端或双端FASTQ文件</span></span><br><span class="line">        <span class="keyword">if</span> sum(<span class="number">1</span> <span class="keyword">for</span> filename <span class="keyword">in</span> os.listdir(FastQdir) <span class="keyword">if</span> fnmatch.fnmatch(filename, <span class="string">f"<span class="subst">&#123;SRR_id&#125;</span>*"</span>)) == <span class="number">1</span>:</span><br><span class="line">            <span class="comment"># 单端文件的情况</span></span><br><span class="line">            fastp_cmd = <span class="string">f'fastp -i <span class="subst">&#123;FastQdir&#125;</span>/<span class="subst">&#123;SRR_id&#125;</span>.fastq.gz -o <span class="subst">&#123;FastQdir&#125;</span>/<span class="subst">&#123;SRR_id&#125;</span>_trimmed.fastq.gz'</span></span><br><span class="line">            subprocess.run(fastp_cmd, shell=<span class="keyword">True</span>, check=<span class="keyword">True</span>)</span><br><span class="line">            cmd4 = <span class="string">f'hisat2 -x <span class="subst">&#123;Referencedir&#125;</span>/<span class="subst">&#123;species&#125;</span>/<span class="subst">&#123;species&#125;</span> -p 4 -U <span class="subst">&#123;FastQdir&#125;</span>/<span class="subst">&#123;SRR_id&#125;</span>_trimmed.fastq.gz | samtools sort -n -@ 5 -o <span class="subst">&#123;BAMdir&#125;</span>/<span class="subst">&#123;SRR_id&#125;</span>.bam'</span></span><br><span class="line">        <span class="keyword">else</span>:</span><br><span class="line">            <span class="comment"># 双端文件的情况</span></span><br><span class="line">            fastp_cmd = <span class="string">f'fastp -i <span class="subst">&#123;FastQdir&#125;</span>/<span class="subst">&#123;SRR_id&#125;</span>_1.fastq.gz -I <span class="subst">&#123;FastQdir&#125;</span>/<span class="subst">&#123;SRR_id&#125;</span>_2.fastq.gz -o <span class="subst">&#123;FastQdir&#125;</span>/<span class="subst">&#123;SRR_id&#125;</span>_1_trimmed.fastq.gz -O <span class="subst">&#123;FastQdir&#125;</span>/<span class="subst">&#123;SRR_id&#125;</span>_2_trimmed.fastq.gz'</span></span><br><span class="line">            subprocess.run(fastp_cmd, shell=<span class="keyword">True</span>, check=<span class="keyword">True</span>)</span><br><span class="line">            cmd4 = <span class="string">f'hisat2 -x <span class="subst">&#123;Referencedir&#125;</span>/<span class="subst">&#123;species&#125;</span>/<span class="subst">&#123;species&#125;</span> -p 4 -1 <span class="subst">&#123;FastQdir&#125;</span>/<span class="subst">&#123;SRR_id&#125;</span>_1_trimmed.fastq.gz -2 <span class="subst">&#123;FastQdir&#125;</span>/<span class="subst">&#123;SRR_id&#125;</span>_2_trimmed.fastq.gz | samtools sort -n -@ 5 -o <span class="subst">&#123;BAMdir&#125;</span>/<span class="subst">&#123;SRR_id&#125;</span>.bam'</span></span><br><span class="line">        </span><br><span class="line">        subprocess.run(cmd4, shell=<span class="keyword">True</span>, check=<span class="keyword">True</span>)</span><br><span class="line">        logger.info(<span class="string">"hisat2 and samtools pipeline completed"</span>)</span><br><span class="line"></span><br><span class="line">    </span><br><span class="line">    <span class="keyword">if</span> <span class="string">f'<span class="subst">&#123;SRR_id&#125;</span>_featureCounts.tsv'</span> <span class="keyword">in</span> os.listdir(Countdir):</span><br><span class="line">        ic(<span class="string">f'<span class="subst">&#123;SRR_id&#125;</span>已定量'</span>)</span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        <span class="keyword">if</span> sum(<span class="number">1</span> <span class="keyword">for</span> filename <span class="keyword">in</span> os.listdir(FastQdir) <span class="keyword">if</span> fnmatch.fnmatch(filename, <span class="string">f"<span class="subst">&#123;SRR_id&#125;</span>*"</span>)) == <span class="number">1</span>:</span><br><span class="line">            cmd8 = <span class="string">f'featureCounts -t exon -g Name \</span></span><br><span class="line"><span class="string">                                -Q 10 --primary -s 0 -T 24 \</span></span><br><span class="line"><span class="string">                                -a <span class="subst">&#123;Referencedir&#125;</span>/<span class="subst">&#123;species&#125;</span>.gff3 \</span></span><br><span class="line"><span class="string">                                -o <span class="subst">&#123;Countdir&#125;</span>/<span class="subst">&#123;SRR_id&#125;</span>_featureCounts.tsv \</span></span><br><span class="line"><span class="string">                                <span class="subst">&#123;BAMdir&#125;</span>/<span class="subst">&#123;SRR_id&#125;</span>.bam &gt; <span class="subst">&#123;Countlogdir&#125;</span>/<span class="subst">&#123;SRR_id&#125;</span>_featureCounts.log  2&gt;&amp;1'</span></span><br><span class="line">            subprocess.run(cmd8, shell=<span class="keyword">True</span>, check=<span class="keyword">True</span>)</span><br><span class="line">            logger.info(<span class="string">f'featureCounts finished'</span>)</span><br><span class="line">        <span class="keyword">else</span>:</span><br><span class="line">            cmd8 = <span class="string">f'featureCounts -t exon -g Name \</span></span><br><span class="line"><span class="string">                    -Q 10 --primary -s 0 -T 24 \</span></span><br><span class="line"><span class="string">                    -a <span class="subst">&#123;Referencedir&#125;</span>/<span class="subst">&#123;species&#125;</span>.gff3 \</span></span><br><span class="line"><span class="string">                    -o <span class="subst">&#123;Countdir&#125;</span>/<span class="subst">&#123;SRR_id&#125;</span>_featureCounts.tsv \</span></span><br><span class="line"><span class="string">                    -p <span class="subst">&#123;BAMdir&#125;</span>/<span class="subst">&#123;SRR_id&#125;</span>.bam &gt; <span class="subst">&#123;Countlogdir&#125;</span>/<span class="subst">&#123;SRR_id&#125;</span>_featureCounts.log  2&gt;&amp;1'</span></span><br><span class="line">            subprocess.run(cmd8, shell=<span class="keyword">True</span>, check=<span class="keyword">True</span>)</span><br><span class="line">            logger.info(<span class="string">f'featureCounts finished'</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">    cmd3 =<span class="string">f'rm <span class="subst">&#123;SRRdir&#125;</span>/<span class="subst">&#123;SRR_id&#125;</span>'</span></span><br><span class="line">    cmd5 = <span class="string">f'rm <span class="subst">&#123;FastQdir&#125;</span>/<span class="subst">&#123;SRR_id&#125;</span>*'</span></span><br><span class="line">    <span class="comment">#cmd7 = f'rm &#123;SAMdir&#125;/&#123;SRR_id&#125;.sam'</span></span><br><span class="line">    cmd9 = <span class="string">f'rm <span class="subst">&#123;BAMdir&#125;</span>/<span class="subst">&#123;SRR_id&#125;</span>.bam'</span></span><br><span class="line"></span><br><span class="line">    <span class="comment">#os.system(cmd3)</span></span><br><span class="line">    subprocess.run(cmd3, shell=<span class="keyword">True</span>, check=<span class="keyword">True</span>)</span><br><span class="line">    <span class="comment">#os.system(cmd5)</span></span><br><span class="line">    subprocess.run(cmd5, shell=<span class="keyword">True</span>, check=<span class="keyword">True</span>)</span><br><span class="line">    <span class="comment">#os.system(cmd7)</span></span><br><span class="line">    <span class="comment">#subprocess.run(cmd7, shell=True, check=True)</span></span><br><span class="line">    <span class="comment">#logger.info(f'rm sam finished')</span></span><br><span class="line">    <span class="comment">#os.system(cmd9)</span></span><br><span class="line">    subprocess.run(cmd9, shell=<span class="keyword">True</span>, check=<span class="keyword">True</span>)</span><br><span class="line">    logger.info(<span class="string">f'rm bam finished'</span>)</span><br><span class="line"></span><br><span class="line">    logger.info(<span class="string">f'process <span class="subst">&#123;SRR_id&#125;</span> finished'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">#HISAT2 比对速度确实很快，一个样本转录组数据比对 2G 的核糖体s RNA 参考基因组约 25 分钟，bowtie2 需要 170 分钟（线程数都是 4）。 bowtie2 的 local 模式比对出 21% 的 rRNA 污染链宴，而 HISAT2 比对 2% 的 rRNA 污染，差异也挺大的……</span></span><br><span class="line"><span class="comment">#但是，HISAT2 的线程数不能设置太高，用户手册中建议对人类参考基因组进行比对时，线程数设置在 1-8 之间。我用文献悉握  Transcript-level expression analysis of RNA-seq</span></span><br><span class="line"><span class="comment">#experiments with HISAT, StringTie and Ballgown  中的数据测试也发现当线程数超过 12 时整个比对步骤消耗的时间反而会增加。</span></span><br><span class="line"></span><br><span class="line"><span class="comment">##比对前的去接头和比对后的过滤去重???</span></span><br></pre></td></tr></table></figure>
<p>发现问题停止运行的脚本stop.sh,如果停止运行了，要把srr,fastq,sam,bam中间文件删除，以免下一次报错</p>
<figure class="highlight bash"><table><tr><td class="code"><pre><span class="line">pkill -f <span class="string">"python3 /home/win/16t2/study/deep_learning/gene/course_try/1Sequence/RNASeq_allprocess.py"</span></span><br><span class="line">pkill -f <span class="string">"/usr/bin/perl /usr/bin/hisat2"</span></span><br><span class="line">pkill -f <span class="string">"sh -c '/usr/bin/hisat2-align-s'"</span></span><br><span class="line">pkill -f <span class="string">"/usr/bin/hisat2-align-s --wrapper"</span></span><br><span class="line">pkill -f <span class="string">"samtools sort"</span></span><br><span class="line">pkill -f <span class="string">"fastq-dump"</span></span><br><span class="line">pkill -f <span class="string">"/usr/bin/fastq-dump"</span></span><br></pre></td></tr></table></figure>
<h2 id="四、定量文件拼接与数据标准化"><a href="#四、定量文件拼接与数据标准化" class="headerlink" title="四、定量文件拼接与数据标准化"></a>四、定量文件拼接与数据标准化</h2><p>将定量出的tsv文件进行拼接以及转为FPKM</p>
<figure class="highlight python"><table><tr><td class="code"><pre><span class="line"></span><br><span class="line"><span class="comment">#-------------------------------------拼接各个tsv脚本---------------------------------------------</span></span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> os</span><br><span class="line"><span class="keyword">from</span> icecream <span class="keyword">import</span> install,ic</span><br><span class="line">install()</span><br><span class="line"><span class="keyword">from</span> tqdm <span class="keyword">import</span> tqdm</span><br><span class="line"><span class="comment"># 指定你的文件夹路径</span></span><br><span class="line">folder_path = <span class="string">'/home/win/16t2/study/deep_learning/gene/RNA_Seq/genecount/genecount_japonica_3400'</span></span><br><span class="line">output_dir =<span class="string">'/home/win/16t2/study/deep_learning/gene/RNA_Seq/genecount/genecount_japonica_3400_df'</span></span><br><span class="line">species = <span class="string">'Oryza_japonica'</span></span><br><span class="line"><span class="comment"># 初始化用于合并所有文件的DataFrame，此时还不添加数据</span></span><br><span class="line">merged_df = pd.DataFrame()</span><br><span class="line"></span><br><span class="line"><span class="comment"># 遍历指定文件夹查找TSV文件</span></span><br><span class="line"><span class="keyword">for</span> filename <span class="keyword">in</span> tqdm(os.listdir(folder_path)):</span><br><span class="line">    <span class="keyword">if</span> filename.endswith(<span class="string">'.tsv'</span>):</span><br><span class="line">        file_path = os.path.join(folder_path, filename)</span><br><span class="line">        </span><br><span class="line">        <span class="comment"># 假设文件有标头，根据实际情况读取Geneid和Length列，以及第七列</span></span><br><span class="line">        <span class="comment"># 这里我们假设标头分别是Geneid,Length,...，根据需要调整列名或列索引</span></span><br><span class="line">        df = pd.read_csv(file_path, sep=<span class="string">'\t'</span>, skiprows=<span class="number">1</span>, header=<span class="keyword">None</span>)</span><br><span class="line">        df.columns = [<span class="string">'Geneid'</span>, <span class="string">'Unknown1'</span>, <span class="string">'Start'</span>, <span class="string">'End'</span>, <span class="string">'Strand'</span>, <span class="string">'Length'</span>, <span class="string">'Count'</span>]</span><br><span class="line">        <span class="comment">#df = pd.read_csv(file_path, sep='\t', usecols=['Geneid', 'Length', 6])</span></span><br><span class="line">        <span class="comment">#ic(df)</span></span><br><span class="line">        <span class="comment"># 文件名命名新列，假设文件名为‘sample_1.tsv’，分割后取‘sample’</span></span><br><span class="line">        new_column_name = filename.split(<span class="string">'_'</span>)[<span class="number">0</span>]</span><br><span class="line">        </span><br><span class="line">        <span class="comment"># 如果是第一个文件，初始化merged_df</span></span><br><span class="line">        <span class="keyword">if</span> merged_df.empty:</span><br><span class="line">            merged_df[<span class="string">'Geneid'</span>] = df[<span class="string">'Geneid'</span>]</span><br><span class="line">            merged_df[<span class="string">'Length'</span>] = df[<span class="string">'Length'</span>]</span><br><span class="line">            merged_df[new_column_name] = df.iloc[:, <span class="number">6</span>]  <span class="comment"># 添加第七列（此时是df的第3列）</span></span><br><span class="line">        <span class="keyword">else</span>:</span><br><span class="line">            <span class="comment"># 将后续文件的第七列添加到merged_df中，匹配Geneid</span></span><br><span class="line">            <span class="comment"># 假设Geneid完全匹配，直接添加新列</span></span><br><span class="line">            merged_df[new_column_name] = df.iloc[:, <span class="number">6</span>]</span><br><span class="line"></span><br><span class="line"><span class="comment"># 检查结果</span></span><br><span class="line">print(merged_df)</span><br><span class="line">merged_df  = merged_df.iloc[<span class="number">1</span>:,:]</span><br><span class="line">output_path = <span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/genecount_<span class="subst">&#123;species&#125;</span>.pkl'</span></span><br><span class="line"><span class="comment"># 使用to_pickle方法保存DataFrames</span></span><br><span class="line">merged_df.to_pickle(output_path)</span><br><span class="line"></span><br><span class="line"><span class="comment">#-------------------------------------生成表达矩阵和基因长度文件------------------------------------</span></span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"></span><br><span class="line"><span class="comment"># 读取CSV文件</span></span><br><span class="line">df = pd.read_pickle(output_path) <span class="comment">#pd.read_csv('/mnt/h/study/deep_learning/gene/RNA_Seq/genecount/drought_samplecount.csv')</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 创建一个新的DataFrame来保存gene_id和gene_length</span></span><br><span class="line">lengths_df = df.iloc[:, :<span class="number">2</span>]  <span class="comment"># 选择前两列</span></span><br><span class="line">lengths_df.columns = [<span class="string">'FEATURE_ID'</span>, <span class="string">'GENE_LENGTHS'</span>]  <span class="comment"># 更改列名</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 保存第一列和第二列为新的CSV文件</span></span><br><span class="line">lengths_df.to_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/lengths.pkl'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 为了转置剩余的列（除了第二列外），先删除第二列</span></span><br><span class="line">exp_df = df.drop(df.columns[<span class="number">1</span>], axis=<span class="number">1</span>)  <span class="comment"># 删除第二列</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 转置剩余列</span></span><br><span class="line"><span class="comment">#exp_df_transposed = exp_df.T  # 只转置第二列之后的数据</span></span><br><span class="line">exp_df.rename(columns=&#123;exp_df.columns[<span class="number">0</span>]: <span class="string">'FEATURE_ID'</span>&#125;, inplace=<span class="keyword">True</span>)</span><br><span class="line"><span class="comment"># 保存转置后的DataFrame到CSV</span></span><br><span class="line">print(exp_df)</span><br><span class="line">feature_id = exp_df.iloc[:,<span class="number">0</span>]</span><br><span class="line">exp_df = exp_df.iloc[:,<span class="number">1</span>:]</span><br><span class="line"></span><br><span class="line">exp_df.to_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/exp.pkl'</span>)</span><br><span class="line"><span class="comment">#ic(exp_df)</span></span><br><span class="line"><span class="comment">#-------------------------------------标准化------------------------------------</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">from</span> icecream <span class="keyword">import</span> install,ic</span><br><span class="line">install()</span><br><span class="line"><span class="comment"># Load the expression counts data (assuming it's in CSV format)</span></span><br><span class="line"><span class="comment"># The DataFrame should have genes as rows and samples as columns</span></span><br><span class="line">exp_counts_df = pd.read_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/exp.pkl'</span>)</span><br><span class="line"><span class="comment">#exp_counts_df = exp_counts_df.iloc[:,1:]</span></span><br><span class="line"><span class="comment"># Load the gene lengths data</span></span><br><span class="line">lengths_df = pd.read_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/lengths.pkl'</span>)</span><br><span class="line">lengths_df = lengths_df.iloc[:,<span class="number">1</span>:]</span><br><span class="line"><span class="comment">#ic(lengths_df)</span></span><br><span class="line"><span class="comment">#ic(exp_counts_df.shape,lengths_df.shape,lengths_df.values.squeeze().shape)</span></span><br><span class="line"><span class="comment"># --- 计算 CPM ---</span></span><br><span class="line"><span class="comment"># Sum the counts for each sample (column) to get the total counts</span></span><br><span class="line">total_counts_per_sample = exp_counts_df.sum(axis=<span class="number">0</span>)</span><br><span class="line"><span class="comment">#ic(total_counts_per_sample)</span></span><br><span class="line"><span class="comment"># Divide each count by the total counts and multiply by one million</span></span><br><span class="line">cpm_df = exp_counts_df.divide(total_counts_per_sample, axis=<span class="number">1</span>) * <span class="number">10</span>**<span class="number">6</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># --- 计算 FPKM ---</span></span><br><span class="line"><span class="comment"># Divide the counts by the lengths of the genes to get counts per kilobase</span></span><br><span class="line"><span class="comment"># lengths_df.values.squeeze() converts the DataFrame to a single column array</span></span><br><span class="line"><span class="comment"># The lengths are assumed to be in base pairs, so we divide by 1,000 to convert to kilobases</span></span><br><span class="line">ic(exp_counts_df)</span><br><span class="line">ic(lengths_df)</span><br><span class="line"><span class="comment"># 显示数据类型</span></span><br><span class="line">ic(lengths_df.dtypes)</span><br><span class="line"><span class="comment"># 输出所有可能非数值的行</span></span><br><span class="line">ic(lengths_df[pd.to_numeric(lengths_df.squeeze(), errors=<span class="string">'coerce'</span>).isna()])</span><br><span class="line">lengths_df = lengths_df.apply(pd.to_numeric, errors=<span class="string">'coerce'</span>)</span><br><span class="line"><span class="comment"># 再次检查数据</span></span><br><span class="line">ic(lengths_df.dtypes)</span><br><span class="line">exp_counts_df = exp_counts_df.apply(pd.to_numeric, errors=<span class="string">'coerce'</span>)</span><br><span class="line">counts_per_kb = exp_counts_df.divide(lengths_df[<span class="string">'GENE_LENGTHS'</span>] / <span class="number">10</span>**<span class="number">3</span>, axis=<span class="number">0</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Then divide by the total counts (in millions) to get FPKM</span></span><br><span class="line">fpkm_df = counts_per_kb.divide(total_counts_per_sample / <span class="number">10</span>**<span class="number">6</span>, axis=<span class="number">1</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># --- 计算 TPM ---</span></span><br><span class="line"><span class="comment"># First, get the per million scaling factor</span></span><br><span class="line"><span class="comment"># Divide counts per kb (from the FPKM calculation step) by the total counts per sample</span></span><br><span class="line">per_million_scaling_factor = counts_per_kb.divide(total_counts_per_sample, axis=<span class="number">1</span>).sum(axis=<span class="number">0</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Divide counts per kb by the per million scaling factor, and then normalize by multiplying by one million</span></span><br><span class="line">tpm_df = counts_per_kb.divide(per_million_scaling_factor, axis=<span class="number">1</span>) * <span class="number">10</span>**<span class="number">6</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Save the CPM, FPKM, and TPM data to new CSV files</span></span><br><span class="line">cpm_df.to_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/exp_cpm_transcript_<span class="subst">&#123;species&#125;</span>.pkl'</span>)</span><br><span class="line">fpkm_df.to_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/exp_fpkm_transcript_<span class="subst">&#123;species&#125;</span>.pkl'</span>)</span><br><span class="line">tpm_df.to_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/exp_tpm_transcript_<span class="subst">&#123;species&#125;</span>.pkl'</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment">#----------------------------------------转录本和基因名转换-----------------------------------------------------</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"></span><br><span class="line"><span class="comment"># 加载数据</span></span><br><span class="line">df = pd.read_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/exp_fpkm_transcript_<span class="subst">&#123;species&#125;</span>.pkl'</span>)</span><br><span class="line">ic(feature_id)</span><br><span class="line">df.insert(<span class="number">0</span>, <span class="string">'FEATURE_ID'</span>, feature_id)</span><br><span class="line"><span class="comment"># 选择 FEATURE_ID 以 "Os" 开头的行</span></span><br><span class="line">df = df[df[<span class="string">'FEATURE_ID'</span>].str.startswith(<span class="string">'Os'</span>)]</span><br><span class="line"></span><br><span class="line"><span class="comment"># 删掉 FEATURE_ID 中第一个 "-" 字符后面的所有字符</span></span><br><span class="line">df[<span class="string">'FEATURE_ID'</span>] = df[<span class="string">'FEATURE_ID'</span>].apply(<span class="keyword">lambda</span> x: x.split(<span class="string">'-'</span>)[<span class="number">0</span>])</span><br><span class="line"></span><br><span class="line"><span class="comment"># 将 FEATURE_ID 中的 "t" 字符替换为 "g"</span></span><br><span class="line">df[<span class="string">'FEATURE_ID'</span>] = df[<span class="string">'FEATURE_ID'</span>].str.replace(<span class="string">'t'</span>, <span class="string">'g'</span>, n=<span class="number">1</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 将 FEATURE_ID 相同的所有行相加合并为一行</span></span><br><span class="line">df_grouped = df.groupby(<span class="string">'FEATURE_ID'</span>).sum().reset_index()</span><br><span class="line"></span><br><span class="line"><span class="comment"># 保存处理后的数据</span></span><br><span class="line">df_grouped.to_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/exp_fpkm_gene_<span class="subst">&#123;species&#125;</span>.pkl'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">#------------------------------------------------批次效应校正-------------------------------------------------------</span></span><br><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">from</span> icecream <span class="keyword">import</span> install,ic</span><br><span class="line">install()</span><br><span class="line"><span class="comment"># 1. 读取 CSV 文件</span></span><br><span class="line">input_filepath = <span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/exp_fpkm_gene_<span class="subst">&#123;species&#125;</span>.pkl'</span></span><br><span class="line">df = pd.read_pickle(input_filepath)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 2. 转置 DataFrame</span></span><br><span class="line">df_transposed = df.transpose()</span><br><span class="line">ic(df_transposed.shape)</span><br><span class="line">ic(df_transposed.iloc[<span class="number">0</span>,:])</span><br><span class="line"></span><br><span class="line">df_transposed.columns = df_transposed.iloc[<span class="number">0</span>,:]</span><br><span class="line"><span class="comment"># 步骤2: 删除原来的第0行</span></span><br><span class="line">df_transposed = df_transposed.drop(index=<span class="string">'FEATURE_ID'</span>)</span><br><span class="line"><span class="comment">#-------------------------------------拼接各个tsv脚本---------------------------------------------</span></span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> os</span><br><span class="line"><span class="keyword">from</span> icecream <span class="keyword">import</span> install,ic</span><br><span class="line">install()</span><br><span class="line"><span class="keyword">from</span> tqdm <span class="keyword">import</span> tqdm</span><br><span class="line"><span class="comment"># 指定你的文件夹路径</span></span><br><span class="line">folder_path = <span class="string">'/home/win/16t2/study/deep_learning/gene/RNA_Seq/genecount/Zeamays_Count'</span></span><br><span class="line">output_dir =<span class="string">'/home/win/16t2/study/deep_learning/gene/RNA_Seq/genecount/out_table'</span></span><br><span class="line">species = <span class="string">'Zeamays'</span></span><br><span class="line">japonica_csv1134_path=<span class="string">'/home/win/16t2/study/deep_learning/gene/RNA_Seq/genecount/genecount_japonica_1134/exp.csv'</span></span><br><span class="line">japonica_csv1134 = pd.read_csv(japonica_csv1134_path)</span><br><span class="line">japonica_csv1134.columns =[str(i).split(<span class="string">'/'</span>)[<span class="number">-1</span>].split(<span class="string">'.'</span>)[<span class="number">0</span>] <span class="keyword">for</span> i <span class="keyword">in</span> japonica_csv1134.columns]</span><br><span class="line"><span class="comment">#ic(japonica_csv1134.shape)</span></span><br><span class="line"><span class="comment"># 初始化用于合并所有文件的DataFrame，此时还不添加数据</span></span><br><span class="line">merged_df = pd.DataFrame()</span><br><span class="line"></span><br><span class="line"><span class="comment"># 遍历指定文件夹查找TSV文件</span></span><br><span class="line"><span class="keyword">for</span> filename <span class="keyword">in</span> tqdm(os.listdir(folder_path)):</span><br><span class="line">    <span class="keyword">if</span> filename.endswith(<span class="string">'.tsv'</span>):</span><br><span class="line">        file_path = os.path.join(folder_path, filename)</span><br><span class="line">        </span><br><span class="line">        <span class="comment"># 假设文件有标头，根据实际情况读取Geneid和Length列，以及第七列</span></span><br><span class="line">        <span class="comment"># 这里我们假设标头分别是Geneid,Length,...，根据需要调整列名或列索引</span></span><br><span class="line">        df = pd.read_csv(file_path, sep=<span class="string">'\t'</span>, skiprows=<span class="number">1</span>, header=<span class="keyword">None</span>)</span><br><span class="line">        df.columns = [<span class="string">'Geneid'</span>, <span class="string">'Unknown1'</span>, <span class="string">'Start'</span>, <span class="string">'End'</span>, <span class="string">'Strand'</span>, <span class="string">'Length'</span>, <span class="string">'Count'</span>]</span><br><span class="line">        <span class="comment">#df = pd.read_csv(file_path, sep='\t', usecols=['Geneid', 'Length', 6])</span></span><br><span class="line">        <span class="comment">#ic(df)</span></span><br><span class="line">        <span class="comment"># 文件名命名新列，假设文件名为‘sample_1.tsv’，分割后取‘sample’</span></span><br><span class="line">        new_column_name = filename.split(<span class="string">'_'</span>)[<span class="number">0</span>]</span><br><span class="line">        </span><br><span class="line">        <span class="comment"># 如果是第一个文件，初始化merged_df</span></span><br><span class="line">        <span class="keyword">if</span> merged_df.empty:</span><br><span class="line">            merged_df[<span class="string">'Geneid'</span>] = df[<span class="string">'Geneid'</span>]</span><br><span class="line">            merged_df[<span class="string">'Length'</span>] = df[<span class="string">'Length'</span>]</span><br><span class="line">            merged_df[new_column_name] = df.iloc[:, <span class="number">6</span>]  <span class="comment"># 添加第七列（此时是df的第3列）</span></span><br><span class="line">        <span class="keyword">else</span>:</span><br><span class="line">            <span class="comment"># 将后续文件的第七列添加到merged_df中，匹配Geneid</span></span><br><span class="line">            <span class="comment"># 假设Geneid完全匹配，直接添加新列</span></span><br><span class="line">            merged_df[new_column_name] = df.iloc[:, <span class="number">6</span>]</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> species == <span class="string">'Oryza_japonica'</span>:</span><br><span class="line">    merged_df = pd.concat([merged_df,japonica_csv1134.iloc[:,<span class="number">6</span>:]], axis=<span class="number">1</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># 检查结果</span></span><br><span class="line"><span class="comment">#ic(merged_df.shape)</span></span><br><span class="line">merged_df  = merged_df.iloc[<span class="number">1</span>:,:]</span><br><span class="line">output_path = <span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/genecount_<span class="subst">&#123;species&#125;</span>.pkl'</span></span><br><span class="line"><span class="comment"># 使用to_pickle方法保存DataFrames</span></span><br><span class="line">merged_df.to_pickle(output_path)</span><br><span class="line"></span><br><span class="line"><span class="comment"># #-------------------------------------生成表达矩阵和基因长度文件------------------------------------</span></span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"></span><br><span class="line"><span class="comment"># # 读取CSV文件</span></span><br><span class="line">df = pd.read_pickle(output_path) <span class="comment">#pd.read_csv('/mnt/h/study/deep_learning/gene/RNA_Seq/genecount/drought_samplecount.csv')</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 创建一个新的DataFrame来保存gene_id和gene_length</span></span><br><span class="line">lengths_df = df.iloc[:, :<span class="number">2</span>]  <span class="comment"># 选择前两列</span></span><br><span class="line">lengths_df.columns = [<span class="string">'FEATURE_ID'</span>, <span class="string">'GENE_LENGTHS'</span>]  <span class="comment"># 更改列名</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 保存第一列和第二列为新的CSV文件</span></span><br><span class="line">lengths_df.to_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/lengths.pkl'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 为了转置剩余的列（除了第二列外），先删除第二列</span></span><br><span class="line">exp_df = df.drop(df.columns[<span class="number">1</span>], axis=<span class="number">1</span>)  <span class="comment"># 删除第二列</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 转置剩余列</span></span><br><span class="line"><span class="comment">#exp_df_transposed = exp_df.T  # 只转置第二列之后的数据</span></span><br><span class="line">exp_df.rename(columns=&#123;exp_df.columns[<span class="number">0</span>]: <span class="string">'FEATURE_ID'</span>&#125;, inplace=<span class="keyword">True</span>)</span><br><span class="line"><span class="comment"># 保存转置后的DataFrame到CSV</span></span><br><span class="line"><span class="comment">#print(exp_df)</span></span><br><span class="line">feature_id = exp_df.iloc[:,<span class="number">0</span>]</span><br><span class="line">exp_df = exp_df.iloc[:,<span class="number">1</span>:]</span><br><span class="line"></span><br><span class="line">exp_df.to_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/exp.pkl'</span>)</span><br><span class="line"><span class="comment">#ic(exp_df)</span></span><br><span class="line"><span class="comment">#-------------------------------------标准化------------------------------------</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">from</span> icecream <span class="keyword">import</span> install,ic</span><br><span class="line">install()</span><br><span class="line"><span class="comment"># Load the expression counts data (assuming it's in CSV format)</span></span><br><span class="line"><span class="comment"># The DataFrame should have genes as rows and samples as columns</span></span><br><span class="line">exp_counts_df = pd.read_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/exp.pkl'</span>)</span><br><span class="line"><span class="comment">#feature_id = exp_counts_df.iloc[:,0]</span></span><br><span class="line">exp_counts_df = exp_counts_df.fillna(<span class="number">0</span>)</span><br><span class="line">exp_counts_df = exp_counts_df.astype(int)</span><br><span class="line"></span><br><span class="line"><span class="comment">#exp_counts_df = exp_counts_df.iloc[:,1:]</span></span><br><span class="line"><span class="comment"># Load the gene lengths data</span></span><br><span class="line"><span class="comment">#ic(exp_counts_df)</span></span><br><span class="line">lengths_df = pd.read_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/lengths.pkl'</span>)</span><br><span class="line">lengths_df = lengths_df.iloc[:,<span class="number">1</span>:]</span><br><span class="line"><span class="comment">#ic(lengths_df)</span></span><br><span class="line"><span class="comment">#ic(exp_counts_df.shape,lengths_df.shape,lengths_df.values.squeeze().shape)</span></span><br><span class="line"><span class="comment"># --- 计算 CPM ---</span></span><br><span class="line"><span class="comment"># Sum the counts for each sample (column) to get the total counts</span></span><br><span class="line">total_counts_per_sample = exp_counts_df.sum(axis=<span class="number">0</span>)</span><br><span class="line"><span class="comment">#ic(total_counts_per_sample)</span></span><br><span class="line"><span class="comment"># Divide each count by the total counts and multiply by one million</span></span><br><span class="line">cpm_df = exp_counts_df.divide(total_counts_per_sample, axis=<span class="number">1</span>) * <span class="number">10</span>**<span class="number">6</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># --- 计算 FPKM ---</span></span><br><span class="line"><span class="comment"># Divide the counts by the lengths of the genes to get counts per kilobase</span></span><br><span class="line"><span class="comment"># lengths_df.values.squeeze() converts the DataFrame to a single column array</span></span><br><span class="line"><span class="comment"># The lengths are assumed to be in base pairs, so we divide by 1,000 to convert to kilobases</span></span><br><span class="line"><span class="comment">#ic(exp_counts_df)</span></span><br><span class="line"><span class="comment">#ic(lengths_df)</span></span><br><span class="line"><span class="comment"># 显示数据类型</span></span><br><span class="line"><span class="comment">#ic(lengths_df.dtypes)</span></span><br><span class="line"><span class="comment"># 输出所有可能非数值的行</span></span><br><span class="line"><span class="comment">#ic(lengths_df[pd.to_numeric(lengths_df.squeeze(), errors='coerce').isna()])</span></span><br><span class="line">lengths_df = lengths_df.apply(pd.to_numeric, errors=<span class="string">'coerce'</span>)</span><br><span class="line"><span class="comment"># 再次检查数据</span></span><br><span class="line"><span class="comment">#ic(lengths_df.dtypes)</span></span><br><span class="line">exp_counts_df = exp_counts_df.apply(pd.to_numeric, errors=<span class="string">'coerce'</span>)</span><br><span class="line">counts_per_kb = exp_counts_df.divide(lengths_df[<span class="string">'GENE_LENGTHS'</span>] / <span class="number">10</span>**<span class="number">3</span>, axis=<span class="number">0</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Then divide by the total counts (in millions) to get FPKM</span></span><br><span class="line">fpkm_df = counts_per_kb.divide(total_counts_per_sample / <span class="number">10</span>**<span class="number">6</span>, axis=<span class="number">1</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># --- 计算 TPM ---</span></span><br><span class="line"><span class="comment"># First, get the per million scaling factor</span></span><br><span class="line"><span class="comment"># Divide counts per kb (from the FPKM calculation step) by the total counts per sample</span></span><br><span class="line">per_million_scaling_factor = counts_per_kb.divide(total_counts_per_sample, axis=<span class="number">1</span>).sum(axis=<span class="number">0</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Divide counts per kb by the per million scaling factor, and then normalize by multiplying by one million</span></span><br><span class="line">tpm_df = counts_per_kb.divide(per_million_scaling_factor, axis=<span class="number">1</span>) * <span class="number">10</span>**<span class="number">6</span></span><br><span class="line"></span><br><span class="line">ic(cpm_df.shape)</span><br><span class="line">ic(fpkm_df.shape)</span><br><span class="line">ic(tpm_df.shape)</span><br><span class="line"><span class="comment"># Save the CPM, FPKM, and TPM data to new CSV files</span></span><br><span class="line">cpm_df.to_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/exp_cpm_transcript_<span class="subst">&#123;species&#125;</span>.pkl'</span>)</span><br><span class="line">fpkm_df.to_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/exp_fpkm_transcript_<span class="subst">&#123;species&#125;</span>.pkl'</span>)</span><br><span class="line">tpm_df.to_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/exp_tpm_transcript_<span class="subst">&#123;species&#125;</span>.pkl'</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment">#----------------------------------------转录本和基因名转换-----------------------------------------------------</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> species==<span class="string">'Oryza_japonica'</span>:</span><br><span class="line">    <span class="comment"># 加载数据</span></span><br><span class="line">    df = pd.read_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/exp_fpkm_transcript_<span class="subst">&#123;species&#125;</span>.pkl'</span>)</span><br><span class="line">    <span class="comment">#ic(feature_id)</span></span><br><span class="line">    df.insert(<span class="number">0</span>, <span class="string">'FEATURE_ID'</span>, feature_id)</span><br><span class="line">    ic(df[<span class="string">'FEATURE_ID'</span>])</span><br><span class="line">    <span class="comment"># 选择 FEATURE_ID 以 "Os" 开头的行</span></span><br><span class="line">    df = df[df[<span class="string">'FEATURE_ID'</span>].str.startswith(<span class="string">'Os'</span>)]</span><br><span class="line">    <span class="comment"># 删掉 FEATURE_ID 中第一个 "-" 字符后面的所有字符</span></span><br><span class="line">    df[<span class="string">'FEATURE_ID'</span>] = df[<span class="string">'FEATURE_ID'</span>].apply(<span class="keyword">lambda</span> x: x.split(<span class="string">'-'</span>)[<span class="number">0</span>])</span><br><span class="line">    <span class="comment"># 将 FEATURE_ID 中的 "t" 字符替换为 "g"</span></span><br><span class="line">    df[<span class="string">'FEATURE_ID'</span>] = df[<span class="string">'FEATURE_ID'</span>].str.replace(<span class="string">'t'</span>, <span class="string">'g'</span>, n=<span class="number">1</span>)</span><br><span class="line">    ic(df[<span class="string">'FEATURE_ID'</span>])</span><br><span class="line">    <span class="comment"># 将 FEATURE_ID 相同的所有行相加合并为一行</span></span><br><span class="line">    df_grouped = df.groupby(<span class="string">'FEATURE_ID'</span>).sum().reset_index()</span><br><span class="line">    ic(df_grouped.shape)</span><br><span class="line">    <span class="comment"># 保存处理后的数据</span></span><br><span class="line">    df_grouped.to_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/exp_fpkm_gene_<span class="subst">&#123;species&#125;</span>.pkl'</span>)</span><br><span class="line"><span class="keyword">elif</span> species==<span class="string">'Oryza_indica'</span>:</span><br><span class="line">    <span class="comment"># 加载数据</span></span><br><span class="line">    df = pd.read_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/exp_fpkm_transcript_<span class="subst">&#123;species&#125;</span>.pkl'</span>)</span><br><span class="line">    <span class="comment">#ic(feature_id)</span></span><br><span class="line">    df.insert(<span class="number">0</span>, <span class="string">'FEATURE_ID'</span>, feature_id)</span><br><span class="line">    <span class="comment"># 选择 FEATURE_ID 以 "Os" 开头的行</span></span><br><span class="line">    df = df[df[<span class="string">'FEATURE_ID'</span>].str.startswith(<span class="string">'BGIOSGA'</span>)]</span><br><span class="line">    <span class="comment"># 删掉 FEATURE_ID 中第一个 "-" 字符后面的所有字符</span></span><br><span class="line">    df[<span class="string">'FEATURE_ID'</span>] = df[<span class="string">'FEATURE_ID'</span>].apply(<span class="keyword">lambda</span> x: x.split(<span class="string">'-'</span>)[<span class="number">0</span>])</span><br><span class="line">    <span class="comment">#ic(df['FEATURE_ID'])</span></span><br><span class="line">    <span class="comment"># 将 FEATURE_ID 中的 "t" 字符替换为 "g"</span></span><br><span class="line">    <span class="comment">#df['FEATURE_ID'] = df['FEATURE_ID'].str.replace('t', 'g', n=1)</span></span><br><span class="line">    <span class="comment"># 将 FEATURE_ID 相同的所有行相加合并为一行</span></span><br><span class="line">    df_grouped = df.groupby(<span class="string">'FEATURE_ID'</span>).sum().reset_index()</span><br><span class="line">    ic(df_grouped.shape)</span><br><span class="line">    <span class="comment"># 保存处理后的数据</span></span><br><span class="line">    df_grouped.to_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/exp_fpkm_gene_<span class="subst">&#123;species&#125;</span>.pkl'</span>)</span><br><span class="line"><span class="keyword">elif</span> species==<span class="string">'Zeamays'</span>:</span><br><span class="line">    <span class="comment"># 加载数据</span></span><br><span class="line">    df = pd.read_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/exp_fpkm_transcript_<span class="subst">&#123;species&#125;</span>.pkl'</span>)</span><br><span class="line">    <span class="comment">#ic(feature_id)</span></span><br><span class="line">    df.insert(<span class="number">0</span>, <span class="string">'FEATURE_ID'</span>, feature_id)</span><br><span class="line">    <span class="comment"># 选择 FEATURE_ID 以 "Os" 开头的行</span></span><br><span class="line">    df = df[df[<span class="string">'FEATURE_ID'</span>].str.startswith(<span class="string">'Zm'</span>)]</span><br><span class="line">    <span class="comment"># 删掉 FEATURE_ID 中第一个 "-" 字符后面的所有字符</span></span><br><span class="line">    df[<span class="string">'FEATURE_ID'</span>] = df[<span class="string">'FEATURE_ID'</span>].apply(<span class="keyword">lambda</span> x: x.split(<span class="string">'_'</span>)[<span class="number">0</span>])</span><br><span class="line">    <span class="comment">#ic(df['FEATURE_ID'])</span></span><br><span class="line">    <span class="comment"># 将 FEATURE_ID 中的 "t" 字符替换为 "g"</span></span><br><span class="line">    <span class="comment">#df['FEATURE_ID'] = df['FEATURE_ID'].str.replace('t', 'g', n=1)</span></span><br><span class="line">    <span class="comment"># 将 FEATURE_ID 相同的所有行相加合并为一行</span></span><br><span class="line">    df_grouped = df.groupby(<span class="string">'FEATURE_ID'</span>).sum().reset_index()</span><br><span class="line">    ic(df_grouped.shape)</span><br><span class="line">    <span class="comment"># 保存处理后的数据</span></span><br><span class="line">    df_grouped.to_pickle(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/exp_fpkm_gene_<span class="subst">&#123;species&#125;</span>.pkl'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">#------------------------------------------------批次效应校正-------------------------------------------------------</span></span><br><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">from</span> icecream <span class="keyword">import</span> install,ic</span><br><span class="line">install()</span><br><span class="line"><span class="comment"># 1. 读取 CSV 文件</span></span><br><span class="line">input_filepath = <span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/exp_fpkm_gene_<span class="subst">&#123;species&#125;</span>.pkl'</span></span><br><span class="line">df = pd.read_pickle(input_filepath)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 2. 转置 DataFrame</span></span><br><span class="line">df_transposed = df.transpose()</span><br><span class="line"><span class="comment">#ic(df_transposed.shape)</span></span><br><span class="line"><span class="comment">#ic(df_transposed.iloc[0,:])</span></span><br><span class="line"></span><br><span class="line">df_transposed.columns = df_transposed.iloc[<span class="number">0</span>,:]</span><br><span class="line"><span class="comment"># 步骤2: 删除原来的第0行</span></span><br><span class="line">df_transposed = df_transposed.drop(index=<span class="string">'FEATURE_ID'</span>)</span><br><span class="line"><span class="comment">#ic(df_transposed.shape)</span></span><br><span class="line"><span class="comment"># 3. 生成批次信息列</span></span><br><span class="line"><span class="comment"># 提取样本名中的 'SRR' 后的前三个数字</span></span><br><span class="line"><span class="comment">#ic(df_transposed)</span></span><br><span class="line"><span class="comment"># batch_info = df_transposed.index</span></span><br><span class="line"><span class="comment"># batch_info = pd.DataFrame([str(i)[:6] for i in batch_info])</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">batch_info = df_transposed.index.to_series().str[:<span class="number">6</span>]</span><br><span class="line"></span><br><span class="line"><span class="comment">#ic(batch_info)</span></span><br><span class="line"><span class="comment"># 创建一个唯一的批次号对应每个不同的标识</span></span><br><span class="line">batch_info = batch_info.astype(<span class="string">'category'</span>).cat.codes + <span class="number">1</span>  <span class="comment"># 用整数编码类别并从 1 开始编码</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 为转置的 DataFrame 添加批次信息列</span></span><br><span class="line">df_transposed[<span class="string">'batch'</span>] = batch_info.values  <span class="comment"># 注：不改变原始 index</span></span><br><span class="line"><span class="comment">#ic(df_transposed)</span></span><br><span class="line"><span class="comment"># 4. 创建 AnnData 对象</span></span><br><span class="line"><span class="comment"># adata = sc.AnnData(df_transposed)</span></span><br><span class="line"><span class="comment"># adata.obs['batch'] = batch_info.values  # 添加批次到.obs属性</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># # 保存转置前的列名（样本名）以备后用</span></span><br><span class="line"><span class="comment"># original_column_names = df.columns</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># # 5. 使用 scanpy 进行批次效应校正</span></span><br><span class="line"><span class="comment"># sc.pp.combat(adata, key='batch')</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># # 6. 校正后的数据转置并保存为 CSV 文件</span></span><br><span class="line"><span class="comment"># corrected_df = adata.X.transpose()[:,:-1] #  最后含有batch维度，所以多了一维</span></span><br><span class="line"><span class="comment"># ic(corrected_df)</span></span><br><span class="line"><span class="comment"># ic(df)</span></span><br><span class="line"><span class="comment"># # 创建输出 DataFrame，确保列名（样本名）和行名（基因名）与原始数据相符</span></span><br><span class="line"><span class="comment"># ic(corrected_df.shape,df.index.shape,original_column_names.shape)</span></span><br><span class="line"><span class="comment"># corrected_df = pd.DataFrame(corrected_df, index=df.index, columns=original_column_names)</span></span><br><span class="line"><span class="comment"># output_filepath = f'&#123;output_dir&#125;/batch_corrected_exp_fpkm_gene_&#123;species&#125;.pkl'</span></span><br><span class="line"><span class="comment"># corrected_df.to_pickle(output_filepath)</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment">#-------------------------可视化画图，看批次效应是否明显------------------------</span></span><br><span class="line"><span class="keyword">import</span> matplotlib.pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">from</span> sklearn.decomposition <span class="keyword">import</span> PCA</span><br><span class="line"><span class="keyword">from</span> sklearn.preprocessing <span class="keyword">import</span> StandardScaler</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"></span><br><span class="line"><span class="comment"># 假设 df_transposed 和 batch_info 已经定义好了</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 标准化数据（重要步骤，由于 PCA 对数据的尺度敏感）</span></span><br><span class="line">scaler = StandardScaler()</span><br><span class="line">df_scaled = scaler.fit_transform(df_transposed.drop(<span class="string">'batch'</span>, axis=<span class="number">1</span>))</span><br><span class="line"></span><br><span class="line"><span class="comment"># 初始化 PCA 类</span></span><br><span class="line">pca = PCA(n_components=<span class="number">2</span>)  <span class="comment"># 选择仅计算前两个主成分</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 挨个对每个批次的数据进行 PCA 分析</span></span><br><span class="line">batch_unique = batch_info.unique()</span><br><span class="line">fig, ax = plt.subplots(figsize=(<span class="number">8</span>, <span class="number">6</span>))</span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> batch <span class="keyword">in</span> batch_unique:</span><br><span class="line">    <span class="comment"># 选择当前批次的数据</span></span><br><span class="line">    batch_indices = df_transposed[<span class="string">'batch'</span>] == batch</span><br><span class="line">    batch_data = df_scaled[batch_indices, :]</span><br><span class="line">    </span><br><span class="line">    <span class="keyword">if</span> batch_data.shape[<span class="number">0</span>] &lt; <span class="number">2</span>:</span><br><span class="line">        print(<span class="string">f"Batch <span class="subst">&#123;batch&#125;</span> has less than 2 samples, skipping..."</span>)</span><br><span class="line">        <span class="keyword">continue</span></span><br><span class="line"></span><br><span class="line">    <span class="comment"># 对当前批次的数据进行 PCA</span></span><br><span class="line">    pca_result = pca.fit_transform(batch_data)</span><br><span class="line">    </span><br><span class="line">    <span class="comment"># 绘制 PCA 结果</span></span><br><span class="line">    ax.scatter(pca_result[:, <span class="number">0</span>], pca_result[:, <span class="number">1</span>], label=<span class="string">f'Batch <span class="subst">&#123;batch&#125;</span>'</span>, alpha=<span class="number">0.5</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 添加图形元素</span></span><br><span class="line">ax.set_xlabel(<span class="string">'PCA 1'</span>)</span><br><span class="line">ax.set_ylabel(<span class="string">'PCA 2'</span>)</span><br><span class="line">ax.set_title(<span class="string">'PCA of Gene Expression Data by Batch'</span>)</span><br><span class="line">ax.legend()</span><br><span class="line">plt.tight_layout()</span><br><span class="line"></span><br><span class="line"><span class="comment"># 保存图像</span></span><br><span class="line">plt.savefig(<span class="string">f'<span class="subst">&#123;output_dir&#125;</span>/<span class="subst">&#123;species&#125;</span>_pca_by_batch.png'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">#从图中来看，批次效应不明显，故不用批次效应校正</span></span><br></pre></td></tr></table></figure>
<h2 id="五、将数据从一个服务器传到另一个服务器"><a href="#五、将数据从一个服务器传到另一个服务器" class="headerlink" title="五、将数据从一个服务器传到另一个服务器"></a>五、将数据从一个服务器传到另一个服务器</h2><figure class="highlight bash"><table><tr><td class="code"><pre><span class="line"><span class="comment">#进 在本地服务器运行  前面是本地服务器地址，后面是远端服务器    出---相对于本地</span></span><br><span class="line">rsync -avz -e <span class="string">"ssh -p 21822"</span> <span class="string">"/home/win/4T/GeneDL/RNASeq/Count"</span> win10@pcrent365.com:/home/win10/3T/RNASeq/data/</span><br><span class="line">rsync -avz -e <span class="string">"ssh -p 2222"</span> /home/user/data/ user@192.168.1.2:/home/user/backup/</span><br><span class="line"><span class="comment">#出 在本地服务器运行  前面是远端服务器地址，后面是本地服务器    进---相对于本地</span></span><br><span class="line">rsync -avz -e <span class="string">"ssh -p 21822"</span> win10@pcrent365.com:/home/win10/3T/RNASeq/data/ <span class="string">"/home/win/4T/GeneDL/RNASeq/Count"</span></span><br></pre></td></tr></table></figure>
<h2 id="六、差异表达、富集分析、WGCNA分析"><a href="#六、差异表达、富集分析、WGCNA分析" class="headerlink" title="六、差异表达、富集分析、WGCNA分析"></a>六、差异表达、富集分析、WGCNA分析</h2><p><strong>差异分析代码</strong></p>
<figure class="highlight python"><table><tr><td class="code"><pre><span class="line"><span class="keyword">import</span> omicverse <span class="keyword">as</span> ov</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"><span class="keyword">import</span> matplotlib.pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">import</span> seaborn <span class="keyword">as</span> sns</span><br><span class="line"><span class="keyword">from</span> icecream <span class="keyword">import</span> install,ic</span><br><span class="line">install()</span><br><span class="line"><span class="keyword">import</span> pickle</span><br><span class="line">sc.settings.verbosity = <span class="number">3</span>             <span class="comment"># verbosity: errors (0), warnings (1), info (2), hints (3)</span></span><br><span class="line">sc.settings.set_figure_params(dpi=<span class="number">80</span>, facecolor=<span class="string">'white'</span>)</span><br><span class="line">species=<span class="string">"indica"</span></span><br><span class="line">test=<span class="number">0</span></span><br><span class="line"><span class="keyword">if</span> test==<span class="number">1</span>:</span><br><span class="line">    exp_datapath=<span class="string">"/home/win/4T/GeneDL/OSDrought_GCNAT_Link/data/japonica_1134/RNASeq/Train_validation_test/ExpressionData_unique_networkfilter.csv"</span>  </span><br><span class="line"><span class="keyword">else</span>:</span><br><span class="line">    exp_datapath= <span class="string">f"/home/win/4T/GeneDL/OSDrought_GCNAT_Link/data/multispecies/TF_regulation_filter_data/<span class="subst">&#123;species&#125;</span>/ExpressionData_unique_networkfilter.pkl"</span><span class="comment">#</span></span><br><span class="line"></span><br><span class="line">gene_file=pd.read_csv((<span class="string">"/home/win/4T/GeneDL/OSDrought_GCNAT_Link/data/multispecies/TF_regulation_filter_data/zeamays/ALL_Genefile.csv"</span>))</span><br><span class="line"><span class="comment">#ov.utils.download_geneid_annotation_pair()</span></span><br><span class="line"><span class="comment">#data=pd.read_csv('https://raw.githubusercontent.com/Starlitnightly/omicverse/master/sample/counts.txt',index_col=0,sep='\t',header=1)</span></span><br><span class="line">ref_data=pd.read_csv(<span class="string">'/home/win/4T/GeneDL/OSDrought_GCNAT_Link/data/japonica_1134/RNASeq/DEG_data.csv'</span>)  <span class="comment">#count数据</span></span><br><span class="line">ic(ref_data)</span><br><span class="line"><span class="keyword">if</span> test==<span class="number">1</span>:</span><br><span class="line">    data=ref_data</span><br><span class="line"><span class="keyword">else</span>:</span><br><span class="line">    <span class="keyword">with</span> open(<span class="string">f'/home/win/4T/GeneDL/OSDrought_GCNAT_Link/data/multispecies/TF_regulation_filter_data/<span class="subst">&#123;species&#125;</span>/label_sample_genecount.pkl'</span>,<span class="string">'rb'</span>) <span class="keyword">as</span> fp:</span><br><span class="line">        data=pickle.load(fp)</span><br><span class="line">    ic(data)</span><br><span class="line">    new_columns = data.iloc[<span class="number">0</span>]</span><br><span class="line">    data.columns = new_columns</span><br><span class="line">    data = data.iloc[<span class="number">1</span>:].copy()  <span class="comment"># 跳过第一行数据</span></span><br><span class="line">    data = data.astype(&#123;col: int <span class="keyword">for</span> col <span class="keyword">in</span> data.columns[<span class="number">1</span>:]&#125;)  <span class="comment"># 除了第一列外的列转换为int类型</span></span><br><span class="line">    <span class="comment"># 重置索引，这里直接在构造新DataFrame时避免了不必要的操作，因为新DataFrame默认已有连续的索引</span></span><br><span class="line">    data.reset_index(drop=<span class="keyword">True</span>, inplace=<span class="keyword">True</span>)</span><br><span class="line">    <span class="comment">#data.set_index(data.columns[0], inplace=True)</span></span><br><span class="line"></span><br><span class="line">    counter = &#123;&#125;</span><br><span class="line">    new_columns = []</span><br><span class="line">    <span class="keyword">for</span> col <span class="keyword">in</span> data.columns:</span><br><span class="line">        counter[col] = counter.get(col, <span class="number">0</span>) + <span class="number">1</span></span><br><span class="line">        new_col_name = <span class="string">f"<span class="subst">&#123;col&#125;</span>.<span class="subst">&#123;counter[col]&#125;</span>"</span> <span class="keyword">if</span> counter[col] &gt; <span class="number">1</span> <span class="keyword">else</span> col</span><br><span class="line">        new_columns.append(new_col_name)</span><br><span class="line">    <span class="comment">#data.columns = new_columns</span></span><br><span class="line">    <span class="comment">#ic(new_columns)</span></span><br><span class="line">    data=pd.DataFrame(data.values,index=list(data.index),columns=new_columns)</span><br><span class="line">ic(data.shape)</span><br><span class="line"></span><br><span class="line">data.iloc[:,<span class="number">1</span>:] = data.iloc[:,<span class="number">1</span>:].astype(<span class="string">'float'</span>) + <span class="number">1e-16</span></span><br><span class="line"></span><br><span class="line">ic(data)</span><br><span class="line"></span><br><span class="line"><span class="comment">#data = data[data['FEATURE_ID'].isin(gene_file['Gene'])]</span></span><br><span class="line">ic(data.shape)</span><br><span class="line"></span><br><span class="line"><span class="string">'''示例数据格式  count数据</span></span><br><span class="line"><span class="string">GeneID,CD,CD,CD,CD,CD</span></span><br><span class="line"><span class="string">Os01g0100100,521,762,833,672</span></span><br><span class="line"><span class="string">Os01g0100200,1,0,0,2,0,0,0</span></span><br><span class="line"><span class="string">'''</span></span><br><span class="line"></span><br><span class="line">cd_columns = data.filter(like=<span class="string">'CD'</span>, axis=<span class="number">1</span>)</span><br><span class="line">ck_columns = data.filter(like=<span class="string">'CK'</span>, axis=<span class="number">1</span>)</span><br><span class="line">ce_columns = data.filter(like=<span class="string">'CE'</span>, axis=<span class="number">1</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> test==<span class="number">0</span>:</span><br><span class="line">    cd_columns.reset_index(inplace=<span class="keyword">True</span>)</span><br><span class="line">    ck_columns.reset_index(inplace=<span class="keyword">True</span>)</span><br><span class="line">    ce_columns.reset_index(inplace=<span class="keyword">True</span>)</span><br><span class="line">    cd_columns.drop(columns=cd_columns.columns[<span class="number">0</span>], inplace=<span class="keyword">True</span>)</span><br><span class="line">    ck_columns.drop(columns=ck_columns.columns[<span class="number">0</span>], inplace=<span class="keyword">True</span>)</span><br><span class="line">    ce_columns.drop(columns=ce_columns.columns[<span class="number">0</span>], inplace=<span class="keyword">True</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">#ic(ck_columns,cd_columns)</span></span><br><span class="line"><span class="comment">#ic(cd_columns.shape,ck_columns.shape,ce_columns.shape)</span></span><br><span class="line">deg_data_ck_cd =  pd.concat([ck_columns, cd_columns], axis=<span class="number">1</span>).astype(float)</span><br><span class="line">deg_data_ck_ce =  pd.concat([ck_columns, ce_columns], axis=<span class="number">1</span>).astype(float)</span><br><span class="line">deg_data_cd_ce = pd.concat([cd_columns, ce_columns], axis=<span class="number">1</span>).astype(float)</span><br><span class="line"></span><br><span class="line">dds_ck_cd=ov.bulk.pyDEG(deg_data_ck_cd)</span><br><span class="line">dds_ck_cd.drop_duplicates_index()</span><br><span class="line">print(<span class="string">'... drop_duplicates_index success'</span>)</span><br><span class="line">dds_ck_cd.normalize()</span><br><span class="line">print(<span class="string">'... estimateSizeFactors and normalize success'</span>)</span><br><span class="line"></span><br><span class="line">dds_ck_ce=ov.bulk.pyDEG(deg_data_ck_ce)</span><br><span class="line">dds_ck_ce.drop_duplicates_index()</span><br><span class="line">print(<span class="string">'... drop_duplicates_index success'</span>)</span><br><span class="line">dds_ck_ce.normalize()</span><br><span class="line">print(<span class="string">'... estimateSizeFactors and normalize success'</span>)</span><br><span class="line"></span><br><span class="line">dds_cd_ce=ov.bulk.pyDEG(deg_data_cd_ce)</span><br><span class="line">dds_cd_ce.drop_duplicates_index()</span><br><span class="line">print(<span class="string">'... drop_duplicates_index success'</span>)</span><br><span class="line">dds_cd_ce.normalize()</span><br><span class="line">print(<span class="string">'... estimateSizeFactors and normalize success'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># cd_columns = cd_columns.astype('float16')</span></span><br><span class="line"><span class="comment"># ce_columns = ce_columns.astype('float16')</span></span><br><span class="line"><span class="comment"># ck_columns = ck_columns.astype('float16')</span></span><br><span class="line"><span class="comment"># ic(cd_columns,ck_columns,ce_columns)</span></span><br><span class="line">treatment_groups_cd=cd_columns.columns<span class="comment">#.tolist()</span></span><br><span class="line">treatment_groups_ce=ce_columns.columns<span class="comment">#.tolist()</span></span><br><span class="line">control_groups_ck=ck_columns.columns<span class="comment">#.tolist()</span></span><br><span class="line">ic(treatment_groups_cd,treatment_groups_ce,control_groups_ck)</span><br><span class="line"><span class="comment">#ic(dds_ck_cd)</span></span><br><span class="line">result_ck_cd=dds_ck_cd.deg_analysis(treatment_groups_cd,control_groups_ck,method=<span class="string">'ttest'</span>)</span><br><span class="line"><span class="comment">#ic(result_ck_cd)</span></span><br><span class="line">result_ck_cd = pd.DataFrame(result_ck_cd)</span><br><span class="line">combined_df_ck_cd = pd.concat([data.iloc[:, [<span class="number">0</span>]], result_ck_cd], axis=<span class="number">1</span>)</span><br><span class="line"></span><br><span class="line">result_ck_ce=dds_ck_ce.deg_analysis(treatment_groups_ce,control_groups_ck,method=<span class="string">'ttest'</span>)</span><br><span class="line">result_ck_ce = pd.DataFrame(result_ck_ce)</span><br><span class="line">combined_df_ck_ce = pd.concat([ data.iloc[:, [<span class="number">0</span>]], result_ck_ce], axis=<span class="number">1</span>)</span><br><span class="line"></span><br><span class="line">result_cd_ce=dds_cd_ce.deg_analysis(treatment_groups_cd,treatment_groups_ce,method=<span class="string">'ttest'</span>)</span><br><span class="line">result_cd_ce = pd.DataFrame(result_cd_ce)</span><br><span class="line">combined_df_cd_ce = pd.concat([ data.iloc[:, [<span class="number">0</span>]], result_cd_ce], axis=<span class="number">1</span>)</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">label_data</span><span class="params">(row)</span>:</span></span><br><span class="line">    <span class="keyword">if</span> row[<span class="string">'sig'</span>] == <span class="string">'sig'</span> <span class="keyword">and</span> row[<span class="string">'log2FC'</span>] &gt; <span class="number">0</span>:</span><br><span class="line">        <span class="keyword">return</span> <span class="number">1</span></span><br><span class="line">    <span class="keyword">elif</span> row[<span class="string">'sig'</span>] == <span class="string">'sig'</span> <span class="keyword">and</span> row[<span class="string">'log2FC'</span>] &lt; <span class="number">0</span>:</span><br><span class="line">        <span class="keyword">return</span> <span class="number">-1</span></span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        <span class="keyword">return</span> <span class="number">0</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 在 df 上应用该函数，并将结果赋给 'Label' 列</span></span><br><span class="line">combined_df_ck_cd[<span class="string">'Label'</span>] = combined_df_ck_cd.apply(label_data, axis=<span class="number">1</span>)</span><br><span class="line"><span class="keyword">if</span> test==<span class="number">1</span>:</span><br><span class="line">    exp_data = pd.read_csv(<span class="string">f'<span class="subst">&#123;exp_datapath&#125;</span>'</span>)</span><br><span class="line"><span class="keyword">else</span>:</span><br><span class="line">    <span class="keyword">with</span> open(<span class="string">f'<span class="subst">&#123;exp_datapath&#125;</span>'</span>,<span class="string">'rb'</span>) <span class="keyword">as</span> fp:</span><br><span class="line">        exp_data=pickle.load(fp)</span><br><span class="line"></span><br><span class="line">exp_data[<span class="string">'Label'</span>] =combined_df_ck_cd.apply(label_data, axis=<span class="number">1</span>)</span><br><span class="line">combined_df_ck_cd.to_csv(<span class="string">f'/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_DEG/<span class="subst">&#123;species&#125;</span>_CK_CD_DEG_output.csv'</span>,index=<span class="keyword">False</span>)</span><br><span class="line"></span><br><span class="line">df = combined_df_ck_cd</span><br><span class="line">df[<span class="string">'-log10pvalue'</span>] = -np.log10(df[<span class="string">'pvalue'</span>])</span><br><span class="line"><span class="comment"># 筛选-log10pvalue大于0.05的行，并进一步筛选出abs(log2FC) &gt; 0.585的行</span></span><br><span class="line"><span class="comment">#selected_genes = df[(df['-log10pvalue'] &gt; 0.05) &amp; (df['log2FC'].abs() &gt; 0.585)]</span></span><br><span class="line">selected_genes_up = df[(df[<span class="string">'pvalue'</span>] &lt; <span class="number">0.05</span>) &amp; (df[<span class="string">'log2FC'</span>] &gt; <span class="number">2</span>)]</span><br><span class="line">selected_genes_up.to_csv(<span class="string">f'/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_DEG/<span class="subst">&#123;species&#125;</span>_CK_CD_DEG_up.csv'</span>,index=<span class="keyword">False</span>)</span><br><span class="line">selected_genes_down = df[(df[<span class="string">'pvalue'</span>] &lt; <span class="number">0.05</span>) &amp; (df[<span class="string">'log2FC'</span>] &lt; <span class="number">-2</span>)]</span><br><span class="line">selected_genes_down.to_csv(<span class="string">f'/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_DEG/<span class="subst">&#123;species&#125;</span>_CK_CD_DEG_down.csv'</span>,index=<span class="keyword">False</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 根据abs(log2FC)降序排列</span></span><br><span class="line"><span class="comment">#sorted_genes = selected_genes.sort_values(by='abs(log2FC)', key=abs, ascending=False)</span></span><br><span class="line"><span class="comment">#sorted_genes.to_csv(f'/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_DEG/&#123;species&#125;_CK_CD_DEG_ascending.csv',index=False)</span></span><br><span class="line"><span class="comment"># 计数log2FC为正数的行（上调基因）</span></span><br><span class="line"><span class="comment"># upregulated_genes_count = sorted_genes[sorted_genes['log2FC'] &gt; 0].shape[0]</span></span><br><span class="line"><span class="comment"># # 计数log2FC为负数的行（下调基因）</span></span><br><span class="line"><span class="comment"># downregulated_genes_count = sorted_genes[sorted_genes['log2FC'] &lt; 0].shape[0]</span></span><br><span class="line"><span class="comment"># 打印结果</span></span><br><span class="line">print(<span class="string">f"上调基因数量：<span class="subst">&#123;len(selected_genes_up)&#125;</span>"</span>)</span><br><span class="line">print(<span class="string">f"下调基因数量：<span class="subst">&#123;len(selected_genes_down)&#125;</span>"</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">combined_df_ck_ce[<span class="string">'Label'</span>] = combined_df_ck_ce.apply(label_data, axis=<span class="number">1</span>)</span><br><span class="line"><span class="comment"># exp_data = pd.read_csv('/home/win/4T/GeneDL/OSDrought_GCNAT_Link/data/RNASeq/Train_validation_test/ExpressionData_unique_networkfilter.csv')</span></span><br><span class="line"><span class="comment"># exp_data['Label'] =combined_df_ck_ce.apply(label_data, axis=1)</span></span><br><span class="line">combined_df_ck_ce.to_csv(<span class="string">f'/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_DEG/<span class="subst">&#123;species&#125;</span>_CK_CE_DEG_output.csv'</span>,index=<span class="keyword">False</span>)</span><br><span class="line"></span><br><span class="line">df = combined_df_ck_ce</span><br><span class="line">df[<span class="string">'-log10pvalue'</span>] = -np.log10(df[<span class="string">'pvalue'</span>])</span><br><span class="line"><span class="comment"># 筛选-log10pvalue大于0.05的行，并进一步筛选出abs(log2FC) &gt; 0.585的行</span></span><br><span class="line">selected_genes_up = df[(df[<span class="string">'pvalue'</span>] &lt; <span class="number">0.05</span>) &amp; (df[<span class="string">'log2FC'</span>] &gt; <span class="number">2</span>)]</span><br><span class="line">selected_genes_up.to_csv(<span class="string">f'/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_DEG/<span class="subst">&#123;species&#125;</span>_CK_CE_DEG_up.csv'</span>,index=<span class="keyword">False</span>)</span><br><span class="line">selected_genes_down = df[(df[<span class="string">'pvalue'</span>] &lt; <span class="number">0.05</span>) &amp; (df[<span class="string">'log2FC'</span>] &lt; <span class="number">-2</span>)]</span><br><span class="line">selected_genes_down.to_csv(<span class="string">f'/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_DEG/<span class="subst">&#123;species&#125;</span>_CK_CE_DEG_down.csv'</span>,index=<span class="keyword">False</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 根据abs(log2FC)降序排列</span></span><br><span class="line"><span class="comment">#sorted_genes = selected_genes.sort_values(by='abs(log2FC)', key=abs, ascending=False)</span></span><br><span class="line"><span class="comment">#sorted_genes.to_csv(f'/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_DEG/&#123;species&#125;_CK_CD_DEG_ascending.csv',index=False)</span></span><br><span class="line"><span class="comment"># 计数log2FC为正数的行（上调基因）</span></span><br><span class="line"><span class="comment"># upregulated_genes_count = sorted_genes[sorted_genes['log2FC'] &gt; 0].shape[0]</span></span><br><span class="line"><span class="comment"># # 计数log2FC为负数的行（下调基因）</span></span><br><span class="line"><span class="comment"># downregulated_genes_count = sorted_genes[sorted_genes['log2FC'] &lt; 0].shape[0]</span></span><br><span class="line"><span class="comment"># 打印结果</span></span><br><span class="line">print(<span class="string">f"上调基因数量：<span class="subst">&#123;len(selected_genes_up)&#125;</span>"</span>)</span><br><span class="line">print(<span class="string">f"下调基因数量：<span class="subst">&#123;len(selected_genes_down)&#125;</span>"</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">combined_df_cd_ce[<span class="string">'Label'</span>] = combined_df_cd_ce.apply(label_data, axis=<span class="number">1</span>)</span><br><span class="line"><span class="comment"># exp_data = pd.read_csv('/home/win/4T/GeneDL/OSDrought_GCNAT_Link/data/RNASeq/Train_validation_test/ExpressionData_unique_networkfilter.csv')</span></span><br><span class="line"><span class="comment"># exp_data['Label'] =combined_df_ck_ce.apply(label_data, axis=1)</span></span><br><span class="line">combined_df_cd_ce.to_csv(<span class="string">f'/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_DEG/<span class="subst">&#123;species&#125;</span>_CD_CE_DEG_output.csv'</span>,index=<span class="keyword">False</span>)</span><br><span class="line"></span><br><span class="line">df = combined_df_cd_ce</span><br><span class="line">df[<span class="string">'-log10pvalue'</span>] = -np.log10(df[<span class="string">'pvalue'</span>])</span><br><span class="line"><span class="comment"># 筛选-log10pvalue大于0.05的行，并进一步筛选出abs(log2FC) &gt; 0.585的行</span></span><br><span class="line"><span class="comment">#selected_genes = df[(df['-log10pvalue'] &gt; 0.05) &amp; (df['log2FC'].abs() &gt; 0.585)]</span></span><br><span class="line"><span class="comment"># 根据abs(log2FC)降序排列</span></span><br><span class="line">selected_genes_up = df[(df[<span class="string">'pvalue'</span>] &lt; <span class="number">0.05</span>) &amp; (df[<span class="string">'log2FC'</span>] &gt; <span class="number">2</span>)]</span><br><span class="line">selected_genes_up.to_csv(<span class="string">f'/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_DEG/<span class="subst">&#123;species&#125;</span>_CD_CE_DEG_up.csv'</span>,index=<span class="keyword">False</span>)</span><br><span class="line">selected_genes_down = df[(df[<span class="string">'pvalue'</span>] &lt; <span class="number">0.05</span>) &amp; (df[<span class="string">'log2FC'</span>] &lt; <span class="number">-2</span>)]</span><br><span class="line">selected_genes_down.to_csv(<span class="string">f'/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_DEG/<span class="subst">&#123;species&#125;</span>_CD_CE_DEG_down.csv'</span>,index=<span class="keyword">False</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># 根据abs(log2FC)降序排列</span></span><br><span class="line"><span class="comment">#sorted_genes = selected_genes.sort_values(by='abs(log2FC)', key=abs, ascending=False)</span></span><br><span class="line"><span class="comment">#sorted_genes.to_csv(f'/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_DEG/&#123;species&#125;_CK_CD_DEG_ascending.csv',index=False)</span></span><br><span class="line"><span class="comment"># 计数log2FC为正数的行（上调基因）</span></span><br><span class="line"><span class="comment"># upregulated_genes_count = sorted_genes[sorted_genes['log2FC'] &gt; 0].shape[0]</span></span><br><span class="line"><span class="comment"># # 计数log2FC为负数的行（下调基因）</span></span><br><span class="line"><span class="comment"># downregulated_genes_count = sorted_genes[sorted_genes['log2FC'] &lt; 0].shape[0]</span></span><br><span class="line"><span class="comment"># 打印结果</span></span><br><span class="line">print(<span class="string">f"上调基因数量：<span class="subst">&#123;len(selected_genes_up)&#125;</span>"</span>)</span><br><span class="line">print(<span class="string">f"下调基因数量：<span class="subst">&#123;len(selected_genes_down)&#125;</span>"</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">exp_data.to_csv(<span class="string">f'/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_DEG/<span class="subst">&#123;species&#125;</span>_ExpressionData_geneclassification.csv'</span>,index=<span class="keyword">False</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># ic(result.head())</span></span><br><span class="line"><span class="comment"># print(result.shape)</span></span><br><span class="line"><span class="comment"># result=result.loc[result['log2(BaseMean)']&gt;1]</span></span><br><span class="line"><span class="comment"># print(result.shape)</span></span><br><span class="line"><span class="comment">#-1 means automatically calculates</span></span><br><span class="line"><span class="comment"># dds.foldchange_set(fc_threshold=-1,</span></span><br><span class="line"><span class="comment">#                    pval_threshold=0.05,</span></span><br><span class="line"><span class="comment">#                    logp_max=6)</span></span><br><span class="line"><span class="comment"># dds.plot_volcano(title='DEG Analysis',figsize=(4,4),</span></span><br><span class="line"><span class="comment">#                  plot_genes_num=8,plot_genes_fontsize=12,)</span></span><br><span class="line"><span class="comment"># plt.savefig('/mnt/g/gene/course_try/Omicverse/out/DEG_Analysis.png',bbox_inches='tight')</span></span><br><span class="line"><span class="comment"># dds.plot_boxplot(genes=['Os12t0152900-01-E5','Os06t0230301-00-E1'],treatment_groups=treatment_groups,</span></span><br><span class="line"><span class="comment">#                 control_groups=control_groups,figsize=(2,3),fontsize=12,</span></span><br><span class="line"><span class="comment">#                  legend_bbox=(2,0.55))</span></span><br><span class="line"><span class="comment"># plt.savefig('/mnt/g/gene/course_try/Omicverse/out/Ckap2_Lef1.png',bbox_inches='tight')</span></span><br><span class="line"><span class="comment"># dds.plot_boxplot(genes=['Os12t0152900-01-E5'],treatment_groups=treatment_groups,</span></span><br><span class="line"><span class="comment">#                 control_groups=control_groups,figsize=(2,3),fontsize=12,</span></span><br><span class="line"><span class="comment">#                  legend_bbox=(2,0.55))</span></span><br><span class="line"><span class="comment"># plt.savefig('/mnt/g/gene/course_try/Omicverse/out/Ckap2.png',bbox_inches='tight')</span></span><br></pre></td></tr></table></figure>
<p><strong>富集分析代码GO</strong></p>
<figure class="highlight plain"><table><tr><td class="code"><pre><span class="line">import gseapy as gp</span><br><span class="line">import os</span><br><span class="line">from plot_new import barplot, dotplot</span><br><span class="line">import pandas as pd</span><br><span class="line">import matplotlib.pyplot as plt</span><br><span class="line">from icecream import install,ic</span><br><span class="line">install()</span><br><span class="line">plt.rcParams[&apos;font.sans-serif&apos;]=[&apos;Times New Roman&apos;]</span><br><span class="line">plt.rcParams[&apos;mathtext.fontset&apos;] = &apos;stix&apos;#和Times new roman 最像</span><br><span class="line"></span><br><span class="line">#-------------------------------获取Geneid和GOterm---------------------------------------------------</span><br><span class="line"># 假设文件是以制表符分隔的，将 &apos;sep&apos; 参数替换成 &apos;,&apos; 如果是逗号分隔的CSV</span><br><span class="line">#species=&apos;zeamays&apos;</span><br><span class="line">species=&apos;japonica&apos;</span><br><span class="line">#file_path = &apos;/home/win/4T/GeneDL/OSDrought_GCNAT_Link/data/multispecies/initial_data/zeamays/Zma_GO_annotation&apos;</span><br><span class="line">file_path =&apos;/home/win/4T/GeneDL/OSDrought_GCNAT_Link/data/japonica_1134/GOcluster/Osj_GO_annotation&apos; # 或者 &apos;Osj_GO_annotation.csv&apos;</span><br><span class="line"></span><br><span class="line"># 读取文件，只获取前两列</span><br><span class="line"># header参数假设文件有标题行，如果没有请设定为None</span><br><span class="line">GO_term_data = pd.read_csv(file_path, sep=&apos;\t&apos;, usecols=[0, 2,3], header=0)</span><br><span class="line">if species==&apos;zeamays&apos;:</span><br><span class="line">    relation = []</span><br><span class="line">    # 使用open函数和with语句读取文件</span><br><span class="line">    with open(&quot;/home/win/4T/GeneDL/OSDrought_GCNAT_Link/data/multispecies/initial_data/zeamays/gene_id_trans.txt&quot;, &apos;r&apos;) as file:</span><br><span class="line">        for line in file:</span><br><span class="line">            # 分割每行的数据，只取前两个值（原始基因名和第一个映射基因名）</span><br><span class="line">            parts = line.strip().split(&apos;\t&apos;)</span><br><span class="line">            if len(parts) &gt;= 2:  # 确保有足够的部分读取</span><br><span class="line">                original_gene = parts[0]</span><br><span class="line">                mapped_gene = parts[1]</span><br><span class="line">                # 保存结果</span><br><span class="line">                relation.append([original_gene, mapped_gene])</span><br><span class="line"></span><br><span class="line">    # 将列表转换为DataFrame</span><br><span class="line">    df3 = pd.DataFrame(relation, columns=[&apos;Original Gene&apos;, &apos;Mapped Gene&apos;])</span><br><span class="line">    # 创建映射字典</span><br><span class="line">    gene_id_map = pd.Series(df3[&apos;Mapped Gene&apos;].values, index=df3[&apos;Original Gene&apos;].values).to_dict()</span><br><span class="line">    #ic(gene_id_map)</span><br><span class="line">    #GO_term_data[&apos;zeaGeneid&apos;] = GO_term_data[&apos;Gene_id&apos;].apply(convert_msu_to_rap)</span><br><span class="line">    # 替换 column1, column2, column3 中的基因名</span><br><span class="line">    GO_term_data[&apos;zeaGeneid&apos;] = GO_term_data[&apos;Gene_id&apos;].map(gene_id_map).fillna(GO_term_data[&apos;Gene_id&apos;])  # fillna用于保留未找到映射的原始值</span><br><span class="line">    # relation=&#123;&#125;</span><br><span class="line">    # for i in open(&quot;/home/win/4T/GeneDL/OSDrought_GCNAT_Link/data/multispecies/initial_data/zeamays/gene_id_trans.txt&quot;):</span><br><span class="line">    #     rap=str(i.split()[0])</span><br><span class="line">    #     msu=str(i.split()[1])</span><br><span class="line">    #     #ic(rap)</span><br><span class="line">    #     #ic(msu)</span><br><span class="line">    #     if msu!=&quot;None&quot;:</span><br><span class="line">    #         if &quot; &quot; in msu:</span><br><span class="line">    #             for a in msu.split(&quot; &quot;):</span><br><span class="line">    #                 relation[a] = rap</span><br><span class="line">    #         else:</span><br><span class="line">    #             relation[msu] = rap</span><br><span class="line">    # #ic(relation)</span><br><span class="line"></span><br><span class="line">    # # 转换函数: 如果找不到对应的MSU标识符，则返回原RAP标识符</span><br><span class="line">    # def convert_rap_to_msu(rap_id):</span><br><span class="line">    #     return relation.get(rap_id, rap_id)</span><br><span class="line">    # def convert_msu_to_rap(msu_id):</span><br><span class="line">    #     return relation.get(msu_id,None)</span><br><span class="line">    </span><br><span class="line">    #GO_term_data[&apos;zeaGeneid&apos;] = GO_term_data[&apos;Gene_id&apos;].apply(convert_msu_to_rap)</span><br><span class="line"></span><br><span class="line">    def create_gene_sets(df):</span><br><span class="line">        &quot;&quot;&quot;</span><br><span class="line">        基于DataFrame创建基因集字典。DataFrame应包含两列：GO_term和OSGeneid。</span><br><span class="line">        </span><br><span class="line">        :param df: 包含GO_term和OSGeneid的DataFrame。</span><br><span class="line">        :return: 基因集字典，键是GO_term，值是OSGeneid列表。</span><br><span class="line">        &quot;&quot;&quot;</span><br><span class="line">        gene_sets = &#123;&#125;</span><br><span class="line">        grouped = df.groupby(&apos;GO_term&apos;)</span><br><span class="line">        </span><br><span class="line">        for name, group in grouped:</span><br><span class="line">            gene_sets[name] = group[&apos;zeaGeneid&apos;].tolist()</span><br><span class="line">        return gene_sets</span><br><span class="line">else:</span><br><span class="line">    relation=&#123;&#125;</span><br><span class="line">    for i in open(&quot;/home/win/4T/GeneDL/OSDrought_GCNAT_Link/data_proceed/MSU2RAP/RAP-MSU_2018-03-29.txt&quot;):</span><br><span class="line">        rap=str(i.split()[0])</span><br><span class="line">        msu=str(i.split()[1])</span><br><span class="line">        if msu!=&quot;None&quot;:</span><br><span class="line">            if &quot;,&quot; in msu:</span><br><span class="line">                for a in msu.split(&quot;,&quot;):</span><br><span class="line">                    relation[a[0:-2]] = rap</span><br><span class="line">            else:</span><br><span class="line">                relation[msu[0:-2]] = rap</span><br><span class="line">    #ic(relation)</span><br><span class="line"></span><br><span class="line">    # 转换函数: 如果找不到对应的MSU标识符，则返回原RAP标识符</span><br><span class="line">    def convert_rap_to_msu(rap_id):</span><br><span class="line">        return relation.get(rap_id, rap_id)</span><br><span class="line">    def convert_msu_to_rap(msu_id):</span><br><span class="line">        return relation.get(msu_id, msu_id)</span><br><span class="line"></span><br><span class="line">    GO_term_data[&apos;OSGeneid&apos;] = GO_term_data[&apos;Gene_id&apos;].apply(convert_rap_to_msu)</span><br><span class="line"></span><br><span class="line">    def create_gene_sets(df):</span><br><span class="line">        &quot;&quot;&quot;</span><br><span class="line">        基于DataFrame创建基因集字典。DataFrame应包含两列：GO_term和OSGeneid。</span><br><span class="line">        </span><br><span class="line">        :param df: 包含GO_term和OSGeneid的DataFrame。</span><br><span class="line">        :return: 基因集字典，键是GO_term，值是OSGeneid列表。</span><br><span class="line">        &quot;&quot;&quot;</span><br><span class="line">        gene_sets = &#123;&#125;</span><br><span class="line">        grouped = df.groupby(&apos;GO_term&apos;)</span><br><span class="line">        </span><br><span class="line">        for name, group in grouped:</span><br><span class="line">            gene_sets[name] = group[&apos;OSGeneid&apos;].tolist()</span><br><span class="line">        </span><br><span class="line">        return gene_sets</span><br><span class="line">#ic(len(GO_term_data[&apos;Gene_id&apos;].unique()))</span><br><span class="line">#ic(GO_term_data)</span><br><span class="line">#ic(GO_term_data[GO_term_data[&apos;zeaGeneid&apos;].notna()][&apos;zeaGeneid&apos;])</span><br><span class="line">grouped_dataframes = &#123;aspect: group for aspect, group in GO_term_data.groupby(&apos;Aspect&apos;)&#125;</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">all_gene_sets=[]</span><br><span class="line">for aspect, sub_df in grouped_dataframes.items():</span><br><span class="line">    ic(f&quot;Aspect: &#123;aspect&#125;&quot;)</span><br><span class="line">    ic(sub_df)</span><br><span class="line"></span><br><span class="line">    gene_sets=create_gene_sets(sub_df)</span><br><span class="line">    for key in gene_sets.keys():</span><br><span class="line">        gene_sets[key] = [item for item in gene_sets[key] if item != &apos;None&apos;]</span><br><span class="line">    all_gene_sets.append(gene_sets)</span><br><span class="line">#ic(gene_sets)</span><br><span class="line"># max_len = max(len(lst) for lst in gene_sets.values())</span><br><span class="line"># for key in gene_sets.keys():</span><br><span class="line">#     gene_sets[key].extend([None] * (max_len - len(gene_sets[key])))</span><br><span class="line"># df = pd.DataFrame(gene_sets)</span><br><span class="line"># df.T.to_csv(&apos;/mnt/e/GeneDL/OSDrought_GCNAT_Link/plot/data/GO_term_output.csv&apos;, index=True)</span><br><span class="line">#ic(gene_sets)</span><br><span class="line">#自己创建基因集gene_sets=&#123;&apos;term_A&apos;:[&apos;gene1&apos;, &apos;gene2&apos;,...], </span><br><span class="line">           #            &apos;term_B&apos;:[&apos;gene2&apos;, &apos;gene4&apos;,...], ...&#125;</span><br><span class="line"># # # gene_sets = &apos;KEGG_2016&apos;</span><br><span class="line">data_list=[&apos;CK_CD&apos;,</span><br><span class="line">            &apos;CK_CE&apos;,</span><br><span class="line">            &apos;CD_CE&apos;,</span><br><span class="line">            ]</span><br><span class="line">#gene_sets=[&apos;KEGG_2016&apos;,&apos;KEGG_2013&apos;]</span><br><span class="line"></span><br><span class="line"># from bioservices import KEGG</span><br><span class="line"># kegg = KEGG()</span><br><span class="line"># # 查询特定物种的某个通路</span><br><span class="line"># pathway_info = kegg.get_pathway_by_gene(&quot;Os03g0719850&quot;, &quot;osa&quot;)</span><br><span class="line"># print(pathway_info)</span><br><span class="line"># # 或下载通路图</span><br><span class="line"># kegg.save_pathway(&quot;path:osa00010&quot;, filename=&quot;pathway_image.png&quot;)</span><br><span class="line">funcs=[&apos;CC&apos;,&apos;MF&apos;,&apos;BP&apos;]</span><br><span class="line">label=[&apos;up&apos;,&apos;down&apos;]</span><br><span class="line">colors=[&apos;#2e86de&apos;,&apos;#c0392b&apos;,&apos;#28b463&apos;]</span><br><span class="line">for cla in label:</span><br><span class="line">    for data_name in data_list:</span><br><span class="line">        all_results=[]</span><br><span class="line">        for i,gene_sets in enumerate(all_gene_sets):</span><br><span class="line">            data = pd.read_csv(f&apos;/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_DEG/&#123;species&#125;_&#123;data_name&#125;_DEG_&#123;cla&#125;.csv&apos;)#[&apos;DAB2IP&apos;, &apos;AKT1&apos;, &apos;ARNT&apos;, &apos;FLCN&apos;, &apos;XRCC6&apos;, &apos;BTG1&apos;, &apos;STAT1&apos;, &apos;STAT1&apos;]</span><br><span class="line">            #ic(data)</span><br><span class="line">            #data = data[data[&apos;sig&apos;] == &apos;sig&apos;]</span><br><span class="line">            #ic(data)</span><br><span class="line">            gene_list = data[&apos;FEATURE_ID&apos;]#[:5000]##[&apos;GeneID&apos;]</span><br><span class="line">            gene_list = [str(i) for i in gene_list]</span><br><span class="line">            #gene_list = pd.read_csv(&apos;/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/data/top_diff_CD_CE_degree.csv&apos;)</span><br><span class="line">            #gene_list = list(gene_list)</span><br><span class="line">            #gene_list = gene_list[&apos;0&apos;].apply(convert_msu_to_rap)</span><br><span class="line">            # ic(gene_list)</span><br><span class="line">            # ic(gene_sets)</span><br><span class="line">            # GO富集</span><br><span class="line">            enr = gp.enrichr(gene_list=gene_list,#所需查询gene_list，可以是一个列表，也可为文件（一列，每行一个基因）</span><br><span class="line">                            gene_sets= gene_sets,#[&apos;GO_Biological_Process_2018&apos;],#,#gene set library，多个相关的gene set 。如所有GO term组成一个gene set library.</span><br><span class="line">                            #organism=&apos;rice&apos;,#持(human, mouse, yeast, fly, fish, worm)， 自定义gene_set 则无影响。</span><br><span class="line">                            #description=&apos;kegg&apos;,#工作运行描述</span><br><span class="line">                            outdir=f&apos;/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_enrichr/GO/&#123;species&#125;_&#123;data_name&#125;_&#123;cla&#125;&apos;,#输出目录</span><br><span class="line">                            top_term=20,</span><br><span class="line">                            cutoff=0.9,#pvalue阈值</span><br><span class="line">                            )</span><br><span class="line">            ic(enr.results)</span><br><span class="line">            #all_results.append(enr.results)</span><br><span class="line">            dot_png = f&quot;&#123;species&#125;_&#123;data_name&#125;_&#123;cla&#125;_&#123;funcs[i]&#125;&quot; + &quot;_&quot; + &quot;dot&quot; + &quot;.png&quot;</span><br><span class="line">            bar_png = f&quot;&#123;species&#125;_&#123;data_name&#125;_&#123;cla&#125;_&#123;funcs[i]&#125;&quot; + &quot;_&quot; + &quot;bar&quot; + &quot;.png&quot;</span><br><span class="line">            base_path = &quot;/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_enrichr/funcGO&quot;</span><br><span class="line"></span><br><span class="line">            if os.path.exists(base_path+dot_png):</span><br><span class="line">                os.remove(base_path+dot_png)</span><br><span class="line"></span><br><span class="line">            #fig, axs = plt.subplots(nrows=3, ncols=1, figsize=(5, 10))</span><br><span class="line">            cut=0.6</span><br><span class="line">            df= dotplot(enr.results,</span><br><span class="line">                    column=&quot;Adjusted P-value&quot;,</span><br><span class="line">                    size=5,</span><br><span class="line">                    #figsize=(3,5),</span><br><span class="line">                    xticklabels_rot=45,</span><br><span class="line">                    title=f&apos;GO Top10 Pathway &#123;species&#125; &#123;data_name&#125; &#123;cla&#125; &#123;funcs[i]&#125;&apos;, </span><br><span class="line">                    show_ring=True,</span><br><span class="line">                    #marker=&apos;o&apos;,</span><br><span class="line">                    cmap=&apos;RdBu&apos;,</span><br><span class="line">                    top_term=10, legend=&quot;r&quot;,pval=0.9,</span><br><span class="line">                    ofname=base_path+dot_png,</span><br><span class="line">                    cutoff=cut,</span><br><span class="line">                    )</span><br><span class="line">            if os.path.exists(base_path+bar_png):</span><br><span class="line">                os.remove(base_path+bar_png)</span><br><span class="line"></span><br><span class="line">            ax = barplot(enr.results,</span><br><span class="line">                        column=&quot;Adjusted P-value&quot;,</span><br><span class="line">                        title=f&quot;&#123;species&#125; &#123;data_name&#125; &#123;cla&#125; &#123;funcs[i]&#125;&quot;,</span><br><span class="line">                        #group=&apos;Gene_set&apos;, # set group, so you could do a multi-sample/library comparsion</span><br><span class="line">                        size=1,</span><br><span class="line">                        top_term=10,</span><br><span class="line">                        #figsize=(3,5),</span><br><span class="line">                        color=colors[i],#[&apos;#27ae60&apos;, &apos;#34495e&apos;] , # set colors for group</span><br><span class="line">                        ofname=base_path+bar_png,</span><br><span class="line">                        cutoff=cut</span><br><span class="line">                        #color = &#123;&apos;KEGG_2021_Human&apos;: &apos;salmon&apos;, &apos;MSigDB_Hallmark_2020&apos;:&apos;darkblue&apos;&#125;</span><br><span class="line">                        )</span><br><span class="line">            df = df.iloc[::-1].reset_index(drop=True)</span><br><span class="line">            df.to_csv(f&apos;/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_enrichr/GO/&#123;species&#125;_&#123;data_name&#125;_&#123;cla&#125;_&#123;funcs[i]&#125;.csv&apos;,index=False)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">        # from gseapy import enrichment_map</span><br><span class="line">        # import numpy as np</span><br><span class="line">        # # return two dataframe</span><br><span class="line">        # ic(enr.results)</span><br><span class="line">        # nodes, edges = enrichment_map(enr.results,cutoff=cut,top_term=20)</span><br><span class="line">        # ic(edges)</span><br><span class="line">        # import networkx as nx</span><br><span class="line">        # # build graph</span><br><span class="line">        # G = nx.from_pandas_edgelist(edges,</span><br><span class="line">        #                             source=&apos;src_idx&apos;,</span><br><span class="line">        #                             target=&apos;targ_idx&apos;,</span><br><span class="line">        #                             edge_attr=[&apos;jaccard_coef&apos;, &apos;overlap_coef&apos;, &apos;overlap_genes&apos;])</span><br><span class="line"></span><br><span class="line">        # fig, ax = plt.subplots()</span><br><span class="line"></span><br><span class="line">        # # init node cooridnates</span><br><span class="line">        # pos=nx.layout.spiral_layout(G)</span><br><span class="line">        # ic(pos.keys())</span><br><span class="line">        # #node_size = nx.get_node_attributes()</span><br><span class="line">        # # draw node</span><br><span class="line">        # nodelist = list(G)</span><br><span class="line">        # xy = np.asarray([pos[v] for v in nodelist])</span><br><span class="line">        # ic(nodelist)</span><br><span class="line">        # original_list = list(xy[:, 0])</span><br><span class="line">        # nx.draw_networkx_nodes(G,</span><br><span class="line">        #                     pos=pos,</span><br><span class="line">        #                     cmap=plt.cm.RdYlBu,</span><br><span class="line">        #                     node_color= [number * 100 for number in original_list],</span><br><span class="line">        #                     node_size= [number * 1000 for number in original_list])</span><br><span class="line">        # # draw node label</span><br><span class="line">        # ic(nodes.Term)</span><br><span class="line">        # ic(pos)</span><br><span class="line">        # node_term_dict = nodes.Term.to_dict()</span><br><span class="line">        # labels = &#123;v:node_term_dict[v] for v in nodelist&#125;</span><br><span class="line">        # nx.draw_networkx_labels(G,</span><br><span class="line">        #                         pos=pos,</span><br><span class="line">        #                         labels=labels)</span><br><span class="line">        # # draw edge</span><br><span class="line">        # edge_weight = nx.get_edge_attributes(G, &apos;jaccard_coef&apos;).values()</span><br><span class="line">        # nx.draw_networkx_edges(G,</span><br><span class="line">        #                     pos=pos,</span><br><span class="line">        #                     width=list(map(lambda x: x*10, edge_weight)),</span><br><span class="line">        #                     edge_color=&apos;#CDDBD4&apos;)</span><br><span class="line">        # #plt.show()</span><br><span class="line">        # plt.savefig(f&apos;/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_enrichr/GO/&#123;species&#125;_&#123;data_name&#125;_&#123;cla&#125;_grn.png&apos;)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"># ax = dotplot(enr.res2d, title=&apos;KEGG_2021_Human&apos;,cmap=&apos;viridis_r&apos;, size=10, figsize=(3,5))</span><br><span class="line"># ax = barplot(enr.res2d,title=&apos;KEGG_2021_Human&apos;, figsize=(4, 5), color=&apos;darkred&apos;)</span><br><span class="line">#barplot(enr.res2d, title=&apos;CK_CD_degree&apos;,top_term=20,ofname=base_path+bar_png,)</span><br></pre></td></tr></table></figure>
<p><strong>富集分析代码KEGG</strong></p>
<figure class="highlight python"><table><tr><td class="code"><pre><span class="line"><span class="keyword">import</span> gseapy <span class="keyword">as</span> gp</span><br><span class="line"><span class="keyword">import</span> os</span><br><span class="line"><span class="comment">#from gseapy.plot import barplot, dotplot</span></span><br><span class="line"><span class="keyword">from</span> plot_new <span class="keyword">import</span> barplot, dotplot</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> matplotlib.pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">from</span> icecream <span class="keyword">import</span> install,ic</span><br><span class="line">install()</span><br><span class="line">plt.rcParams[<span class="string">'font.sans-serif'</span>]=[<span class="string">'Times New Roman'</span>]</span><br><span class="line"><span class="comment">#-------------------------------获取Geneid和GOterm---------------------------------------------------</span></span><br><span class="line"><span class="keyword">import</span> json</span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line">species=<span class="string">'japonica'</span></span><br><span class="line">basepath=<span class="string">'/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_enrichr/KEGG/'</span></span><br><span class="line"><span class="keyword">with</span> open(<span class="string">f"<span class="subst">&#123;basepath&#125;</span>/<span class="subst">&#123;species&#125;</span>/kegg.json"</span>) <span class="keyword">as</span> f:</span><br><span class="line">    ko_map_data = json.load(f)</span><br><span class="line"></span><br><span class="line"><span class="keyword">with</span> open(<span class="string">f"<span class="subst">&#123;basepath&#125;</span>/<span class="subst">&#123;species&#125;</span>/KEGG_pathway_ko.txt"</span>, <span class="string">"w"</span>) <span class="keyword">as</span> oh:</span><br><span class="line">    line = <span class="string">"level1_pathway_id\tlevel1_pathway_name\tlevel2_pathway_id\tlevel2_pathway_name"</span></span><br><span class="line">    line += <span class="string">"\tlevel3_pathway_id\tlevel3_pathway_name\tko\tko_name\tko_des\tec\n"</span></span><br><span class="line">    oh.write(line)</span><br><span class="line">    <span class="keyword">for</span> level1 <span class="keyword">in</span> ko_map_data[<span class="string">"children"</span>]:</span><br><span class="line">        m = re.match(<span class="string">r"(\S+)\s+([\S\w\s]+)"</span>, level1[<span class="string">"name"</span>])</span><br><span class="line">        level1_pathway_id = m.groups()[<span class="number">0</span>].strip()</span><br><span class="line">        level1_pathway_name = m.groups()[<span class="number">1</span>].strip()</span><br><span class="line">        <span class="keyword">for</span> level2 <span class="keyword">in</span> level1[<span class="string">"children"</span>]:</span><br><span class="line">            m = re.match(<span class="string">r"(\S+)\s+([\S\w\s]+)"</span>, level2[<span class="string">"name"</span>])</span><br><span class="line">            level2_pathway_id = m.groups()[<span class="number">0</span>].strip()</span><br><span class="line">            level2_pathway_name = m.groups()[<span class="number">1</span>].strip()</span><br><span class="line">            <span class="keyword">for</span> level3 <span class="keyword">in</span> level2[<span class="string">"children"</span>]:</span><br><span class="line">                m = re.match(<span class="string">r"(\S+)\s+([^\[]*)"</span>, level3[<span class="string">"name"</span>])</span><br><span class="line">                level3_pathway_id = m.groups()[<span class="number">0</span>].strip()</span><br><span class="line">                level3_pathway_name = m.groups()[<span class="number">1</span>].strip()</span><br><span class="line"></span><br><span class="line">                <span class="keyword">if</span> <span class="string">"children"</span> <span class="keyword">in</span> level3:</span><br><span class="line">                    <span class="keyword">for</span> ko <span class="keyword">in</span> level3[<span class="string">"children"</span>]:</span><br><span class="line">                        <span class="keyword">if</span> species==<span class="string">'zeamays'</span>:</span><br><span class="line">                            pattern = <span class="string">r"(\S+)\s+([\S\s]+?);\s*([\S\s]*?)(?:\[EC:(\S+)\])*"</span></span><br><span class="line">                            m = re.match(pattern, ko[<span class="string">"name"</span>])</span><br><span class="line">                            <span class="keyword">if</span> m:</span><br><span class="line">                                ko_id = m.group(<span class="number">1</span>).strip()</span><br><span class="line">                                ko_des = m.group(<span class="number">2</span>).strip()</span><br><span class="line">                                ko_name = ko_id</span><br><span class="line">                                ec = m.group(<span class="number">4</span>) <span class="keyword">if</span> m.group(<span class="number">4</span>) <span class="keyword">else</span> <span class="string">"-"</span></span><br><span class="line">                            line = level1_pathway_id + <span class="string">"\t"</span> + level1_pathway_name + <span class="string">"\t"</span> + level2_pathway_id + <span class="string">"\t"</span> + level2_pathway_name</span><br><span class="line">                            line += <span class="string">"\t"</span> + level3_pathway_id + <span class="string">"\t"</span> + level3_pathway_name + <span class="string">"\t"</span> + ko_id + <span class="string">"\t"</span> + ko_name +  <span class="string">"\t"</span> + ko_des + <span class="string">"\t"</span> + ec + <span class="string">"\n"</span></span><br><span class="line">                            oh.write(line)</span><br><span class="line">                        <span class="keyword">else</span>:</span><br><span class="line">                            m = re.match(<span class="string">r"(\S+)\s+(\S+);\s+([^\[]+)\s*(\[EC:\S+(?:\s+[^\[\]]+)*\])*"</span>, ko[<span class="string">"name"</span>])</span><br><span class="line">                            <span class="comment">#ic(m)</span></span><br><span class="line">                            <span class="keyword">if</span> m <span class="keyword">is</span> <span class="keyword">not</span> <span class="keyword">None</span>:</span><br><span class="line">                                ko_id = m.groups()[<span class="number">0</span>].strip()</span><br><span class="line">                                ko_name = m.groups()[<span class="number">1</span>].strip()</span><br><span class="line">                                ko_des = m.groups()[<span class="number">2</span>].strip()</span><br><span class="line">                                ec = m.groups()[<span class="number">3</span>]</span><br><span class="line">                                <span class="keyword">if</span> ec==<span class="keyword">None</span>:</span><br><span class="line">                                    ec = <span class="string">"-"</span></span><br><span class="line">                            line = level1_pathway_id + <span class="string">"\t"</span> + level1_pathway_name + <span class="string">"\t"</span> + level2_pathway_id + <span class="string">"\t"</span> + level2_pathway_name</span><br><span class="line">                            line += <span class="string">"\t"</span> + level3_pathway_id + <span class="string">"\t"</span> + level3_pathway_name + <span class="string">"\t"</span> + ko_id + <span class="string">"\t"</span> + ko_name + <span class="string">"\t"</span> + ko_des + <span class="string">"\t"</span> + ec + <span class="string">"\n"</span></span><br><span class="line">                            oh.write(line)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"></span><br><span class="line">data = pd.read_csv(<span class="string">f"<span class="subst">&#123;basepath&#125;</span>/<span class="subst">&#123;species&#125;</span>/KEGG_pathway_ko.txt"</span>, sep=<span class="string">"\t"</span>,dtype=str)</span><br><span class="line"></span><br><span class="line">data = data.drop_duplicates()</span><br><span class="line"></span><br><span class="line">data.to_csv(<span class="string">f"<span class="subst">&#123;basepath&#125;</span>/<span class="subst">&#123;species&#125;</span>/KEGG_pathway_ko_uniq.csv"</span>, index=<span class="keyword">False</span>)</span><br><span class="line"></span><br><span class="line">file_path = <span class="string">f'<span class="subst">&#123;basepath&#125;</span>/<span class="subst">&#123;species&#125;</span>/KEGG_pathway_ko_uniq.csv'</span></span><br><span class="line"><span class="comment"># 读取CSV文件中的特定两列</span></span><br><span class="line">df = pd.read_csv(file_path, usecols=[<span class="string">'level3_pathway_id'</span>, <span class="string">'ko'</span>])</span><br><span class="line">ic(len(df[<span class="string">'level3_pathway_id'</span>].unique()))</span><br><span class="line">ic(df)</span><br><span class="line"><span class="keyword">if</span> species==<span class="string">'zeamays'</span>:</span><br><span class="line">    <span class="comment">#df['ko'].to_csv('/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_enrichr/KEGG/zeamays/entre_id.csv',index=False)</span></span><br><span class="line">    mapdata=pd.read_csv(<span class="string">'/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_enrichr/KEGG/zeamays/stableid_entreid.txt'</span>,sep=<span class="string">'\t'</span>)</span><br><span class="line">    ic(mapdata)</span><br><span class="line">    new_df = mapdata[[<span class="string">'Gene stable ID'</span>, <span class="string">'NCBI gene (formerly Entrezgene) ID'</span>]].drop_duplicates()</span><br><span class="line">    mapping = pd.Series(new_df[<span class="string">'Gene stable ID'</span>].values, index=new_df[<span class="string">'NCBI gene (formerly Entrezgene) ID'</span>]).to_dict()</span><br><span class="line">    df[<span class="string">'ko'</span>] = df[<span class="string">'ko'</span>].map(mapping)</span><br><span class="line">    ic(df[<span class="string">'ko'</span>])</span><br><span class="line"></span><br><span class="line">pd.DataFrame(df[<span class="string">'level3_pathway_id'</span>].unique()).to_csv(<span class="string">f'<span class="subst">&#123;basepath&#125;</span>/<span class="subst">&#123;species&#125;</span>/unique_kegg_term.csv'</span>,index=<span class="keyword">False</span>)</span><br><span class="line"><span class="comment"># 初始化基因集字典</span></span><br><span class="line">gene_sets = &#123;&#125;</span><br><span class="line"><span class="comment"># 遍历DataFrame来构建基因集字典</span></span><br><span class="line"><span class="keyword">for</span> index, row <span class="keyword">in</span> df.iterrows():</span><br><span class="line">    <span class="comment"># 从当前行获取pathway ID和KO基因ID</span></span><br><span class="line">    pathway_id = row[<span class="string">'level3_pathway_id'</span>]</span><br><span class="line">    ko_gene = row[<span class="string">'ko'</span>]</span><br><span class="line">    <span class="comment"># 如果pathway ID已经在字典中，添加KO基因到对应的列表中</span></span><br><span class="line">    <span class="comment"># 否则，在字典中创建新的pathway ID键，并添加KO基因</span></span><br><span class="line">    <span class="keyword">if</span> pathway_id <span class="keyword">in</span> gene_sets:</span><br><span class="line">        <span class="comment"># 假设一个pathway可以有多个相同的ko基因，我们不重复添加</span></span><br><span class="line">        <span class="keyword">if</span> ko_gene <span class="keyword">not</span> <span class="keyword">in</span> gene_sets[pathway_id]:</span><br><span class="line">            gene_sets[pathway_id].append(ko_gene)</span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        <span class="comment"># 创建新键，初始化为包含当前KO基因的列表</span></span><br><span class="line">        gene_sets[pathway_id] = [ko_gene]</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">label=[<span class="string">'up'</span>,<span class="string">'down'</span>]</span><br><span class="line">data_list=[<span class="string">'CK_CD'</span>,</span><br><span class="line">            <span class="string">'CK_CE'</span>,</span><br><span class="line">            <span class="string">'CD_CE'</span>,</span><br><span class="line">            ]</span><br><span class="line">colors=[<span class="string">'#2e86de'</span>,<span class="string">'#c0392b'</span>,<span class="string">'#28b463'</span>]</span><br><span class="line"><span class="keyword">for</span> cla <span class="keyword">in</span> label:</span><br><span class="line">    <span class="keyword">for</span> i,data_name <span class="keyword">in</span> enumerate(data_list):</span><br><span class="line">        all_results=[]</span><br><span class="line">        data = pd.read_csv(<span class="string">f'/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_DEG/<span class="subst">&#123;species&#125;</span>_<span class="subst">&#123;data_name&#125;</span>_DEG_<span class="subst">&#123;cla&#125;</span>.csv'</span>)</span><br><span class="line">        <span class="comment">#ic(data)</span></span><br><span class="line">        <span class="comment">#data = data[data['sig'] == 'sig']</span></span><br><span class="line">        <span class="comment">#ic(data)</span></span><br><span class="line">        gene_list = data[<span class="string">'FEATURE_ID'</span>]<span class="comment">#[:5000]##['GeneID']</span></span><br><span class="line">        gene_list = [str(i) <span class="keyword">for</span> i <span class="keyword">in</span> gene_list]</span><br><span class="line">        <span class="comment">#gene_list = pd.read_csv('/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/data/top_diff_CD_CE_degree.csv')</span></span><br><span class="line">        <span class="comment">#gene_list = list(gene_list)</span></span><br><span class="line">        <span class="comment">#gene_list = gene_list['0'].apply(convert_msu_to_rap)</span></span><br><span class="line">        <span class="comment"># ic(gene_list)</span></span><br><span class="line">        <span class="comment"># ic(gene_sets)</span></span><br><span class="line">        <span class="comment"># GO富集</span></span><br><span class="line">        enr = gp.enrichr(gene_list=gene_list,<span class="comment">#所需查询gene_list，可以是一个列表，也可为文件（一列，每行一个基因）</span></span><br><span class="line">                        gene_sets= gene_sets,<span class="comment">#['GO_Biological_Process_2018'],#,#gene set library，多个相关的gene set 。如所有GO term组成一个gene set library.</span></span><br><span class="line">                        <span class="comment">#organism='rice',#持(human, mouse, yeast, fly, fish, worm)， 自定义gene_set 则无影响。</span></span><br><span class="line">                        <span class="comment">#description='kegg',#工作运行描述</span></span><br><span class="line">                        outdir=<span class="string">f'/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_enrichr/KEGG/<span class="subst">&#123;species&#125;</span>/<span class="subst">&#123;species&#125;</span>_<span class="subst">&#123;data_name&#125;</span>_<span class="subst">&#123;cla&#125;</span>'</span>,<span class="comment">#输出目录</span></span><br><span class="line">                        top_term=<span class="number">20</span>,</span><br><span class="line">                        cutoff=<span class="number">0.9</span>,<span class="comment">#pvalue阈值</span></span><br><span class="line">                        )</span><br><span class="line">        ic(enr.results)</span><br><span class="line">        <span class="comment">#all_results.append(enr.results)</span></span><br><span class="line">        dot_png = <span class="string">f"<span class="subst">&#123;species&#125;</span>_<span class="subst">&#123;data_name&#125;</span>_<span class="subst">&#123;cla&#125;</span>"</span> + <span class="string">"_"</span> + <span class="string">"dot"</span> + <span class="string">".png"</span></span><br><span class="line">        bar_png = <span class="string">f"<span class="subst">&#123;species&#125;</span>_<span class="subst">&#123;data_name&#125;</span>_<span class="subst">&#123;cla&#125;</span>"</span> + <span class="string">"_"</span> + <span class="string">"bar"</span> + <span class="string">".png"</span></span><br><span class="line">        base_path = <span class="string">f"/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_enrichr/KEGG/<span class="subst">&#123;species&#125;</span>/funcKEGG"</span></span><br><span class="line"></span><br><span class="line">        <span class="keyword">if</span> os.path.exists(base_path+dot_png):</span><br><span class="line">            os.remove(base_path+dot_png)</span><br><span class="line"></span><br><span class="line">        <span class="comment">#fig, axs = plt.subplots(nrows=3, ncols=1, figsize=(5, 10))</span></span><br><span class="line">        cut=<span class="number">0.6</span></span><br><span class="line">        df= dotplot(enr.results,</span><br><span class="line">                column=<span class="string">"Adjusted P-value"</span>,</span><br><span class="line">                size=<span class="number">5</span>,</span><br><span class="line">                <span class="comment">#figsize=(3,5),</span></span><br><span class="line">                xticklabels_rot=<span class="number">45</span>,</span><br><span class="line">                title=<span class="string">f'GO Top10 Pathway <span class="subst">&#123;species&#125;</span> <span class="subst">&#123;data_name&#125;</span> <span class="subst">&#123;cla&#125;</span>'</span>, </span><br><span class="line">                show_ring=<span class="keyword">True</span>,</span><br><span class="line">                <span class="comment">#marker='o',</span></span><br><span class="line">                cmap=<span class="string">'RdBu'</span>,</span><br><span class="line">                top_term=<span class="number">10</span>, legend=<span class="string">"r"</span>,pval=<span class="number">0.9</span>,</span><br><span class="line">                ofname=base_path+dot_png,</span><br><span class="line">                cutoff=cut,</span><br><span class="line">                )</span><br><span class="line">        <span class="keyword">if</span> os.path.exists(base_path+bar_png):</span><br><span class="line">            os.remove(base_path+bar_png)</span><br><span class="line"></span><br><span class="line">        ax = barplot(enr.results,</span><br><span class="line">                    column=<span class="string">"Adjusted P-value"</span>,</span><br><span class="line">                    title=<span class="string">f"<span class="subst">&#123;species&#125;</span> <span class="subst">&#123;data_name&#125;</span> <span class="subst">&#123;cla&#125;</span>"</span>,</span><br><span class="line">                    <span class="comment">#group='Gene_set', # set group, so you could do a multi-sample/library comparsion</span></span><br><span class="line">                    size=<span class="number">1</span>,</span><br><span class="line">                    top_term=<span class="number">10</span>,</span><br><span class="line">                    <span class="comment">#figsize=(3,5),</span></span><br><span class="line">                    color=colors[i],<span class="comment">#['#27ae60', '#34495e'] , # set colors for group</span></span><br><span class="line">                    ofname=base_path+bar_png,</span><br><span class="line">                    cutoff=cut</span><br><span class="line">                    <span class="comment">#color = &#123;'KEGG_2021_Human': 'salmon', 'MSigDB_Hallmark_2020':'darkblue'&#125;</span></span><br><span class="line">                    )</span><br><span class="line">        df = df.iloc[::<span class="number">-1</span>].reset_index(drop=<span class="keyword">True</span>)</span><br><span class="line">        df.to_csv(<span class="string">f'/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_enrichr/KEGG/<span class="subst">&#123;species&#125;</span>/<span class="subst">&#123;species&#125;</span>_<span class="subst">&#123;data_name&#125;</span>_<span class="subst">&#123;cla&#125;</span>.csv'</span>,index=<span class="keyword">False</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">    <span class="comment"># from gseapy import enrichment_map</span></span><br><span class="line">    <span class="comment"># import numpy as np</span></span><br><span class="line">    <span class="comment"># # return two dataframe</span></span><br><span class="line">    <span class="comment"># ic(enr.results)</span></span><br><span class="line">    <span class="comment"># nodes, edges = enrichment_map(enr.results,cutoff=cut,top_term=20)</span></span><br><span class="line">    <span class="comment"># ic(edges)</span></span><br><span class="line">    <span class="comment"># import networkx as nx</span></span><br><span class="line">    <span class="comment"># # build graph</span></span><br><span class="line">    <span class="comment"># G = nx.from_pandas_edgelist(edges,</span></span><br><span class="line">    <span class="comment">#                             source='src_idx',</span></span><br><span class="line">    <span class="comment">#                             target='targ_idx',</span></span><br><span class="line">    <span class="comment">#                             edge_attr=['jaccard_coef', 'overlap_coef', 'overlap_genes'])</span></span><br><span class="line"></span><br><span class="line">    <span class="comment"># fig, ax = plt.subplots()</span></span><br><span class="line"></span><br><span class="line">    <span class="comment"># # init node cooridnates</span></span><br><span class="line">    <span class="comment"># pos=nx.layout.spiral_layout(G)</span></span><br><span class="line">    <span class="comment"># ic(pos.keys())</span></span><br><span class="line">    <span class="comment"># #node_size = nx.get_node_attributes()</span></span><br><span class="line">    <span class="comment"># # draw node</span></span><br><span class="line">    <span class="comment"># nodelist = list(G)</span></span><br><span class="line">    <span class="comment"># xy = np.asarray([pos[v] for v in nodelist])</span></span><br><span class="line">    <span class="comment"># ic(nodelist)</span></span><br><span class="line">    <span class="comment"># original_list = list(xy[:, 0])</span></span><br><span class="line">    <span class="comment"># nx.draw_networkx_nodes(G,</span></span><br><span class="line">    <span class="comment">#                     pos=pos,</span></span><br><span class="line">    <span class="comment">#                     cmap=plt.cm.RdYlBu,</span></span><br><span class="line">    <span class="comment">#                     node_color= [number * 100 for number in original_list],</span></span><br><span class="line">    <span class="comment">#                     node_size= [number * 1000 for number in original_list])</span></span><br><span class="line">    <span class="comment"># # draw node label</span></span><br><span class="line">    <span class="comment"># ic(nodes.Term)</span></span><br><span class="line">    <span class="comment"># ic(pos)</span></span><br><span class="line">    <span class="comment"># node_term_dict = nodes.Term.to_dict()</span></span><br><span class="line">    <span class="comment"># labels = &#123;v:node_term_dict[v] for v in nodelist&#125;</span></span><br><span class="line">    <span class="comment"># nx.draw_networkx_labels(G,</span></span><br><span class="line">    <span class="comment">#                         pos=pos,</span></span><br><span class="line">    <span class="comment">#                         labels=labels)</span></span><br><span class="line">    <span class="comment"># # draw edge</span></span><br><span class="line">    <span class="comment"># edge_weight = nx.get_edge_attributes(G, 'jaccard_coef').values()</span></span><br><span class="line">    <span class="comment"># nx.draw_networkx_edges(G,</span></span><br><span class="line">    <span class="comment">#                     pos=pos,</span></span><br><span class="line">    <span class="comment">#                     width=list(map(lambda x: x*10, edge_weight)),</span></span><br><span class="line">    <span class="comment">#                     edge_color='#CDDBD4')</span></span><br><span class="line">    <span class="comment"># #plt.show()</span></span><br><span class="line">    <span class="comment"># plt.savefig(f'/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_enrichr/KEGG/&#123;data_name&#125;_kegg_grn.png')</span></span><br></pre></td></tr></table></figure>
<p><strong>WGCNA分析R</strong></p>
<figure class="highlight r"><table><tr><td class="code"><pre><span class="line"><span class="comment">#Rscript script_name.R arg1 arg2</span></span><br><span class="line"><span class="comment"># install.packages("devtools", repos='https://cran.r-project.org')</span></span><br><span class="line"><span class="comment"># devtools::install_version("ggplot2", version="2.2.1")</span></span><br><span class="line"><span class="comment"># install.packages("optparse", repos='https://cran.r-project.org')</span></span><br><span class="line"><span class="comment"># install.packages("jsonlite", repos='https://cran.r-project.org')</span></span><br><span class="line"><span class="comment"># install.packages('amap', repos='https://cran.r-project.org')</span></span><br><span class="line"><span class="comment"># install.packages("dendextend", repos='https://cran.r-project.org')</span></span><br><span class="line"><span class="comment"># install.packages("WGCNA", repos='https://cran.r-project.org')</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">library</span>(optparse)</span><br><span class="line">option_list &lt;- list(</span><br><span class="line">  make_option(c(<span class="string">"-i"</span>, <span class="string">"--infile"</span>), type = <span class="string">"character"</span>, default = <span class="string">'/pub1/data/mg_projects/projects/web_script/tool_runing/8ed926747e04570e4a5345bfa133185b/input.json'</span>,</span><br><span class="line">              action = <span class="string">"store"</span>, help = <span class="string">"Input a exp file path!"</span></span><br><span class="line">  ),</span><br><span class="line">  make_option(c(<span class="string">"-o"</span>, <span class="string">"--outfile"</span>), type = <span class="string">"character"</span>, default = <span class="string">'/pub1/data/mg_projects/projects/web_script/web_file_catche/runing/dae42f7681eca28355c6ba9105a3b5ff'</span>,</span><br><span class="line">              action = <span class="string">"store"</span>, help = <span class="string">"Input a outfolder path!"</span></span><br><span class="line">  )</span><br><span class="line">)</span><br><span class="line">logs=c()</span><br><span class="line"></span><br><span class="line"><span class="keyword">tryCatch</span>(&#123;</span><br><span class="line">  Args &lt;- commandArgs()</span><br><span class="line">  opt = parse_args(OptionParser(option_list = option_list, usage = <span class="string">"Data press"</span>))</span><br><span class="line">  <span class="comment">#logs=c(logs,paste0('geting data:',paste0(paste0(names(opt),'=',opt),collapse = ',')))</span></span><br><span class="line">  logs=c(logs,paste0(<span class="string">'run wgcna.R-'</span>,basename(opt$outfile)))</span><br><span class="line"></span><br><span class="line">  <span class="keyword">library</span>(jsonlite)</span><br><span class="line">  data&lt;-jsonlite::stream_in(file(opt$infile),pagesize = <span class="number">1000</span>)</span><br><span class="line">  </span><br><span class="line"><span class="comment">#exp_path='/pub1/data/user_data/13456826965/GSE1561/GSE1561_GPL96_sample_exp.txt'</span></span><br><span class="line"><span class="comment">#geneMode='mad'</span></span><br><span class="line"><span class="comment">#geneModeValue='50'</span></span><br><span class="line"><span class="comment">#sampleCut='1'</span></span><br><span class="line"><span class="comment">#netType='unsigned'</span></span><br><span class="line"><span class="comment">#geneModeValue=as.numeric(geneModeValue)</span></span><br><span class="line"><span class="comment">#outFolder='/pub1/data/mg_projects/projects/web_script/tool_runing/test_data/WGCNA1'</span></span><br><span class="line"><span class="comment">#minModuleSize=30</span></span><br><span class="line"><span class="comment">#deepSplit=4</span></span><br><span class="line"><span class="comment">#mergeCutHeight=0.5</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#oldFolder='/pub1/data/mg_projects/projects/web_script/tool_runing/test_data/WGCNA'</span></span><br><span class="line">method=unlist(data$method)<span class="comment">#'cli'#module</span></span><br><span class="line"><span class="comment">#cliDatas=list()</span></span><br><span class="line"><span class="comment">#list(A=list(name:xxx,sample:c(),value:c()),B=list())</span></span><br><span class="line">oldFolder=unlist(data$oldFolder)</span><br><span class="line"></span><br><span class="line"><span class="keyword">library</span>(<span class="string">'amap'</span>)</span><br><span class="line"><span class="keyword">library</span>(dendextend)</span><br><span class="line">getTreeNode=<span class="keyword">function</span>(hc_obj)&#123;</span><br><span class="line">  dend=as.dendrogram(hc_obj)</span><br><span class="line">  capture.output(str(dend))-&gt;stx2</span><br><span class="line">  <span class="comment">#write.table(stx,row.names = F,col.names = F ,paste0(outFolder,'/wgcna_dissTOM_tree.text'))</span></span><br><span class="line">  gsub(<span class="string">'^ \\s+'</span>,<span class="string">''</span>,stx2)-&gt;stx2</span><br><span class="line">  nodeMap=rbind()</span><br><span class="line">  leafs=rbind()</span><br><span class="line">  all_node_obj=rbind()</span><br><span class="line">  <span class="keyword">for</span>(i <span class="keyword">in</span> length(stx2):<span class="number">1</span>)&#123;</span><br><span class="line">    <span class="comment">#  i=length(stx2)</span></span><br><span class="line">    <span class="comment">#i=547</span></span><br><span class="line">    <span class="keyword">if</span>(length(grep(<span class="string">'--leaf'</span>,stx2[i]))==<span class="number">0</span>)&#123;</span><br><span class="line">      h=unlist(stringr::str_split(stx2[i],<span class="string">'members at h = '</span>))[<span class="number">2</span>]</span><br><span class="line">      h=gsub(<span class="string">']'</span>,<span class="string">''</span>,h)</span><br><span class="line">      lefNode=unlist(stringr::str_split(stx2[i],<span class="string">' members at h = '</span>))[<span class="number">1</span>]</span><br><span class="line">      lefNode=unlist(stringr::str_split(lefNode,<span class="string">'branches and '</span>))[<span class="number">2</span>]</span><br><span class="line">      sub1=paste0(<span class="string">'C'</span>,i+<span class="number">1</span>)</span><br><span class="line">      all_node_obj[which(all_node_obj[,<span class="number">1</span>]==paste0(<span class="string">'C'</span>,i+<span class="number">1</span>)),<span class="number">4</span>]=paste0(<span class="string">'C'</span>,i)</span><br><span class="line">      <span class="comment">#sub2=all_node_obj[which(all_node_obj[,4]==''),1]</span></span><br><span class="line">      <span class="keyword">for</span>(j <span class="keyword">in</span> <span class="number">1</span>:nrow(all_node_obj))&#123;</span><br><span class="line">        <span class="comment">#t.ind=which(all_node_obj[,1]==paste0('C',i+j+1))</span></span><br><span class="line">        <span class="comment">#if(length(t.ind))</span></span><br><span class="line">        sub_mx=all_node_obj[which(all_node_obj[,<span class="number">1</span>]==paste0(<span class="string">'C'</span>,i+j+<span class="number">1</span>)),]</span><br><span class="line">        <span class="keyword">if</span>(sub_mx[<span class="number">4</span>]==<span class="string">''</span>)&#123;</span><br><span class="line">          sub2=sub_mx[<span class="number">1</span>]</span><br><span class="line">          <span class="keyword">break</span>()</span><br><span class="line">        &#125;</span><br><span class="line">      &#125;</span><br><span class="line">      all_node_obj[which(all_node_obj[,<span class="number">1</span>]==sub2),<span class="number">4</span>]=paste0(<span class="string">'C'</span>,i)</span><br><span class="line">      <span class="comment">#all_node_obj=c(all_node_obj,list(Node=paste0('C',i),H=h,Members=lefNode,subNode1=paste0('C',i+1),subNode12=''))</span></span><br><span class="line">      all_node_obj=rbind(all_node_obj,c(paste0(<span class="string">'C'</span>,i),sub1,sub2,<span class="string">''</span>,lefNode))</span><br><span class="line">      nodeMap=rbind(nodeMap,c(<span class="string">''</span>,paste0(<span class="string">'C'</span>,i),paste0(<span class="string">'C'</span>,i+<span class="number">1</span>),sub2,h))</span><br><span class="line">    &#125;<span class="keyword">else</span>&#123;</span><br><span class="line">      node=gsub(<span class="string">' '</span>,<span class="string">''</span>,unlist(stringr::str_split(stx2[i],<span class="string">'--leaf'</span>))[<span class="number">2</span>])</span><br><span class="line">      node=gsub(<span class="string">'^ \\s+'</span>,<span class="string">''</span>,node)</span><br><span class="line">      node=gsub(<span class="string">'\\s+ $'</span>,<span class="string">''</span>,node)</span><br><span class="line">      node=gsub(<span class="string">'^"'</span>,<span class="string">''</span>,node)</span><br><span class="line">      node=gsub(<span class="string">'"$'</span>,<span class="string">''</span>,node)</span><br><span class="line">      <span class="comment">#node=colnames(adjacency)[as.numeric(node)]</span></span><br><span class="line">      leafs=rbind(leafs,c(node,paste0(<span class="string">'C'</span>,i)))</span><br><span class="line">      all_node_obj=rbind(all_node_obj,c(paste0(<span class="string">'C'</span>,i),<span class="string">''</span>,<span class="string">''</span>,<span class="string">''</span>,<span class="number">1</span>))</span><br><span class="line">      nodeMap=rbind(nodeMap,c(node,paste0(<span class="string">'C'</span>,i),<span class="string">''</span>,<span class="string">''</span>,<span class="number">0</span>))</span><br><span class="line">    &#125;</span><br><span class="line">  &#125;</span><br><span class="line">  <span class="comment">#all_node_obj[all_node_obj[,4]=='',]</span></span><br><span class="line">  <span class="comment">#sort(leafs[,1])[1:10]</span></span><br><span class="line">  od=rep(nrow(nodeMap),nrow(nodeMap))</span><br><span class="line">  od[match(leafs[,<span class="number">1</span>],nodeMap[,<span class="number">1</span>])]=nrow(leafs):<span class="number">1</span></span><br><span class="line">  nodeMap=nodeMap[order(od),]</span><br><span class="line">  <span class="keyword">return</span>(nodeMap)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">getTreeNode_backup=<span class="keyword">function</span>(hc_obj)&#123;</span><br><span class="line">  dend=as.dendrogram(hc_obj)</span><br><span class="line">  dend_label=partition_leaves(dend)</span><br><span class="line">  dend_height=get_nodes_attr(dend,<span class="string">'height'</span>)</span><br><span class="line">  nodeMap=rbind()</span><br><span class="line">  <span class="keyword">for</span>(i <span class="keyword">in</span> length(dend_height):<span class="number">1</span>)&#123;</span><br><span class="line">    cls=dend_label[[i]]</span><br><span class="line">    node1=paste0(cls,collapse = <span class="string">'\t'</span>)</span><br><span class="line">    snode1=<span class="string">''</span></span><br><span class="line">    snode2=<span class="string">''</span></span><br><span class="line">    <span class="keyword">if</span>(length(cls)&gt;<span class="number">1</span>)&#123;</span><br><span class="line">      node1=openssl::md5(node1)</span><br><span class="line">      snode1=paste0(<span class="string">'C'</span>,i+<span class="number">1</span>)</span><br><span class="line">      node2_ts=dend_label[[i+<span class="number">1</span>]]</span><br><span class="line">      cls1=cls[which(!cls%<span class="keyword">in</span>%node2_ts)]</span><br><span class="line">      node2=paste0(cls1,collapse = <span class="string">'\t'</span>)</span><br><span class="line">      <span class="keyword">if</span>(length(cls1)&gt;<span class="number">1</span>)&#123;</span><br><span class="line">        node2=openssl::md5(node2)</span><br><span class="line">      &#125;</span><br><span class="line">      snode2=nodeMap[which(nodeMap[,<span class="number">1</span>]==node2),<span class="number">2</span>]</span><br><span class="line">    &#125;</span><br><span class="line">    nodeMap=rbind(nodeMap,c(node1,paste0(<span class="string">'C'</span>,i),snode1,snode2,dend_height[i]))</span><br><span class="line">  &#125;</span><br><span class="line">  od=rep(<span class="number">10000000</span>,nrow(nodeMap))</span><br><span class="line">  od[match(hc_obj$labels[hc_obj$order],nodeMap[,<span class="number">1</span>])]=<span class="number">1</span>:length(hc_obj$labels)</span><br><span class="line">  nodeMap=nodeMap[order(od),]</span><br><span class="line">  nodeMap[which(nodeMap[,<span class="number">3</span>]!=<span class="string">''</span>),<span class="number">1</span>]=<span class="string">''</span></span><br><span class="line">  <span class="comment">#head(nodeMap)</span></span><br><span class="line">  <span class="comment">#head(nodeMap1)</span></span><br><span class="line">  <span class="comment">#sum(nodeMap1[,1]==nodeMap1[,1])</span></span><br><span class="line">  <span class="comment">#nodeMap1[which(nodeMap1[,3]!=''),1]</span></span><br><span class="line">  <span class="keyword">return</span>(nodeMap)</span><br><span class="line">&#125;</span><br><span class="line">getClusterTree=<span class="keyword">function</span>(exp,dMethod=<span class="string">'euclidean'</span>,hMethod=<span class="string">'complete'</span>)&#123;</span><br><span class="line">  cNames=colnames(exp)</span><br><span class="line">  colnames(exp)=paste0(<span class="string">'C'</span>,<span class="number">1</span>:ncol(exp))</span><br><span class="line">  hc=hcluster(t(exp),method =dMethod,link=hMethod )</span><br><span class="line">  nodeMap=getTreeNode(hc)</span><br><span class="line">  nodeMap2=getTreeNode_backup(hc)</span><br><span class="line">  <span class="comment">#dim(nodeMap1)</span></span><br><span class="line">  <span class="comment">#head(nodeMap1)</span></span><br><span class="line">  sum(nodeMap2[,<span class="number">2</span>]==nodeMap[,<span class="number">2</span>])</span><br><span class="line">  sum(nodeMap2[,<span class="number">1</span>]==nodeMap[,<span class="number">1</span>])</span><br><span class="line">  sum(nodeMap2[,<span class="number">4</span>]==nodeMap[,<span class="number">4</span>])</span><br><span class="line">  cbind(nodeMap2[which(nodeMap2[,<span class="number">4</span>]!=nodeMap[,<span class="number">4</span>]),],nodeMap[which(nodeMap2[,<span class="number">4</span>]!=nodeMap[,<span class="number">4</span>]),])</span><br><span class="line">  </span><br><span class="line">  <span class="comment">#dim(nodeMap)</span></span><br><span class="line">  nodeMap[,<span class="number">1</span>]=cNames[match(nodeMap[,<span class="number">1</span>],colnames(exp))]</span><br><span class="line">  nodeMap[is.na(nodeMap[,<span class="number">1</span>]),<span class="number">1</span>]=<span class="string">''</span></span><br><span class="line">  <span class="comment">#nodeMap[which(nodeMap[,3]!=''),1]</span></span><br><span class="line">  <span class="comment">#head(exp)</span></span><br><span class="line">  <span class="comment">#dend=as.dendrogram(hc)</span></span><br><span class="line">  <span class="comment">#dend_label=partition_leaves(dend)</span></span><br><span class="line">  <span class="comment">#dend_height=get_nodes_attr(dend,'height')</span></span><br><span class="line">  <span class="comment">#nodeMap=rbind()</span></span><br><span class="line">  <span class="comment">#for(i in length(dend_height):1)&#123;</span></span><br><span class="line">  <span class="comment">#  cls=dend_label[[i]]</span></span><br><span class="line">  <span class="comment">#  node1=paste0(cls,collapse = '\t')</span></span><br><span class="line">  <span class="comment">#  snode1=''</span></span><br><span class="line">  <span class="comment">#  snode2=''</span></span><br><span class="line">  <span class="comment">#  if(length(cls)&gt;1)&#123;</span></span><br><span class="line">  <span class="comment">#    node1=openssl::md5(node1)</span></span><br><span class="line">  <span class="comment">#    snode1=paste0('C',i+1)</span></span><br><span class="line">  <span class="comment">#    node2_ts=dend_label[[i+1]]</span></span><br><span class="line">  <span class="comment">#    cls1=cls[which(!cls%in%node2_ts)]</span></span><br><span class="line">  <span class="comment">#    node2=paste0(cls1,collapse = '\t')</span></span><br><span class="line">  <span class="comment">#    if(length(cls1)&gt;1)&#123;</span></span><br><span class="line">  <span class="comment">#      node2=openssl::md5(node2)</span></span><br><span class="line">  <span class="comment">#    &#125;</span></span><br><span class="line">  <span class="comment">#    snode2=nodeMap[which(nodeMap[,1]==node2),2]</span></span><br><span class="line">  <span class="comment">#  &#125;</span></span><br><span class="line">  <span class="comment">#  nodeMap=rbind(nodeMap,c(node1,paste0('C',i),snode1,snode2,dend_height[i]))</span></span><br><span class="line">  <span class="comment">#&#125;</span></span><br><span class="line">  <span class="comment">#od=rep(10000000,nrow(nodeMap))</span></span><br><span class="line">  <span class="comment">#od[match(hc$labels[hc$order],nodeMap[,1])]=1:length(hc$labels)</span></span><br><span class="line">  <span class="comment">#nodeMap=nodeMap[order(od),]</span></span><br><span class="line">  <span class="comment">#nodeMap[which(nodeMap[,3]!=''),1]=''</span></span><br><span class="line">  <span class="comment">#head(nodeMap)</span></span><br><span class="line">  <span class="comment">#head(nodeMap1)</span></span><br><span class="line">  <span class="comment">#sum(nodeMap1[,1]==nodeMap1[,1])</span></span><br><span class="line">  <span class="comment">#nodeMap1[which(nodeMap1[,3]!=''),1]</span></span><br><span class="line">  <span class="comment">#head(nodeMap)</span></span><br><span class="line">  <span class="keyword">return</span>(nodeMap)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">library</span>(WGCNA)</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span>(method==<span class="string">'module'</span>)&#123;</span><br><span class="line"></span><br><span class="line">  exp_path=unlist(data$exp_path)</span><br><span class="line">  geneMode=paste0(as.character(unlist(data$geneMode)),<span class="string">''</span>)</span><br><span class="line">  geneModeValue=unlist(data$geneModeValue)</span><br><span class="line">  sampleCut=unlist(data$sampleCut)</span><br><span class="line">  netType=unlist(data$netType)</span><br><span class="line">  outFolder=opt$outfile</span><br><span class="line">  minModuleSize=unlist(data$minModuleSize)</span><br><span class="line">  deepSplit=unlist(data$deepSplit)</span><br><span class="line">  oldFolder=unlist(data$oldFolder)</span><br><span class="line">  mergeCutHeight=unlist(data$mergeCutHeight)</span><br><span class="line">  </span><br><span class="line">  geneModeValue=as.numeric(geneModeValue)</span><br><span class="line">  </span><br><span class="line"><span class="keyword">if</span>(oldFolder!=<span class="string">''</span>&amp;oldFolder!=outFolder&amp;file.exists(paste0(oldFolder,<span class="string">'/wgcna_TOM.RData'</span>)))&#123;</span><br><span class="line">  load(paste0(oldFolder,<span class="string">'/wgcna_TOM.RData'</span>))</span><br><span class="line">  file.copy(paste0(oldFolder,<span class="string">'/wgcna_pre_tree.txt'</span>),paste0(outFolder,<span class="string">'/wgcna_pre_tree.txt'</span>),overwrite = <span class="literal">T</span>)</span><br><span class="line">  file.copy(paste0(oldFolder,<span class="string">'/wgcna_exp_powers.txt'</span>),paste0(outFolder,<span class="string">'/wgcna_exp_powers.txt'</span>),overwrite = <span class="literal">T</span>)</span><br><span class="line">  file.copy(paste0(oldFolder,<span class="string">'/wgcna_exp_data.txt'</span>),paste0(outFolder,<span class="string">'/wgcna_exp_data.txt'</span>),overwrite = <span class="literal">T</span>)</span><br><span class="line">  file.copy(paste0(oldFolder,<span class="string">'/wgcna_dissTOM_tree.txt'</span>),paste0(outFolder,<span class="string">'/wgcna_dissTOM_tree.txt'</span>),overwrite = <span class="literal">T</span>)</span><br><span class="line">  file.copy(paste0(oldFolder,<span class="string">'/wgcna_module_net.fst'</span>),paste0(outFolder,<span class="string">'/wgcna_module_net.fst'</span>),overwrite = <span class="literal">T</span>)</span><br><span class="line">  file.copy(paste0(oldFolder,<span class="string">'/wgcna_TOM.RData'</span>),paste0(outFolder,<span class="string">'/wgcna_TOM.RData'</span>),overwrite = <span class="literal">T</span>)</span><br><span class="line">  logs=c(logs,paste0(<span class="string">'SampleList='</span>,paste0(row.names(filter.exp),collapse = <span class="string">'#,#'</span>)))</span><br><span class="line">  logs=c(logs,paste0(<span class="string">'saved module network data:'</span>,paste0(cytq,collapse = <span class="string">','</span>)))</span><br><span class="line">  logs=c(logs,paste0(<span class="string">'set power='</span>,cutPower,<span class="string">',By sample count='</span>,nrow(filter.exp)))</span><br><span class="line">  logs=c(logs,paste0(<span class="string">'saved TOM data'</span>))</span><br><span class="line">  <span class="comment">#wgcna_module_net.fst</span></span><br><span class="line">  </span><br><span class="line">&#125;<span class="keyword">else</span>&#123;</span><br><span class="line"></span><br><span class="line">  </span><br><span class="line">dat=data.table::fread(exp_path, sep = <span class="string">"\t"</span>,header = <span class="literal">T</span>,stringsAsFactors = <span class="literal">F</span>,check.names = <span class="literal">F</span></span><br><span class="line">                      ,na.strings=<span class="string">"NA"</span>,data.table = <span class="literal">F</span>)  </span><br><span class="line">dat=dat[match(unique(dat[,<span class="number">1</span>]),dat[,<span class="number">1</span>]),]</span><br><span class="line">row.names(dat)=dat[,<span class="number">1</span>]</span><br><span class="line">dat=dat[,-<span class="number">1</span>]</span><br><span class="line">logs=c(logs,<span class="string">'filter SD=0'</span>)</span><br><span class="line">sds=apply(dat, <span class="number">1</span>, sd)</span><br><span class="line">dat=dat[which(sds&gt;<span class="number">0</span>),]</span><br><span class="line">logs=c(logs,paste0(<span class="string">'filtered SD=0,row='</span>,nrow(dat),<span class="string">',col='</span>,ncol(dat)))</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span>(geneMode==<span class="string">'mad'</span>)&#123;</span><br><span class="line">  logs=c(logs,paste0(<span class="string">'filter MAD&gt;'</span>,ceiling(geneModeValue),<span class="string">'%'</span>))</span><br><span class="line">  mads=apply(dat, <span class="number">1</span>, mad)</span><br><span class="line">  t_inds=which(mads&gt;=quantile(mads,seq(<span class="number">0</span>,<span class="number">1</span>,<span class="number">0.01</span>))[ceiling(geneModeValue)+<span class="number">1</span>])</span><br><span class="line">  dat=dat[t_inds,]</span><br><span class="line">  logs=c(logs,paste0(<span class="string">'filtered MAD,row='</span>,nrow(dat),<span class="string">',col='</span>,ncol(dat)))</span><br><span class="line">&#125;<span class="keyword">else</span> <span class="keyword">if</span>(geneMode==<span class="string">'sd'</span>)&#123;</span><br><span class="line">  logs=c(logs,paste0(<span class="string">'filter SD&gt;'</span>,ceiling(geneModeValue),<span class="string">'%'</span>))</span><br><span class="line">  mads=apply(dat, <span class="number">1</span>, sd)</span><br><span class="line">  t_inds=which(mads&gt;=quantile(mads,seq(<span class="number">0</span>,<span class="number">1</span>,<span class="number">0.01</span>))[ceiling(geneModeValue)+<span class="number">1</span>])</span><br><span class="line">  dat=dat[t_inds,]</span><br><span class="line">  logs=c(logs,paste0(<span class="string">'filtered SD,row='</span>,nrow(dat),<span class="string">',col='</span>,ncol(dat)))</span><br><span class="line">&#125;<span class="keyword">else</span> <span class="keyword">if</span>(geneMode==<span class="string">'mean'</span>)&#123;</span><br><span class="line">  logs=c(logs,paste0(<span class="string">'filter mean&gt;'</span>,(geneModeValue),<span class="string">'%'</span>))</span><br><span class="line">  mads=apply(dat, <span class="number">1</span>, mean)</span><br><span class="line">  t_inds=which(mads&gt;geneModeValue)</span><br><span class="line">  dat=dat[t_inds,]</span><br><span class="line">  logs=c(logs,paste0(<span class="string">'filtered mean,row='</span>,nrow(dat),<span class="string">',col='</span>,ncol(dat)))</span><br><span class="line">&#125;</span><br><span class="line"><span class="comment">#logs=c()</span></span><br><span class="line"><span class="comment">#print('=====')</span></span><br><span class="line">filter.exp=t(dat)</span><br><span class="line">dim(filter.exp)</span><br><span class="line">logs=c(logs,paste0(<span class="string">'Sample clustering'</span>))</span><br><span class="line">sTree=getClusterTree(dat,hMethod = <span class="string">'average'</span>)</span><br><span class="line">write.table(sTree,file = paste0(outFolder,<span class="string">'/wgcna_pre_tree.txt'</span>),quote = <span class="literal">F</span>,row.names = <span class="literal">F</span>,sep = <span class="string">'\t'</span>)</span><br><span class="line">logs=c(logs,paste0(<span class="string">'saved sample clustered'</span>))</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">gsg = goodSamplesGenes(filter.exp, verbose = <span class="number">3</span>);</span><br><span class="line"><span class="comment">#gsg$allOK</span></span><br><span class="line"><span class="comment"># if FALSE:</span></span><br><span class="line"><span class="keyword">if</span> (!gsg$allOK)</span><br><span class="line">&#123;</span><br><span class="line">  <span class="comment"># Optionally, print the gene and sample names that were removed:</span></span><br><span class="line">  <span class="keyword">if</span> (sum(!gsg$goodGenes)&gt;<span class="number">0</span>)</span><br><span class="line">    logs(logs,paste(<span class="string">"Removing genes:"</span>, paste(colnames(filter.exp)[!gsg$goodGenes], collapse = <span class="string">", "</span>)));</span><br><span class="line">  <span class="keyword">if</span> (sum(!gsg$goodSamples)&gt;<span class="number">0</span>)</span><br><span class="line">    logs(logs,paste(<span class="string">"Removing samples:"</span>, paste(rownames(filter.exp)[!gsg$goodSamples], collapse = <span class="string">", "</span>)));</span><br><span class="line">  <span class="comment"># Remove the offending genes and samples from the data:</span></span><br><span class="line">  filter.exp = filter.exp[gsg$goodSamples, gsg$goodGenes]</span><br><span class="line">&#125;</span><br><span class="line">logs=c(logs,paste0(<span class="string">'SampleFiltered sampleNum='</span>,nrow(filter.exp),<span class="string">'geneNum='</span>,ncol(filter.exp)))</span><br><span class="line"></span><br><span class="line"><span class="comment">#if(sampleCut=='1')&#123;</span></span><br><span class="line"><span class="comment">#  logs=c(logs,paste0('SampleFiltering'))</span></span><br><span class="line"><span class="comment">#  logs=c(logs,paste0('Sample clustering'))</span></span><br><span class="line"><span class="comment">#  sampleTree = hclust(dist(filter.exp), method = "average")</span></span><br><span class="line">  </span><br><span class="line"><span class="comment">#  logs=c(logs,paste0('Sample clustered'))</span></span><br><span class="line"><span class="comment">#  q1=quantile(sampleTree$height, c(0,0.25, 0.5, 0.75,1), na.rm=T)</span></span><br><span class="line"><span class="comment">#  otl=1.5*(q1[4]-q1[2])</span></span><br><span class="line"><span class="comment">#  otl.u=ifelse(q1[4]+otl&gt;q1[5],q1[5],q1[4]+otl)</span></span><br><span class="line"><span class="comment">#  otl.d=ifelse(q1[2]-otl&lt;q1[1],q1[1],q1[2]-otl)</span></span><br><span class="line"><span class="comment">#  logs=c(logs,paste0('SampleFiltering,outlier upper=',otl.u))</span></span><br><span class="line">  <span class="comment">#outlier=which(sampleTree$height&gt;=otl.d&amp;sampleTree$height&lt;=otl.u)</span></span><br><span class="line"><span class="comment">#  ctr=cutree(sampleTree,h = otl.u+1)</span></span><br><span class="line"><span class="comment">#  ctr.tb=table(ctr)</span></span><br><span class="line"><span class="comment">#  outlier=match(names(which(ctr==names(ctr.tb)[which.max(ctr.tb)])),row.names(filter.exp))</span></span><br><span class="line"><span class="comment">#  logs=c(logs,paste0('SampleFiltered sampleNum=',length(outlier)))</span></span><br><span class="line"><span class="comment">#  filter.exp=filter.exp[outlier,]</span></span><br><span class="line"><span class="comment">#&#125;</span></span><br><span class="line"></span><br><span class="line">write.table(cbind(Tag=colnames(filter.exp),t(filter.exp)),file = paste0(outFolder,<span class="string">'/wgcna_exp_data.txt'</span>),quote = <span class="literal">F</span>,row.names = <span class="literal">F</span>,sep = <span class="string">'\t'</span>)</span><br><span class="line">logs=c(logs,paste0(<span class="string">'SampleList='</span>,paste0(row.names(filter.exp),collapse = <span class="string">'#,#'</span>)))</span><br><span class="line"></span><br><span class="line"><span class="comment">#pdf(file = paste0(MG_GENE_FOLDER,'/SampleCluster.pdf'),width = 12,height = 6)</span></span><br><span class="line"><span class="comment">#plot(sampleTree, main = "Sample clustering to detect outliers", sub="", xlab="")</span></span><br><span class="line"><span class="comment">#dev.off()</span></span><br><span class="line"><span class="comment">#height=140000</span></span><br><span class="line"><span class="comment">#print(height)</span></span><br><span class="line"><span class="comment">#print(min(sampleTree$height))</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#print('===1')</span></span><br><span class="line">logs=c(logs,paste0(<span class="string">'run pickSoftThresholding'</span>))</span><br><span class="line">powers = c(c(<span class="number">1</span>:<span class="number">10</span>), seq(from = <span class="number">12</span>, to=<span class="number">30</span>, by=<span class="number">2</span>))</span><br><span class="line"><span class="comment">#print('===2')</span></span><br><span class="line"><span class="comment">#print(head(filter.exp))</span></span><br><span class="line">sft = pickSoftThreshold((filter.exp), powerVector=powers, </span><br><span class="line">                        networkType=netType, verbose=<span class="number">5</span>,blockSize = <span class="number">20000</span>)</span><br><span class="line">cutPower=sft$powerEstimate</span><br><span class="line">logs=c(logs,paste0(<span class="string">'run pickSoftThresholded power='</span>,cutPower))</span><br><span class="line"><span class="keyword">if</span>(is.na(cutPower))&#123;</span><br><span class="line">  <span class="keyword">if</span>(netType==<span class="string">'unsigned'</span>|netType==<span class="string">'signed_hybrid'</span>)&#123;</span><br><span class="line">    <span class="keyword">if</span>(nrow(filter.exp)&lt;<span class="number">20</span>)&#123;</span><br><span class="line">      cutPower=<span class="number">9</span></span><br><span class="line">    &#125;<span class="keyword">else</span> <span class="keyword">if</span>(nrow(filter.exp)&lt;<span class="number">30</span>)&#123;</span><br><span class="line">      cutPower=<span class="number">8</span></span><br><span class="line">    &#125;<span class="keyword">else</span> <span class="keyword">if</span>(nrow(filter.exp)&lt;<span class="number">40</span>)&#123;</span><br><span class="line">      cutPower=<span class="number">7</span></span><br><span class="line">    &#125;<span class="keyword">else</span>&#123;</span><br><span class="line">      cutPower=<span class="number">6</span></span><br><span class="line">    &#125;</span><br><span class="line">  &#125;<span class="keyword">else</span>&#123;</span><br><span class="line">    <span class="keyword">if</span>(nrow(filter.exp)&lt;<span class="number">20</span>)&#123;</span><br><span class="line">      cutPower=<span class="number">18</span></span><br><span class="line">    &#125;<span class="keyword">else</span> <span class="keyword">if</span>(nrow(filter.exp)&lt;<span class="number">30</span>)&#123;</span><br><span class="line">      cutPower=<span class="number">16</span></span><br><span class="line">    &#125;<span class="keyword">else</span> <span class="keyword">if</span>(nrow(filter.exp)&lt;<span class="number">40</span>)&#123;</span><br><span class="line">      cutPower=<span class="number">14</span></span><br><span class="line">    &#125;<span class="keyword">else</span>&#123;</span><br><span class="line">      cutPower=<span class="number">12</span></span><br><span class="line">    &#125;</span><br><span class="line">  &#125;</span><br><span class="line">&#125;</span><br><span class="line">logs=c(logs,paste0(<span class="string">'set power='</span>,cutPower,<span class="string">',By '</span>,netType,<span class="string">' sample count='</span>,nrow(filter.exp)))</span><br><span class="line"></span><br><span class="line">sft.tab=sft$fitIndices</span><br><span class="line">sft.tab$SFTR2=-sign(sft$fitIndices[,<span class="number">3</span>])*sft$fitIndices[,<span class="number">2</span>]</span><br><span class="line">write.table(sft.tab,file = paste0(outFolder,<span class="string">'/wgcna_exp_powers.txt'</span>),quote = <span class="literal">F</span>,row.names = <span class="literal">F</span>,sep = <span class="string">'\t'</span>)</span><br><span class="line">logs=c(logs,paste0(<span class="string">'saved powerEstimateed'</span>))</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">logs=c(logs,paste0(<span class="string">'runing adjacency'</span>))</span><br><span class="line">adjacency = adjacency(filter.exp, power = cutPower,type = netType)</span><br><span class="line">logs=c(logs,paste0(<span class="string">'runing TOMsimilarity'</span>))</span><br><span class="line">TOM = TOMsimilarity(adjacency)</span><br><span class="line">colnames(TOM)=colnames(adjacency)</span><br><span class="line">row.names(TOM)=row.names(adjacency)</span><br><span class="line"><span class="comment">#head(TOM[,1:10])</span></span><br><span class="line">dissTOM = <span class="number">1</span>-TOM</span><br><span class="line">plotTOM = dissTOM^<span class="number">7</span></span><br><span class="line"><span class="comment">#dim(TOM)</span></span><br><span class="line">diag(plotTOM) = <span class="literal">NA</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#dim(adjacency)</span></span><br><span class="line"><span class="comment">#colnames(plotTOM)</span></span><br><span class="line"><span class="comment"># Call the plot function</span></span><br><span class="line"><span class="comment">#colnames(dissTOM)=colnames(adjacency)</span></span><br><span class="line"><span class="comment">#row.names(dissTOM)=colnames(adjacency)</span></span><br><span class="line">cName_TOM=colnames(dissTOM)</span><br><span class="line">rName_TOM=row.names(dissTOM)</span><br><span class="line">colnames(dissTOM)=paste0(<span class="string">'C'</span>,<span class="number">1</span>:ncol(dissTOM))</span><br><span class="line">row.names(dissTOM)=paste0(<span class="string">'R'</span>,<span class="number">1</span>:nrow(dissTOM))</span><br><span class="line">logs=c(logs,paste0(<span class="string">'hclust dissTOM'</span>))</span><br><span class="line">geneTree = hclust(as.dist(dissTOM), method = <span class="string">"average"</span>)</span><br><span class="line">dim(dissTOM)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment">#plot(geneTree)</span></span><br><span class="line">logs=c(logs,paste0(<span class="string">'hclusted dissTOM'</span>))</span><br><span class="line">nodeMap=getTreeNode(geneTree)</span><br><span class="line"><span class="comment">#head(nodeMap)</span></span><br><span class="line">nodeMap[,<span class="number">1</span>]=rName_TOM[match(nodeMap[,<span class="number">1</span>],row.names(dissTOM))]</span><br><span class="line">nodeMap[is.na(nodeMap[,<span class="number">1</span>]),<span class="number">1</span>]=<span class="string">''</span></span><br><span class="line">colnames(dissTOM)=cName_TOM</span><br><span class="line">row.names(dissTOM)=rName_TOM</span><br><span class="line"><span class="comment">#dend=as.dendrogram(geneTree)</span></span><br><span class="line"><span class="comment">#capture.output(str(dend))-&gt;stx2</span></span><br><span class="line"><span class="comment">#write.table(stx,row.names = F,col.names = F ,paste0(outFolder,'/wgcna_dissTOM_tree.text'))</span></span><br><span class="line"><span class="comment">#gsub('^ \\s+','',stx2)-&gt;stx2</span></span><br><span class="line"><span class="comment">#nodeMap=rbind()</span></span><br><span class="line"><span class="comment">#leafs=rbind()</span></span><br><span class="line"><span class="comment">#for(i in length(stx2):1)&#123;</span></span><br><span class="line"><span class="comment">#  i=length(stx2)</span></span><br><span class="line"><span class="comment">#  if(length(grep('--leaf',stx2[i]))==0)&#123;</span></span><br><span class="line"><span class="comment">#    h=unlist(stringr::str_split(stx2[i],'members at h = '))[2]</span></span><br><span class="line"><span class="comment">#    h=gsub(']','',h)</span></span><br><span class="line"><span class="comment">#    nodeMap=rbind(nodeMap,c('',paste0('C',i),paste0('C',i+1),paste0('C',i+2),h))</span></span><br><span class="line"><span class="comment">#  &#125;else&#123;</span></span><br><span class="line"><span class="comment">#    node=gsub(' ','',unlist(stringr::str_split(stx2[i],'--leaf'))[2])</span></span><br><span class="line"><span class="comment">#    node=gsub('^ \\s+','',node)</span></span><br><span class="line"><span class="comment">#    node=gsub('\\s+ $','',node)</span></span><br><span class="line"><span class="comment">#    node=gsub('^"','',node)</span></span><br><span class="line"><span class="comment">#    node=gsub('"$','',node)</span></span><br><span class="line">    <span class="comment">#node=colnames(adjacency)[as.numeric(node)]</span></span><br><span class="line"><span class="comment">#    leafs=rbind(leafs,c(node,paste0('C',i)))</span></span><br><span class="line"><span class="comment">#    nodeMap=rbind(nodeMap,c(node,paste0('C',i),'','',0))</span></span><br><span class="line"><span class="comment">#  &#125;</span></span><br><span class="line"><span class="comment">#&#125;</span></span><br><span class="line"><span class="comment">#sort(leafs[,1])[1:10]</span></span><br><span class="line"><span class="comment">#od=rep(nrow(nodeMap),nrow(nodeMap))</span></span><br><span class="line"><span class="comment">#od[match(leafs[,1],nodeMap[,1])]=nrow(leafs):1</span></span><br><span class="line"><span class="comment">#nodeMap=nodeMap[order(od),]</span></span><br><span class="line"><span class="comment">#head(nodeMap)</span></span><br><span class="line"><span class="comment">#geneTree$labels</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#match(colnames(dissTOM)[geneTree$order],nodeMap[,1])</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#which(nodeMap[,3]=='')</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#print(geneTree$height)</span></span><br><span class="line"><span class="comment">#class(dend)</span></span><br><span class="line"><span class="comment">#summary(dend)</span></span><br><span class="line"><span class="comment">#attr(dend,x = length)</span></span><br><span class="line"><span class="comment">#plot(geneTree,hang = -1, cex = 0.6)</span></span><br><span class="line"><span class="comment">#plot(as.phylo(geneTree), type = "fan")</span></span><br><span class="line"><span class="comment">#dend_label=partition_leaves(dend)</span></span><br><span class="line"><span class="comment">#dend_height=get_nodes_attr(dend,'height')</span></span><br><span class="line"><span class="comment">#nodeMap=rbind()</span></span><br><span class="line"><span class="comment">#for(i in length(dend_height):1)&#123;</span></span><br><span class="line"><span class="comment">#  cls=dend_label[[i]]</span></span><br><span class="line"><span class="comment">#  node1=paste0(cls,collapse = '\t')</span></span><br><span class="line"><span class="comment">#  snode1=''</span></span><br><span class="line"><span class="comment">#  snode2=''</span></span><br><span class="line"><span class="comment">#  if(length(cls)&gt;1)&#123;</span></span><br><span class="line"><span class="comment">#    node1=openssl::md5(node1)</span></span><br><span class="line"><span class="comment">#    snode1=paste0('C',i+1)</span></span><br><span class="line"><span class="comment">#    node2_ts=dend_label[[i+1]]</span></span><br><span class="line"><span class="comment">#    cls1=cls[which(!cls%in%node2_ts)]</span></span><br><span class="line"><span class="comment">#    node2=paste0(cls1,collapse = '\t')</span></span><br><span class="line"><span class="comment">#    if(length(cls1)&gt;1)&#123;</span></span><br><span class="line"><span class="comment">#      node2=openssl::md5(node2)</span></span><br><span class="line"><span class="comment">#    &#125;</span></span><br><span class="line"><span class="comment">#    snode2=nodeMap[which(nodeMap[,1]==node2),2]</span></span><br><span class="line"><span class="comment">#  &#125;</span></span><br><span class="line"><span class="comment">#  nodeMap=rbind(nodeMap,c(node1,paste0('C',i),snode1,snode2,dend_height[i]))</span></span><br><span class="line"><span class="comment">#&#125;</span></span><br><span class="line"><span class="comment">#nodeMap[which(nodeMap[,3]!=''),1]=''</span></span><br><span class="line"><span class="comment">#od=rep(10000,nrow(nodeMap))</span></span><br><span class="line"><span class="comment">#od[match(geneTree$labels[geneTree$order],nodeMap[,1])]=1:length(geneTree$labels)</span></span><br><span class="line"><span class="comment">#nodeMap=nodeMap[order(od),]</span></span><br><span class="line">write.table(nodeMap,file = paste0(outFolder,<span class="string">'/wgcna_dissTOM_tree.txt'</span>),quote = <span class="literal">F</span>,row.names = <span class="literal">F</span>,sep = <span class="string">'\t'</span>)</span><br><span class="line"><span class="comment">#head(nodeMap)</span></span><br><span class="line">logs=c(logs,paste0(<span class="string">'saved hclust dissTOM tree'</span>))</span><br><span class="line">logs=c(logs,paste0(<span class="string">'exporting module network data'</span>))</span><br><span class="line"><span class="comment">#colnames(TOM)</span></span><br><span class="line"><span class="comment">#row.names(TOM)</span></span><br><span class="line">cyt = exportNetworkToCytoscape(TOM,threshold =<span class="number">0</span>)</span><br><span class="line"><span class="comment">#head(cyt$edgeData)</span></span><br><span class="line">tidyfst::export_fst(cyt$edgeData[,<span class="number">1</span>:<span class="number">3</span>],path = paste0(outFolder,<span class="string">'/wgcna_module_net.fst'</span>)) </span><br><span class="line">cytq=quantile(cyt$edgeData[,<span class="number">3</span>])</span><br><span class="line">logs=c(logs,paste0(<span class="string">'saved module network data:'</span>,paste0(cytq,collapse = <span class="string">','</span>)))</span><br><span class="line">save(filter.exp,adjacency,dissTOM,geneTree,cytq,cutPower,file = paste0(outFolder,<span class="string">'/wgcna_TOM.RData'</span>))</span><br><span class="line">logs=c(logs,paste0(<span class="string">'saved TOM data'</span>))</span><br><span class="line">&#125;</span><br><span class="line">  </span><br><span class="line"><span class="comment">#load(paste0('/pub1/data/mg_projects/projects/web_script/tool_runing/97d971bc04475222ec72f4a05e66910f/wgcna_TOM.RData'))</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#  plotDendroAndColors(geneTree,dynamicColors ,</span></span><br><span class="line"><span class="comment">#                      "Module colors",</span></span><br><span class="line"><span class="comment">#                      dendroLabels = FALSE, hang = 0.03,</span></span><br><span class="line"><span class="comment">#                      addGuide = TRUE, guideHang = 0.05)</span></span><br><span class="line"><span class="comment">#################cal module#################</span></span><br><span class="line"><span class="comment">#dim(dissTOM)</span></span><br><span class="line"><span class="comment">#tom_tree=readMatrix(paste0(oldFolder,'/wgcna_dissTOM_tree.txt'),row=F)</span></span><br><span class="line"><span class="comment">#head(tom_tree)</span></span><br><span class="line"><span class="comment">#setdiff(colnames(filter.exp),unique(tom_tree[,1]))[1:10]</span></span><br><span class="line"><span class="comment">#setdiff(unique(tom_tree[,1]),colnames(filter.exp))[1:10]</span></span><br><span class="line"></span><br><span class="line">dynamicMods = cutreeDynamic(dendro = geneTree, distM = dissTOM</span><br><span class="line">                            ,deepSplit = as.numeric(deepSplit), pamRespectsDendro = <span class="literal">FALSE</span>,</span><br><span class="line">                            minClusterSize = as.numeric(minModuleSize))</span><br><span class="line">dynamicColors = labels2colors(dynamicMods)</span><br><span class="line"><span class="comment">#MEList = moduleEigengenes(filter.exp, colors = dynamicColors)</span></span><br><span class="line">merge = mergeCloseModules(filter.exp, dynamicColors, cutHeight = as.numeric(mergeCutHeight), verbose = <span class="number">3</span>)</span><br><span class="line"><span class="comment"># The merged module colors</span></span><br><span class="line">mergedColors = merge$colors</span><br><span class="line"><span class="comment">#dim(m_dat_sum)</span></span><br><span class="line"><span class="comment">#length(dynamicColors)</span></span><br><span class="line">m_dat_sum=as.data.frame(table(mergedColors))</span><br><span class="line">logs=c(logs,paste0(<span class="string">'module count:'</span>,paste0(paste0(m_dat_sum[,<span class="number">1</span>],<span class="string">'-'</span>,m_dat_sum[,<span class="number">2</span>]),collapse = <span class="string">','</span>)))</span><br><span class="line">write.table(m_dat_sum,file = paste0(outFolder,<span class="string">'/wgcna_module_summary.txt'</span>),quote = <span class="literal">F</span>,row.names = <span class="literal">F</span>,sep = <span class="string">'\t'</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">#head(cbind(Gene=colnames(filter.exp),module=mergedColors))</span></span><br><span class="line">write.table(cbind(Gene=colnames(filter.exp),dynamicModule=dynamicColors,mergeModule=mergedColors)</span><br><span class="line">            ,file = paste0(outFolder,<span class="string">'/wgcna_module.txt'</span>),quote = <span class="literal">F</span>,row.names = <span class="literal">F</span>,sep = <span class="string">'\t'</span>)</span><br><span class="line">logs=c(logs,paste0(<span class="string">'saved hclust mergedColors data'</span>))</span><br><span class="line"></span><br><span class="line">MEs=merge$newMEs</span><br><span class="line">colnames(MEs)=gsub(<span class="string">'^ME'</span>,<span class="string">''</span>,colnames(MEs))</span><br><span class="line">write.table(cbind(Tag=row.names(MEs),MEs),file = paste0(outFolder,<span class="string">'/wgcna_MEs.txt'</span>),quote = <span class="literal">F</span>,row.names = <span class="literal">F</span>,sep = <span class="string">'\t'</span>)</span><br><span class="line">logs=c(logs,paste0(<span class="string">'saved hclust ME data'</span>))</span><br><span class="line">MEDiss = <span class="number">1</span>-cor(MEs)</span><br><span class="line"></span><br><span class="line">sTree2=getClusterTree(MEDiss,dMethod = <span class="string">'pearson'</span>,hMethod = <span class="string">'average'</span>)</span><br><span class="line">write.table(sTree2,file = paste0(outFolder,<span class="string">'/wgcna_ME_tree.txt'</span>),quote = <span class="literal">F</span>,row.names = <span class="literal">F</span>,sep = <span class="string">'\t'</span>)</span><br><span class="line">m_inds=match(sTree2[which(sTree2[,<span class="number">1</span>]!=<span class="string">''</span>)],colnames(MEDiss))</span><br><span class="line">write.table(MEDiss[m_inds,m_inds],file = paste0(outFolder,<span class="string">'/wgcna_ME_tree_heat.txt'</span>),quote = <span class="literal">F</span>,row.names = <span class="literal">F</span>,sep = <span class="string">'\t'</span>)</span><br><span class="line">logs=c(logs,paste0(<span class="string">'saved hclust ME tree'</span>))</span><br><span class="line"><span class="comment">#####################Module membership########################</span></span><br><span class="line">geneModuleMembership = as.data.frame(cor(filter.exp, MEs, use = <span class="string">"p"</span>));</span><br><span class="line">MMPvalue = as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), nrow(filter.exp)));</span><br><span class="line"><span class="comment">#melt(geneModuleMembership)</span></span><br><span class="line"><span class="comment">#colnames(geneModuleMembership)</span></span><br><span class="line"><span class="comment">#cbind(colnames(filter.exp),mergedColors)</span></span><br><span class="line">mmr=reshape::melt(cbind(Tag=row.names(geneModuleMembership),geneModuleMembership),id.vars=c(<span class="string">'Tag'</span>))</span><br><span class="line">mmp=reshape::melt(cbind(Tag=row.names(MMPvalue),MMPvalue),id.vars=c(<span class="string">'Tag'</span>))</span><br><span class="line">mmr=mmr[paste0(mmr[,<span class="number">1</span>],<span class="string">'\t'</span>,mmr[,<span class="number">2</span>])%<span class="keyword">in</span>%paste0(colnames(filter.exp),<span class="string">'\t'</span>,mergedColors),]</span><br><span class="line">mmr$pvalue=mmp[match(paste0(mmr[,<span class="number">1</span>],<span class="string">'\t'</span>,mmr[,<span class="number">2</span>]),paste0(mmp[,<span class="number">1</span>],<span class="string">'\t'</span>,mmp[,<span class="number">2</span>])),<span class="number">3</span>]</span><br><span class="line">colnames(mmr)=c(<span class="string">'Gene'</span>,<span class="string">'Module'</span>,<span class="string">'R'</span>,<span class="string">'pvalue'</span>)</span><br><span class="line">mmr=mmr[match(colnames(filter.exp),mmr[,<span class="number">1</span>]),]</span><br><span class="line">write.table(mmr,file = paste0(outFolder,<span class="string">'/wgcna_MM_value.txt'</span>),quote = <span class="literal">F</span>,row.names = <span class="literal">F</span>,sep = <span class="string">'\t'</span>)</span><br><span class="line">logs=c(logs,paste0(<span class="string">'saved Module membership data'</span>))</span><br><span class="line"><span class="comment">#mergedColors</span></span><br><span class="line"><span class="keyword">if</span>(!file.exists(paste0(outFolder,<span class="string">'/clini.json'</span>)))&#123;</span><br><span class="line">  write.table(<span class="string">''</span>,file = paste0(outFolder,<span class="string">'/clini.json'</span>),quote = <span class="literal">F</span>,row.names = <span class="literal">F</span>,sep = <span class="string">'\t'</span>)</span><br><span class="line">&#125;</span><br><span class="line">tidyfst::export_fst(as.data.frame(cbind(Gene=colnames(filter.exp),mergeModule=mergedColors)),path = paste0(outFolder,<span class="string">'/wgcna_module.fst'</span>)) </span><br><span class="line">tidyfst::export_fst(as.data.frame(cbind(Sample=row.names(MEs),MEs)),path = paste0(outFolder,<span class="string">'/wgcna_MEs.fst'</span>)) </span><br><span class="line">tidyfst::export_fst(mmr,path = paste0(outFolder,<span class="string">'/wgcna_MM_value.fst'</span>)) </span><br><span class="line">tidyfst::export_fst(as.data.frame(cbind(Sample=row.names(filter.exp),filter.exp)),path = paste0(outFolder,<span class="string">'/filter.exp.fst'</span>)) </span><br><span class="line">logs=c(logs,paste0(<span class="string">'saved all module data'</span>))</span><br><span class="line"></span><br><span class="line">&#125;<span class="keyword">else</span> <span class="keyword">if</span>(method==<span class="string">'cli'</span>)&#123;</span><br><span class="line">  </span><br><span class="line">  outFolder=opt$outfile</span><br><span class="line">  cNames=unlist(data$cNames)</span><br><span class="line">  cTypes=unlist(data$cTypes)</span><br><span class="line">  cliDatas=(data$cliDatas)</span><br><span class="line">  oldFolder=unlist(data$oldFolder)</span><br><span class="line">  <span class="comment">#cliDatas[1][[2]]</span></span><br><span class="line">  </span><br><span class="line">  cliDatas=cliDatas[[<span class="number">1</span>]]</span><br><span class="line">  </span><br><span class="line">  <span class="comment">#table(cliDatas[,4])</span></span><br><span class="line">  </span><br><span class="line">  file.copy(opt$infile,paste0(oldFolder,<span class="string">'/clini.json'</span>),overwrite = <span class="literal">T</span>)</span><br><span class="line">  <span class="comment">#data&lt;-jsonlite::stream_in(file(paste0(oldFolder,'/clini.json')),pagesize = 1000)</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">#oldFolder='/pub1/data/mg_projects/projects/web_script/tool_runing/97d971bc04475222ec72f4a05e66910f'</span></span><br><span class="line">  </span><br><span class="line">  baseFolder=oldFolder</span><br><span class="line">  <span class="comment">#geneModule=tidyfst::import_fst(paste0(baseFolder,'/wgcna_module.fst'),as.data.table = F)</span></span><br><span class="line">  mes=tidyfst::import_fst(paste0(baseFolder,<span class="string">'/wgcna_MEs.fst'</span>),as.data.table = <span class="literal">F</span>)</span><br><span class="line">  <span class="comment">#head(mes)</span></span><br><span class="line">  row.names(mes)=mes[,<span class="number">1</span>]</span><br><span class="line">  mes=mes[,-<span class="number">1</span>]</span><br><span class="line">  cmp=intersect(row.names(mes),cliDatas[,<span class="number">1</span>])</span><br><span class="line">  mes=mes[match(cmp,row.names(mes)),]</span><br><span class="line">  cliDatas=cliDatas[match(cmp,cliDatas[,<span class="number">1</span>]),]</span><br><span class="line">  nColNames=c()</span><br><span class="line">  nColSubNames=c()</span><br><span class="line">  nColData=cbind()</span><br><span class="line">  <span class="keyword">for</span>(i <span class="keyword">in</span> <span class="number">1</span>:length(cTypes))&#123;</span><br><span class="line">    <span class="keyword">if</span>(cTypes[i]&gt;<span class="number">0</span>)&#123;</span><br><span class="line">      nColNames=c(nColNames,cNames[i])  </span><br><span class="line">      nColSubNames=c(nColSubNames,cNames[i])</span><br><span class="line">      nColData=cbind(nColData,cliDatas[,i+<span class="number">1</span>])</span><br><span class="line">    &#125;<span class="keyword">else</span>&#123;</span><br><span class="line">      dt=cliDatas[,i+<span class="number">1</span>]</span><br><span class="line">      udt=unique(dt)</span><br><span class="line">      udt=udt[which(!is.na(udt)&amp;udt!=<span class="string">''</span>)]</span><br><span class="line">      <span class="keyword">for</span>(u <span class="keyword">in</span> udt)&#123;</span><br><span class="line">        ndt=ifelse(dt==u,<span class="number">1</span>,<span class="number">0</span>)</span><br><span class="line">        ndt[which(is.na(dt))]=<span class="literal">NA</span></span><br><span class="line">        nColData=cbind(nColData,ndt)</span><br><span class="line">        nColSubNames=c(nColSubNames,u)</span><br><span class="line">        nColNames=c(nColNames,cNames[i])</span><br><span class="line">      &#125;</span><br><span class="line">    &#125;</span><br><span class="line">  &#125;</span><br><span class="line">  nColData=apply(nColData, <span class="number">2</span>, as.numeric)</span><br><span class="line">  <span class="comment">#nColData=nColData[,1]</span></span><br><span class="line">  cliModuleMembership = as.data.frame(cor(mes,nColData, use = <span class="string">"p"</span>));</span><br><span class="line">  MMPvalue = as.data.frame(corPvalueStudent(as.matrix(cliModuleMembership), nrow(mes)));</span><br><span class="line">  colnames(cliModuleMembership)=paste0(nColNames,<span class="string">'('</span>,nColSubNames,<span class="string">')'</span>)</span><br><span class="line">  colnames(MMPvalue)=paste0(nColNames,<span class="string">'('</span>,nColSubNames,<span class="string">')'</span>)</span><br><span class="line">  </span><br><span class="line">  cNs=rbind(c(-<span class="number">1</span>,<span class="string">''</span>,nColNames),c(-<span class="number">1</span>,<span class="string">''</span>,nColSubNames))</span><br><span class="line">  colnames(cNs)=c(<span class="string">'isPvalue'</span>,<span class="string">'Module'</span>,colnames(cliModuleMembership))</span><br><span class="line">  cdt2=rbind(cbind(isPvalue=<span class="number">0</span>,Module=row.names(cliModuleMembership),cliModuleMembership),</span><br><span class="line">        cbind(isPvalue=<span class="number">1</span>,Module=row.names(MMPvalue),MMPvalue))</span><br><span class="line">  cdt2=apply(cdt2, <span class="number">2</span>,as.character)</span><br><span class="line">  </span><br><span class="line">  write.table(rbind(as.data.frame(cNs),as.data.frame(cdt2)),file = paste0(baseFolder,<span class="string">'/wgcna_ME_cli_cor.txt'</span>)</span><br><span class="line">              ,quote = <span class="literal">F</span>,row.names = <span class="literal">F</span>,sep = <span class="string">'\t'</span>)</span><br><span class="line">  </span><br><span class="line">  <span class="comment">#rbind(cliModuleMembership,MMPvalue)</span></span><br><span class="line">  <span class="comment">#geneModule=tidyfst::import_fst(paste0(baseFolder,'/wgcna_module.fst'),as.data.table = F)</span></span><br><span class="line">  mmr=tidyfst::import_fst(paste0(baseFolder,<span class="string">'/wgcna_MM_value.fst'</span>),as.data.table = <span class="literal">F</span>)</span><br><span class="line">  filter.exp=tidyfst::import_fst(paste0(baseFolder,<span class="string">'/filter.exp.fst'</span>),as.data.table = <span class="literal">F</span>)</span><br><span class="line">  row.names(filter.exp)=filter.exp[,<span class="number">1</span>]</span><br><span class="line">  filter.exp=filter.exp[,-<span class="number">1</span>]</span><br><span class="line">  filter.exp=filter.exp[match(cmp,row.names(filter.exp)),]</span><br><span class="line">  <span class="comment">#head(filter.exp[,1:10])</span></span><br><span class="line">  </span><br><span class="line">  cliGeneMembership = as.data.frame(cor(filter.exp,nColData, use = <span class="string">"p"</span>));</span><br><span class="line">  MMPvalue = as.data.frame(corPvalueStudent(as.matrix(cliGeneMembership), nrow(mes)));</span><br><span class="line">  colnames(cliGeneMembership)=paste0(nColNames,<span class="string">'('</span>,nColSubNames,<span class="string">')'</span>)</span><br><span class="line">  colnames(MMPvalue)=paste0(nColNames,<span class="string">'('</span>,nColSubNames,<span class="string">')'</span>)</span><br><span class="line">  <span class="comment">#head(gs)</span></span><br><span class="line">  gs=cbind(mmr,cliGeneMembership[match(mmr[,<span class="number">1</span>],row.names(cliGeneMembership)),]</span><br><span class="line">           ,MMPvalue[match(mmr[,<span class="number">1</span>],row.names(MMPvalue)),])</span><br><span class="line">  write.table(gs,file = paste0(baseFolder,<span class="string">'/wgcna_MM_GS_cor.txt'</span>)</span><br><span class="line">              ,quote = <span class="literal">F</span>,row.names = <span class="literal">F</span>,sep = <span class="string">'\t'</span>)  </span><br><span class="line">  <span class="keyword">if</span>(ncol(cliGeneMembership)==<span class="number">1</span>)&#123;</span><br><span class="line">    ngm=cbind(abs(cliGeneMembership[match(mmr[,<span class="number">1</span>],row.names(cliGeneMembership)),]))</span><br><span class="line">    colnames(ngm)=colnames(cliGeneMembership)</span><br><span class="line">  &#125;<span class="keyword">else</span>&#123;</span><br><span class="line">    ngm=abs(cliGeneMembership[match(mmr[,<span class="number">1</span>],row.names(cliGeneMembership)),])</span><br><span class="line">  &#125;</span><br><span class="line">  <span class="comment">#head(ngm)</span></span><br><span class="line">  all_cr=rbind()</span><br><span class="line">  <span class="keyword">for</span>(u <span class="keyword">in</span> unique(as.character(mmr[,<span class="number">2</span>])))&#123;</span><br><span class="line">    <span class="comment">#u=unique(as.character(mmr[,2]))[1]</span></span><br><span class="line">    t_inds=which(mmr[,<span class="number">2</span>]==u)</span><br><span class="line">    <span class="keyword">if</span>(length(t_inds)&gt;<span class="number">3</span>)&#123;</span><br><span class="line">      cr=as.data.frame(cor(cbind(abs(mmr[t_inds,<span class="number">3</span>])),ngm[t_inds,], use = <span class="string">"p"</span>));</span><br><span class="line">      cr_p=as.data.frame(corPvalueStudent(as.matrix(cr), length(t_inds)));</span><br><span class="line">      cdt=cbind(M=u,Clinical=<span class="number">1</span>:ncol(ngm),t(rbind(cr,cr_p)))</span><br><span class="line">      colnames(cdt)=c(<span class="string">'Module'</span>,<span class="string">'Clinical'</span>,<span class="string">'R'</span>,<span class="string">'P'</span>)</span><br><span class="line">      all_cr=rbind(all_cr,cdt)</span><br><span class="line">    &#125;<span class="keyword">else</span>&#123;</span><br><span class="line">      cdt=cbind(rep(u,ncol(ngm)),<span class="number">1</span>:ncol(ngm),rep(<span class="literal">NA</span>,ncol(ngm)),rep(<span class="literal">NA</span>,ncol(ngm)))</span><br><span class="line">      colnames(cdt)=c(<span class="string">'Module'</span>,<span class="string">'Clinical'</span>,<span class="string">'R'</span>,<span class="string">'P'</span>)</span><br><span class="line">      all_cr=rbind(all_cr,cdt)</span><br><span class="line">    &#125;</span><br><span class="line">  &#125;</span><br><span class="line">  write.table(all_cr,file = paste0(baseFolder,<span class="string">'/wgcna_MM_GS_cor_RP.txt'</span>),quote = <span class="literal">F</span>,row.names = <span class="literal">F</span>,sep = <span class="string">'\t'</span>) </span><br><span class="line">  </span><br><span class="line">&#125;<span class="keyword">else</span> <span class="keyword">if</span>(method==<span class="string">'export'</span>)&#123;</span><br><span class="line">  </span><br><span class="line">  weight=as.numeric(unlist(data$weight))</span><br><span class="line">  modules=unlist(data$modules)</span><br><span class="line">  fileTree=as.numeric(unlist(data$fileTree))</span><br><span class="line">  baseFolder=unlist(data$oldFolder)</span><br><span class="line">  <span class="comment">#cliDatas[1][[2]]</span></span><br><span class="line">  <span class="comment">#table(cliDatas[,4])</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">#data&lt;-jsonlite::stream_in(file(paste0(oldFolder,'/export.json')),pagesize = 1000)</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment">#oldFolder='/pub1/data/mg_projects/projects/web_script/tool_runing/97d971bc04475222ec72f4a05e66910f'</span></span><br><span class="line">  <span class="keyword">if</span>(fileTree==<span class="number">1</span>)&#123;</span><br><span class="line">    al.dt=c()</span><br><span class="line">    <span class="keyword">for</span>(fl <span class="keyword">in</span> dir(oldFolder))&#123;</span><br><span class="line">      <span class="keyword">if</span>(length(grep(<span class="string">'.json$'</span>,fl))&gt;<span class="number">0</span>|length(grep(<span class="string">'.log$'</span>,fl))&gt;<span class="number">0</span>)&#123;</span><br><span class="line">      &#125;<span class="keyword">else</span>&#123;</span><br><span class="line">        al.dt=c(al.dt,fl)    </span><br><span class="line">      &#125;</span><br><span class="line">    &#125;</span><br><span class="line">    write.table(al.dt,file = paste0(opt$outfile,<span class="string">'/file.tree'</span>),quote = <span class="literal">F</span>,row.names = <span class="literal">F</span>,col.names = <span class="literal">F</span>,sep = <span class="string">'\t'</span>)</span><br><span class="line">  &#125;<span class="keyword">else</span>&#123;</span><br><span class="line">    file.copy(opt$infile,paste0(oldFolder,<span class="string">'/export.json'</span>),overwrite = <span class="literal">T</span>)</span><br><span class="line">    <span class="comment">#weight=0.2</span></span><br><span class="line">    <span class="comment">#modules=c('darkgreen','turquoise');</span></span><br><span class="line">    net=tidyfst::import_fst(paste0(baseFolder,<span class="string">'/wgcna_module_net.fst'</span>),as.data.table = <span class="literal">F</span>)</span><br><span class="line">    net=net[net[,<span class="number">3</span>]&gt;weight,]</span><br><span class="line">    <span class="comment">#dim(net)</span></span><br><span class="line">    geneModule=tidyfst::import_fst(paste0(baseFolder,<span class="string">'/wgcna_module.fst'</span>),as.data.table = <span class="literal">F</span>)</span><br><span class="line">    genes=as.character(geneModule[which(as.character(geneModule[,<span class="number">2</span>])%<span class="keyword">in</span>%modules),<span class="number">1</span>])</span><br><span class="line">    t1.inds=which(as.character(net[,<span class="number">1</span>])%<span class="keyword">in</span>%genes)</span><br><span class="line">    t2.inds=which(as.character(net[,<span class="number">2</span>])%<span class="keyword">in</span>%genes)</span><br><span class="line">    t.inds=intersect(t1.inds,t2.inds)</span><br><span class="line">    <span class="keyword">if</span>(length(t.inds)==<span class="number">0</span>)&#123;</span><br><span class="line">      write.table(rbind(colnames(net)),file = paste0(baseFolder,<span class="string">'/network_edge.txt'</span>),quote = <span class="literal">F</span>,row.names = <span class="literal">F</span></span><br><span class="line">                  ,sep = <span class="string">'\t'</span>,col.names = <span class="literal">F</span>)</span><br><span class="line">      write.table(rbind(colnames(geneModule)),file = paste0(baseFolder,<span class="string">'/network_node.txt'</span>),quote = <span class="literal">F</span>,row.names = <span class="literal">F</span></span><br><span class="line">                  ,sep = <span class="string">'\t'</span>,col.names = <span class="literal">F</span>)</span><br><span class="line">    &#125;<span class="keyword">else</span>&#123;</span><br><span class="line">      net=net[t.inds,]</span><br><span class="line">      ug=unique(c(as.character(net[,<span class="number">1</span>]),as.character(net[,<span class="number">2</span>])))</span><br><span class="line">      node=geneModule[match(ug,as.character(geneModule[,<span class="number">1</span>])),]</span><br><span class="line">    </span><br><span class="line">      write.table(net,file = paste0(baseFolder,<span class="string">'/network_edge.txt'</span>),quote = <span class="literal">F</span>,row.names = <span class="literal">F</span>,col.names = <span class="literal">T</span>,sep = <span class="string">'\t'</span>)</span><br><span class="line">      write.table(node,file = paste0(baseFolder,<span class="string">'/network_node.txt'</span>),quote = <span class="literal">F</span>,row.names = <span class="literal">F</span>,col.names = <span class="literal">T</span>,sep = <span class="string">'\t'</span>)</span><br><span class="line">    &#125;</span><br><span class="line">  &#125;</span><br><span class="line">  </span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">&#125;,error = <span class="keyword">function</span>(e) &#123;</span><br><span class="line">  print(conditionMessage(e))</span><br><span class="line">  logs=c(logs,paste0(<span class="string">'error:'</span>,conditionMessage(e)))</span><br><span class="line">&#125;, finally = &#123;</span><br><span class="line">  write.table(logs,file = paste0(opt$outfile,<span class="string">'/run.log'</span>),quote = <span class="literal">F</span>,row.names = <span class="literal">T</span>,col.names = <span class="literal">T</span>,sep = <span class="string">'\t'</span>)</span><br><span class="line">&#125;)</span><br></pre></td></tr></table></figure>
<p>input.json如下</p>
<figure class="highlight json"><table><tr><td class="code"><pre><span class="line">&#123;<span class="attr">"exp_path"</span>:<span class="string">"/home/win/4T/GeneDL/OSDrought_GCNAT_Link/plot/multispecies_WGCNA/indica/data/indica_CD_WGCNA.txt"</span>,<span class="attr">"geneMode"</span>:<span class="string">"mad"</span>,<span class="attr">"geneModeValue"</span>:<span class="string">"50"</span>,<span class="attr">"sampleCut"</span>:<span class="string">"1"</span>,<span class="attr">"netType"</span>:<span class="string">"unsigned"</span>,<span class="attr">"oldFolder"</span>:<span class="string">""</span>,<span class="attr">"minModuleSize"</span>:<span class="string">"30"</span>,<span class="attr">"deepSplit"</span>:<span class="string">"3"</span>,<span class="attr">"mergeCutHeight"</span>:<span class="string">"0.25"</span>,<span class="attr">"method"</span>:<span class="string">"module"</span>&#125;</span><br></pre></td></tr></table></figure>
<p>后续就可以用数据做深度学习模型了~</p>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/RNASeq数据下载处理全流程/" rel="tag">✿✿✿ RNASeq数据下载处理全流程</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2019/01/07/opencv/" rel="next" title="opencv基础">
                <i class="fa fa-chevron-left"></i> opencv基础
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2019/01/07/服务器配置LDAP/" rel="prev" title="服务器配置LDAP">
                服务器配置LDAP <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </div>

<div>
    
    <div>
    
        <div style="text-align:center;color: #aaa;font-size:14px;margin-top:2rem;">------ 本文结束 🎉🎉 谢谢观看  ------</div>
    
</div>

    
</div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  
    <div class="comments" id="comments">
      <div id="disqus_thread">
        <noscript>
          Please enable JavaScript to view the
          <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a>
        </noscript>
      </div>
    </div>

  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
      <div id="sidebar-dimmer"></div>
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            文章目录
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            站点概览
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope="" itemtype="http://schema.org/Person">
            
              <img class="site-author-image" itemprop="image" src="/uploads/avatar.jpg" alt="sylvia">
            
              <p class="site-author-name" itemprop="name">sylvia</p>
              <p class="site-description motion-element" itemprop="description">君がいるだから、今の僕は幸せです。今日もありがとう。</p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives">
              
                  <span class="site-state-item-count">114</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">16</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">104</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          
            <div class="feed-link motion-element">
              <a href="/atom.xml" rel="alternate">
                <i class="fa fa-rss"></i>
                RSS
              </a>
            </div>
          

          
            <div class="links-of-author motion-element">
                
                  <span class="links-of-author-item">
                    <a href="mailto:836753560@qq.com" target="_blank" title="E-Mail">
                      
                        <i class="fa fa-fw fa-envelope"></i>E-Mail</a>
                  </span>
                
                  <span class="links-of-author-item">
                    <a href="skype:live:836753560?call|chat" target="_blank" title="Skype">
                      
                        <i class="fa fa-fw fa-skype"></i>Skype</a>
                  </span>
                
            </div>
          

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#RNASeq数据下载处理全流程"><span class="nav-number">1.</span> <span class="nav-text">RNASeq数据下载处理全流程</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#一、下载数据"><span class="nav-number">1.1.</span> <span class="nav-text">一、下载数据</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#二、参考基因组索引-建立"><span class="nav-number">1.2.</span> <span class="nav-text">二、参考基因组索引 建立</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#三、数据全流程自动处理"><span class="nav-number">1.3.</span> <span class="nav-text">三、数据全流程自动处理</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#四、定量文件拼接与数据标准化"><span class="nav-number">1.4.</span> <span class="nav-text">四、定量文件拼接与数据标准化</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#五、将数据从一个服务器传到另一个服务器"><span class="nav-number">1.5.</span> <span class="nav-text">五、将数据从一个服务器传到另一个服务器</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#六、差异表达、富集分析、WGCNA分析"><span class="nav-number">1.6.</span> <span class="nav-text">六、差异表达、富集分析、WGCNA分析</span></a></li></ol></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; <span itemprop="copyrightYear">2024</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">sylvia</span>

  
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item-icon">
      <i class="fa fa-area-chart"></i>
    </span>
    
      <span class="post-meta-item-text">Site words total count&#58;</span>
    
    <span title="Site words total count">144.5k</span>
  
</div>

<!--
  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Muse</a> v5.1.4</div>

-->



        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
          <span id="scrollpercent"><span>0</span>%</span>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  
    <script type="text/javascript" src="//cdn.jsdelivr.net/jquery/2.1.3/jquery.min.js"></script>
  

  
  
    <script type="text/javascript" src="//cdn.jsdelivr.net/fastclick/1.0.6/fastclick.min.js"></script>
  

  
  
    <script type="text/javascript" src="//cdn.jsdelivr.net/jquery.lazyload/1.9.3/jquery.lazyload.min.js"></script>
  

  
  
    <script type="text/javascript" src="//cdn.jsdelivr.net/velocity/1.2.3/velocity.min.js"></script>
  

  
  
    <script type="text/javascript" src="//cdn.jsdelivr.net/velocity/1.2.3/velocity.ui.min.js"></script>
  

  
  
    <script type="text/javascript" src="//cdn.jsdelivr.net/fancybox/2.1.5/jquery.fancybox.pack.js"></script>
  


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.4"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.4"></script>



  
  

  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.4"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.4"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.4"></script>



  


  

    
      <script id="dsq-count-scr" src="https://sylvia.disqus.com/count.js" async></script>
    

    
      <script type="text/javascript">
        var disqus_config = function () {
          this.page.url = 'https://janvia.github.io/2019/01/07/RNASeq数据下载处理全流程/';
          this.page.identifier = '2019/01/07/RNASeq数据下载处理全流程/';
          this.page.title = 'RNASeq数据下载处理全流程';
        };
        var d = document, s = d.createElement('script');
        s.src = 'https://sylvia.disqus.com/embed.js';
        s.setAttribute('data-timestamp', '' + +new Date());
        (d.head || d.body).appendChild(s);
      </script>
    

  




	





  














  

  <script type="text/javascript">
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (/json$/i.test(search_path)) {
      isXml = false;
    }
    var path = "/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url);
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x" /></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x" /></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script>





  

  

  

  
  

  
  
    <script type="text/x-mathjax-config">
      MathJax.Hub.Config({
        tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
        }
      });
    </script>

    <script type="text/x-mathjax-config">
      MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax(), i;
        for (i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
        }
      });
    </script>
    <script type="text/javascript" src="//cdn.bootcss.com/mathjax/2.7.1/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
  


  

  

<script src="/live2dw/lib/L2Dwidget.min.js?0c58a1486de42ac6cc1c59c7d98ae887"></script><script>L2Dwidget.init({"pluginRootPath":"live2dw/","pluginJsPath":"lib/","pluginModelPath":"assets/","tagMode":false,"debug":false,"model":{"jsonPath":"live2d-widget-model-tsumiki"},"display":{"superSample":2,"position":"left","width":75,"height":120,"hOffset":0,"vOffset":-5},"mobile":{"show":true,"scale":0.1},"react":{"opacityDefault":0.7,"opacityOnHover":0.2},"log":false});</script><script async>window.onload=function(){var a=document.createElement('script'),b=document.getElementsByTagName('script')[0];a.type='text/javascript',a.async=!0,a.src='/sw-register.js?v='+Date.now(),b.parentNode.insertBefore(a,b)};</script></body></html>